---
title: "scRNA-seq analysis of MCF-7 cells treated with TAM for 0, 3, 6, 9 weeks"
author: "Keita Iida"
date: "`r Sys.Date()`"
output:
#  rmarkdown::github_document
  rmarkdown::html_vignette:
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```



<br>

# Introduction
In this vignette, single-cell RNA sequencing (scRNA-seq) data obtained from
MCF-7 cells with tamoxifen treatment for 0 (control), 3, 6, and 9 weeks
(Magi et al., Sci Rep, 2021) are analyzed.



<br>

# Load the raw count data
Load the data.
```{r, eval = FALSE}
# Count matrix
f <- "data/tamoxifen_MCF7_scRNAseq/DrMagi_F0551_GA06071-4.genematrix.csv"
data <- read.csv(file = f, sep = ",", header = TRUE, row.names = 1)
data <- t(data)

# Cell barcodes
f <- "data/tamoxifen_MCF7_scRNAseq/DrMagi_F0551_GA06071-4.report_red.csv"
meta <- read.csv(file = f, sep = ",", header = TRUE, row.names = 1)

# Mitochondrial genes
f <- "data/mitochondrial_genes.txt"
mt_genes <- read.csv(file = f, sep = "\t", header = TRUE)
colnames(mt_genes) <- c("ensemble", "symbol", "description")
```

Remove cells from the meta data, which are not registered in `data`.
```{r, eval = FALSE}
cells <- colnames(data)
meta <- meta[which(rownames(meta) %in% cells), ]
```

Check the order of cell barcodes.
```{r, eval = FALSE}
identical(colnames(data), rownames(meta))
```

```
[1] TRUE
```

Make a dictionary converting Ensembl IDs to gene symbols.
```{r, eval = FALSE}
dict <- AnnotationDbi::select(org.Hs.eg.db::org.Hs.eg.db,
                              key = rownames(data),
                              columns = c("SYMBOL"), keytype = "ENSEMBL")
for(i in seq_along(dict$ENSEMBL)){
  dict$SYMBOL[i] <-
    ifelse(dict$ENSEMBL[i] %in% mt_genes$ensembl,
           mt_genes[which(mt_genes$ensembl == dict$ENSEMBL[i]),]$symbol,
           dict$SYMBOL[i])
}
dict <- dict[!duplicated(dict$ENSEMBL), ]
dict <- dict[-which(is.na(dict$SYMBOL)), ]
```

Change Ensembl IDs into gene symbols for the data.
```{r, eval = FALSE}
data <- data[which(rownames(data) %in% dict$ENSEMBL), ]
data <- data[dict$ENSEMBL, ]
rownames(data) <- make.unique(dict$SYMBOL)
```

Create a SingleCellExperiment object.
```{r, eval = FALSE}
mcf7 <- SingleCellExperiment(assays = list(counts = data),
                             rowData = data.frame(ensemble = dict$ENSEMBL),
                             colData = data.frame(type = meta$Type))
```

```{r, eval = FALSE}
dim(mcf7)
```
```
[1] 22053  1108
```

Save the data.
```{r, eval = FALSE}
saveRDS(mcf7, file = "backup/01_001_mcf7_sce.rds")

#Load the data.
#mcf7 <- readRDS("backup/01_001_mcf7_sce.rds")
```



<br>

# Preprocessing
## Control data quality
Remove genes and cells with low quality, by processing the following steps:

1. remove genes based on expression profiles across cells,
2. remove cells based on the numbers of reads and nonzero expressed genes,

First of all, add metadata for both genes and cells using ASURAT function
`add_metadata()`.
```{r, eval = FALSE}
mcf7 <- ASURAT::add_metadata(sce = mcf7, mitochondria_symbol = "^MT-")
```



<br>

### Remove genes based on expression profiles
ASURAT function `remove_variables()` removes genes data such that the numbers of
non-zero expressing cells are less than `min_nsamples`.
```{r, eval = FALSE}
mcf7 <- ASURAT::remove_variables(sce = mcf7, min_nsamples = 10)
```



<br>

### Remove cells based on expression profiles
Qualities of cell data are confirmed based on proper visualization of
`colData(sce)`.
```{r, eval = FALSE}
title <- "MCF-7"
df <- data.frame(x = colData(mcf7)$nReads, y = colData(mcf7)$nGenes)
p <- ggplot2::ggplot() +
  ggplot2::geom_point(aes(x = df$x, y = df$y), size = 1, alpha = 1) +
  ggplot2::scale_x_log10(
    breaks = scales::trans_breaks("log10", function(x) 10^x),
    labels = scales::trans_format("log10", scales::math_format(10^.x))
  ) +
  ggplot2::labs(title = title, x = "Number of reads", y = "Number of genes") +
  ggplot2::theme_classic(base_size = 20) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, size = 20))
filename <- "figures/figure_01_0010.png"
ggplot2::ggsave(file = filename, plot = p, dpi = 80, width = 4.8, height = 5)

df <- data.frame(x = colData(mcf7)$nReads, y = colData(mcf7)$percMT)
p <- ggplot2::ggplot() +
  ggplot2::geom_point(aes(x = df$x, y = df$y), size = 1, alpha = 1) +
  ggplot2::scale_x_log10(
    breaks = scales::trans_breaks("log10", function(x) 10^x),
    labels = scales::trans_format("log10", scales::math_format(10^.x))
  ) +
  ggplot2::labs(title = title, x = "Number of reads", y = "Perc of MT reads") +
  ggplot2::theme_classic(base_size = 20) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, size = 20))
filename <- "figures/figure_01_0011.png"
ggplot2::ggsave(file = filename, plot = p, dpi = 80, width = 4.6, height = 5)
```

<img src="figures/figure_01_0010.png" width="200px">
<img src="figures/figure_01_0011.png" width="200px">

ASURAT function `remove_samples()` removes cell data by setting cutoff values
for the metadata.
```{r, eval = FALSE}
mcf7 <- ASURAT::remove_samples(sce = mcf7, min_nReads = 1000, max_nReads = 2e+5,
                               min_nGenes = 0, max_nGenes = 1e+10,
                               min_percMT = 0, max_percMT = 55)
```

```{r, eval = FALSE}
dim(mcf7)
```
```
[1] 14476  1000
```

```{r, eval = FALSE}
title <- "MCF-7"
df <- data.frame(x = colData(mcf7)$nReads, y = colData(mcf7)$nGenes)
p <- ggplot2::ggplot() +
  ggplot2::geom_point(aes(x = df$x, y = df$y), size = 1, alpha = 1) +
  ggplot2::scale_x_log10(
    breaks = scales::trans_breaks("log10", function(x) 10^x),
    labels = scales::trans_format("log10", scales::math_format(10^.x))
  ) +
  ggplot2::labs(title = title, x = "Number of reads", y = "Number of genes") +
  ggplot2::theme_classic(base_size = 20) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, size = 20))
filename <- "figures/figure_01_0012.png"
ggplot2::ggsave(file = filename, plot = p, dpi = 80, width = 4.8, height = 5)

df <- data.frame(x = colData(mcf7)$nReads, y = colData(mcf7)$percMT)
p <- ggplot2::ggplot() +
  ggplot2::geom_point(aes(x = df$x, y = df$y), size = 1, alpha = 1) +
  ggplot2::scale_x_log10(
    breaks = scales::trans_breaks("log10", function(x) 10^x),
    labels = scales::trans_format("log10", scales::math_format(10^.x))
  ) +
  ggplot2::labs(title = title, x = "Number of reads", y = "Perc of MT reads") +
  ggplot2::theme_classic(base_size = 20) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, size = 20))
filename <- "figures/figure_01_0013.png"
ggplot2::ggsave(file = filename, plot = p, dpi = 80, width = 4.6, height = 5)
```

<img src="figures/figure_01_0012.png" width="200px">
<img src="figures/figure_01_0013.png" width="200px">

Save the data.
```{r, eval = FALSE}
saveRDS(mcf7, file = "backup/01_002_mcf7_dataqc.rds")

#Load the data.
#mcf7 <- readRDS("backup/01_002_mcf7_dataqc.rds")
```



<br>

## Normalize the data
Normalize the data using bayNorm, where we use `Prior_type = "LL"` preventing
the impact of data integration from becoming too strong.
```{r, eval = FALSE}
mat <- as.matrix(assay(mcf7, "counts"))
BETA <- bayNorm::BetaFun(Data = mat, MeanBETA = 0.06)
bayout <- bayNorm::bayNorm(Data = mat, Conditions = mcf7$type,
                           BETA_vec = BETA[["BETA"]],
                           Prior_type = "LL", mode_version = TRUE)
```

Save the data.
```{r, eval = FALSE}
saveRDS(bayout, file = "backup/01_003_mcf7_baynorm.rds")

#Load the data.
#bayout <- readRDS("backup/01_003_mcf7_baynorm.rds")
```



<br>

# Analysis using Seurat
Create a Seurat object.
```{r, eval = FALSE}
mat <- do.call("cbind", bayout[["Bay_out_list"]])
mat <- mat[, colnames(mcf7)]
mcf7 <- Seurat::CreateSeuratObject(counts = mat,
                                   meta.data = as.data.frame(colData(mcf7)),
                                   project = "MCF7")
```



<br>

## Perform Seurat preprocessing
According to the Seurat protocol, normalize data, perform variance stabilizing
transform by setting the number of variable feature, scale data, and reduce
dimension using principal component analysis.
```{r, eval = FALSE}
# Normalization
mcf7 <- Seurat::NormalizeData(mcf7, normalization.method = "LogNormalize")
# Variance stabilizing transform
n <- round(5 * ncol(mcf7))
mcf7 <- Seurat::FindVariableFeatures(mcf7, selection.method = "vst",
                                     nfeatures = n)
# Scale data
mcf7 <- Seurat::ScaleData(mcf7)
# Principal component analysis
mcf7 <- Seurat::RunPCA(mcf7, features = Seurat::VariableFeatures(mcf7))
```



<br>

## Cluster cells
Compute the cumulative sum of variances, which is used for determining
the number of the principal components (PCs).
```{r, eval = FALSE}
pc <- which(cumsum(mcf7@reductions[["pca"]]@stdev) /
              sum(mcf7@reductions[["pca"]]@stdev) > 0.9)[1]
```

Perform cell clustering.
```{r, eval = FALSE}
# Create k-nearest neighbor graph.
mcf7 <- Seurat::FindNeighbors(mcf7, reduction = "pca", dim = seq_len(pc))
# Cluster cells.
mcf7 <- Seurat::FindClusters(mcf7, resolution = 0.6)
# Run UMAP.
mcf7 <- Seurat::RunUMAP(mcf7, dims = seq_len(pc))
```

Visualize the clustering results.
```{r, eval = FALSE}
title <- "MCF-7"
df <- as.data.frame(Seurat::Embeddings(object = mcf7, reduction = "umap"))

# Conditions
df$label <- factor(mcf7$type, levels = unique(sort(mcf7$type)))
p <- ggplot2::ggplot() +
  ggplot2::geom_point(ggplot2::aes(x = df[, 1], y = df[, 2], color = df$label),
                      size = 1, alpha = 1) +
  ggplot2::labs(title = title, x = "UMAP_1", y = "UMAP_2", color = "") +
  ggplot2::theme_classic(base_size = 20) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  ggplot2::guides(colour = guide_legend(override.aes = list(size = 4))) +
  ggplot2::scale_color_manual(breaks = unique(sort(df$label)),
                              values = c("Control" = "grey20",
                                         "TAM_3weeks" = "grey40",
                                         "TAM_6weeks" = "grey60",
                                         "TAM_9weeks" = "grey80"))
filename <- "figures/figure_01_0020.png"
ggplot2::ggsave(file = filename, plot = p, dpi = 80, width = 7, height = 5)

# Seurat clusters
df$label <- mcf7$seurat_clusters
p <- ggplot2::ggplot() +
  ggplot2::geom_point(ggplot2::aes(x = df[, 1], y = df[, 2], color = df$label),
                      size = 1, alpha = 1) +
  ggplot2::labs(title = title, x = "UMAP_1", y = "UMAP_2", color = "") +
  ggplot2::theme_classic(base_size = 20) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  ggplot2::guides(colour = guide_legend(override.aes = list(size = 4)))
filename <- "figures/figure_01_0021.png"
ggplot2::ggsave(file = filename, plot = p, dpi = 80, width = 5.8, height = 5)

# QC metrics
df$ncounts <- mcf7$nCount_RNA
p <- ggplot2::ggplot() +
  ggplot2::geom_point(ggplot2::aes(x = df[, 1], y = df[, 2], color=df$ncounts),
                      size = 1, alpha = 1) +
  ggplot2::labs(title = title, x = "UMAP_1", y = "UMAP_2", color="nCount_RNA") +
  ggplot2::scale_color_continuous() + ggplot2::theme_classic(base_size = 20) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, size = 20))
filename <- "figures/figure_01_0022.png"
ggplot2::ggsave(file = filename, plot = p, dpi = 80, width = 7, height = 5)

df$ncounts <- mcf7$percMT
p <- ggplot2::ggplot() +
  ggplot2::geom_point(ggplot2::aes(x = df[, 1], y = df[, 2], color=df$ncounts),
                      size = 1, alpha = 1) +
  ggplot2::labs(title = title, x = "UMAP_1", y = "UMAP_2", color = "percMT") +
  ggplot2::scale_color_continuous() + ggplot2::theme_classic(base_size = 20) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, size = 20))
filename <- "figures/figure_01_0023.png"
ggplot2::ggsave(file = filename, plot = p, dpi = 80, width = 6.2, height = 5)
```

<img src="figures/figure_01_0020.png" width="200px">
<img src="figures/figure_01_0021.png" width="200px">
<img src="figures/figure_01_0022.png" width="200px">
<img src="figures/figure_01_0023.png" width="200px">



<br>

## Cell cycle inference
Assign each cell a cell cycle score using `CellCycleScoring()`.
```{r, eval = FALSE}
s.genes <- Seurat::cc.genes$s.genes
g2m.genes <- Seurat::cc.genes$g2m.genes
mcf7 <- Seurat::CellCycleScoring(mcf7, s.features = Seurat::cc.genes$s.genes,
                                 g2m.features = Seurat::cc.genes$g2m.genes)
```

Visualize the cell cycle phases.
```{r, eval = FALSE}
title <- "MCF-7"
df <- as.data.frame(Seurat::Embeddings(object = mcf7, reduction = "umap"))
df$label <- factor(mcf7$Phase, levels = c("G1", "S", "G2M"))
p <- ggplot2::ggplot() +
  ggplot2::geom_point(ggplot2::aes(x = df[, 1], y = df[, 2], color = df$label),
                      size = 1, alpha = 1) +
  ggplot2::labs(title = title, x = "UMAP_1", y = "UMAP_2", color = "Phase") +
  ggplot2::theme_classic(base_size = 20) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  ggplot2::guides(colour = guide_legend(override.aes = list(size = 4))) +
  ggplot2::scale_color_manual(breaks = unique(sort(df$label)),
                              values = c("G1" = rainbow(3)[3],
                                         "S" = rainbow(3)[2],
                                         "G2M" = rainbow(3)[1]))
filename <- "figures/figure_01_0024.png"
ggplot2::ggsave(file = filename, plot = p, dpi = 80, width = 6.2, height = 5)
```

<img src="figures/figure_01_0024.png" width="200px">

Save the data.
```{r, eval = FALSE}
saveRDS(mcf7, file = "backup/01_004_mcf7_seurat.rds")

#Load the data.
#mcf7 <- readRDS("backup/01_004_mcf7_seurat.rds")
```



<br>

# Pseudotime analysis using MERLoT
Attach R libraries.
```{r, eval = FALSE}
library(plot3D)
```



<br>

## Diffusion map
Perform a dimensionality reduction using a diffusion map.
```{r, eval = FALSE}
mat <- Seurat::Embeddings(object = mcf7, reduction = "pca")
set.seed(1)
res <- destiny::DiffusionMap(mat, sigma = "local", distance = "euclidean")
```

Check the results.
```{r, eval = FALSE, echo = FALSE}
df <- as.data.frame(res@eigenvectors)
df$label <- factor(mcf7$type, levels = unique(sort(mcf7$type)))
plotly::plot_ly(df, x = ~df[, 1], y = ~df[, 2], z = ~df[, 3],
                marker = list(size = 5)) %>% add_markers(color = ~df$label)
```
```{r, eval = FALSE}
df <- as.data.frame(res@eigenvectors)[, seq_len(3)]
df$label <- factor(mcf7$type, levels = unique(sort(mcf7$type)))
colors <- as.character(df$label)
colors[which(colors == "Control")] <- "grey20"
colors[which(colors == "TAM_3weeks")] <- "grey40"
colors[which(colors == "TAM_6weeks")] <- "grey60"
colors[which(colors == "TAM_9weeks")] <- "grey80"
filename <- "figures/figure_01_0030.png"
png(file = filename, height = 500, width = 500, res = 80)
scatter3D(df[, 1], df[, 2], df[, 3],
          main = "MCF-7", xlab = "DC_1", ylab = "DC_2", zlab = "DC_3",
          box = TRUE, bty = "b2", axes = TRUE, nticks = 5,
          theta = 230, phi = 30, pch = 16, cex = 1.0, alpha = 1.0,
          col = colors, colvar = NA, colkey = FALSE)
graphics::legend("bottomright", legend = unique(sort(df$label)),
                 pch = 16, col = unique(sort(colors)), cex = 1,
                 inset = c(-0.0))
dev.off()
```

<img src="figures/figure_01_0030.png" width="200px">



mat <- Seurat::Embeddings(object = mcf7, reduction = "pca")
set.seed(1)
res <- DiffusionMap(mat, sigma = "local", distance = "euclidean")
df <- as.data.frame(res@eigenvectors)
df$label <- mcf7$type
#plot_ly(df, x = ~df[, 1], y = ~df[, 2], z = ~df[, 3],
#        marker = list(size = 5)) %>% add_markers(color=~df$label)
labels <- factor(df$label, levels = c("Control", "TAM_3weeks", "TAM_6weeks", "TAM_9weeks"))
colors <- df$label
n <- length(unique(labels))
colors[which(colors == "Control")] <- "grey20" #"green4"
colors[which(colors == "TAM_3weeks")] <- "grey40" #"magenta"
colors[which(colors == "TAM_6weeks")] <- "grey60" #"cyan"
colors[which(colors == "TAM_9weeks")] <- "grey80" #"gold"
df <- as.data.frame(res@eigenvectors)[, seq_len(3)]
filename <- "figures/figure_01_0025.png"
png(file = filename, height = 1500, width = 1500, res = 300)
scatter3D(df[, 1], df[, 2], df[, 3],
          main = "MCF-7", xlab = "", ylab = "", zlab = "",
          box = TRUE, bty = "b2", axes = TRUE, nticks = 5,
          theta = 230, phi = 30, pch = 16, cex = 1.0, alpha = 1.0,
          col = colors, colvar = NA, colkey = FALSE)
graphics::legend("bottomright", legend = unique(sort(labels)),
                 pch = 16, col = unique(sort(colors)), cex = 0.7,
                 inset = c(-0.0))
dev.off()



<br>

## MERLoT
Compute a pseudotime using MERLoT.
```{r, eval = FALSE}
df <- as.data.frame(res@eigenvectors)

merlot <- list()
merlot[["ScaffoldTree"]] <- merlot::CalculateScaffoldTree(
  CellCoordinates = df[, seq_len(4)], NEndpoints = 4, random_seed = 1
)
merlot[["ElasticTree"]] <- merlot::CalculateElasticTree(
  ScaffoldTree = merlot[["ScaffoldTree"]], N_yk = 60 # Not too few not too many.
)
merlot[["GenesSpaceEmbedding"]] <- merlot::GenesSpaceEmbedding(
  ExpressionMatrix = t(as.matrix(mcf7@assays[["RNA"]]@data)),
  ElasticTree = merlot[["ElasticTree"]], NCores = 3
)
merlot[["Pseudotimes_lowdim"]] <- merlot::CalculatePseudotimes(
  InputTree = merlot[["ElasticTree"]], T0 = 2
)
merlot[["Pseudotimes_highdim"]] <- CalculatePseudotimes(
  InputTree = merlot[["GenesSpaceEmbedding"]], T0 = 2
)
```

Save the data.
```{r, eval = FALSE}
saveRDS(merlot, file = "backup/01_005_mcf7_merlot_nky60.rds")

#Load the data.
#merlot <- readRDS("backup/01_005_mcf7_merlot_nky60.rds")
merlot <- readRDS("robustness_abcdConstraint_C1/merlot_data/01_004_mcf7_merlot_C3_Nyk60.rds")
```

Visualize the MERLoT clustering results.
```{r, eval = FALSE}
# MERLoT cluster
# Note the followings:
# identical(merlot[["GenesSpaceEmbedding"]][["Cells2Branches"]],
#           merlot[["Pseudotimes_highdim"]][["Cells2Branches"]]) = TRUE
# identical(merlot[["Pseudotimes_lowdim"]][["Cells2Branches"]],
#           merlot[["ElasticTree"]][["Cells2Branches"]]) = TRUE
# identical(merlot[["GenesSpaceEmbedding"]][["Cells2Branches"]],
#           merlot[["ElasticTree"]][["Cells2Branches"]]) = FALSE
df <- as.data.frame(res@eigenvectors)[, seq_len(3)]
df$branch <- merlot[["GenesSpaceEmbedding"]][["Cells2Branches"]]
dg <- as.data.frame(merlot[["ElasticTree"]][["Nodes"]])
dh <- as.data.frame(merlot[["ElasticTree"]][["EndpointsCoords"]])
dh <- rbind(dh, merlot[["ElasticTree"]][["BranchpointsCoords"]])
dj <- list()
for(j in unique(sort(merlot[["GenesSpaceEmbedding"]][["Cells2Branches"]]))){
  dj[[j]] <- as.data.frame(dg[merlot[["ElasticTree"]][["Branches"]][[j]],])
}
colors <- scales::hue_pal()(length(unique(df$branch)))[df$branch]
df$color <- scales::hue_pal()(length(unique(df$branch)))[df$branch]
df <- df[order(df$branch), ]
filename <- "figures/figure_01_0026.png"
png(file = filename, height = 500, width = 500, res = 80)
scatter3D(df[, 1], df[, 2], df[, 3],
          main = "MCF-7", xlab = "DC_1", ylab = "DC_2", zlab = "DC_3",
          box = FALSE, bty = "b2", axes = FALSE, nticks = 5,
          theta = 230, phi = 30, pch = 16, cex = 1.0, alpha = 1.0,
          col = df$color, colvar = NA, colkey = FALSE)
graphics::legend("bottomright", legend=unique(df$branch),
                 pch = 16, col = unique(df$color), cex = 1, inset = c(0.02))
dev.off()
```



# Elastic tree
source("R/plot_additional.R")
labels <- merlot[["GenesSpaceEmbedding"]][["Cells2Branches"]]
colors <- scales::hue_pal()(length(unique(labels)))[labels]
df <- merlot$ElasticTree
dg <- merlot[["GenesSpaceEmbedding"]]
#df[["CellCoords"]][, 1] <- -df[["CellCoords"]][, 1]
#df[["CellCoords"]][, 3] <- -df[["CellCoords"]][, 3]
#df[["EndpointsCoords"]][, 1] <- -df[["EndpointsCoords"]][, 1]
#df[["EndpointsCoords"]][, 3] <- -df[["EndpointsCoords"]][, 3]
#df[["BranchpointsCoords"]][, 1] <- -df[["BranchpointsCoords"]][, 1]
#df[["BranchpointsCoords"]][, 3] <- -df[["BranchpointsCoords"]][, 3]
#df[["Nodes"]][, 1] <- -df[["Nodes"]][, 1]
#df[["Nodes"]][, 3] <- -df[["Nodes"]][, 3]
filename <- "figures/figure_01_0027.png"
filename <- "figures/figure_01_0027b.png"
png(file = filename, height = 1500, width = 1500, res = 300)
plot_elasticTree3D(ElasticTree = df, GeneSpaceEmbedding = dg,
                   labels = labels, colors = colors, theta = 230, phi = 30,
                   title = "MCF-7", xlabel = "", ylabel = "",
                   zlabel = "")
dev.off()

# Pseudotime
filename <- "figures/figure_01_0028b.png"
png(file = filename, height = 1500, width = 1500, res = 300)
plot_pseudotime3D(ElasticTree = df, Pseudotimes = merlot$Pseudotimes_highdim,
                  labels = labels, colors = colors, theta = 230, phi = 30,
                  title = "MCF-7", xlabel = "", ylabel = "",
                  zlabel = "")
dev.off()

# Cell cycle
labels <- factor(mcf7$Phase, levels = unique(mcf7$Phase))
colors <- mcf7$Phase
colors[which(colors == "G1")] <- rainbow(3)[3]
colors[which(colors == "S")] <- rainbow(3)[2]
colors[which(colors == "G2M")] <- rainbow(3)[1]
df <- as.data.frame(res@eigenvectors)[, seq_len(3)]
filename <- "figures/figure_01_0029.png"
png(file = filename, height = 1500, width = 1500, res = 300)
scatter3D(df[, 1], df[, 2], df[, 3],
          main = "MCF-7", xlab = "DC_1", ylab = "DC_2", zlab = "DC_3",
          box = FALSE, bty = "b2", axes = FALSE, nticks = 5,
          theta = 230, phi = 30, pch = 16, cex = 1.0, alpha = 1.0,
          col = colors, colvar = NA, colkey = FALSE)
graphics::legend("bottomright", legend=unique(labels),
                 pch = 16, col = unique(colors), cex = 1,
                 inset = c(0.0))
dev.off()



# Set MERLoT information into Seurat object 
## Change the MERLoT cluster names
#df <- merlot[["ElasticTree"]][["Cells2Branches"]]
df <- merlot[["Pseudotimes_highdim"]][["Cells2Branches"]]
df[which(df == 1)] <- 44
df[which(df == 2)] <- 55
df[which(df == 3)] <- 11
df[which(df == 4)] <- 33
df[which(df == 5)] <- 22
df[which(df == 11)] <- 1
df[which(df == 22)] <- 2
df[which(df == 33)] <- 3
df[which(df == 44)] <- 4
df[which(df == 55)] <- 5
mcf7$merlot_clusters <- factor(df, levels = seq_along(unique(df)))

## Set pseudotime
df <- merlot[["Pseudotimes_highdim"]][["Times_cells"]]
mcf7$merlot_pseudotime <- df

## Differentially expressed genes
mcf7 <- SetIdent(mcf7, value = "merlot_clusters")
mcf7@misc[["FindAllMarkers"]] <- FindAllMarkers(mcf7, only.pos = TRUE,
                                                min.pct = 0.25,
                                                logfc.threshold = 0.25)
View(mcf7@misc[["FindAllMarkers"]])

Save the data.
```{r, eval = FALSE}
saveRDS(mcf7, file = "backup/01_005_mcf7_seurat.rds")

#Load the data.
#mcf7 <- readRDS("backup/01_005_mcf7_seurat.rds")
```

Show heatmaps for variably expressed genes.
```{r, eval = FALSE}
topn <- mcf7@misc[["FindAllMarkers"]]
topn <- topn[which(topn$gene %in% mcf7@assays[["RNA"]]@var.features), ]
#topn <- topn[!grepl("LOC102724852", topn$gene), ]
topn <- dplyr::group_by(topn, cluster)
tmp <- topn
topn$p_val_adj <- -log10(topn$p_val_adj)
topn_new <- topn
topn_new <- dplyr::top_n(topn, n = 8, p_val_adj)      # Selecting by p_val_adj
topn_new <- dplyr::top_n(topn_new, n = 8, avg_log2FC) # Selecting by avg_log2FC
topn_tmp <- tmp[which(tmp$gene %in% topn_new$gene), ]

mycolor <- scales::hue_pal()(length(unique(mcf7$merlot_clusters)))
#mycolor <- scales::brewer_pal(palette = "Set2")(3)
mycolor <- c("1" = mycolor[3], "2" = mycolor[5], "3" = mycolor[4],
             "4" = mycolor[1], "5" = mycolor[2])
set.seed(1)
p <- Seurat::DoHeatmap(mcf7, features = topn_tmp$gene,
                       group.by = "merlot_clusters",
                       group.colors = mycolor, angle = 0, size = 4) +
  ggplot2::guides(color = "none") +
  ggplot2::labs(fill = "Scaled\nnormalized\nexpression")
filename <- "figures/figure_01_0030.png"
ggplot2::ggsave(file = filename, plot = p, dpi = 200, width = 6, height = 6)
```



# ----------------------------------------
# Show gene expressions as a function of pseudotime
# ----------------------------------------
source("R/myfuncs.R")
genes <- rownames(mcf7)

# Bar chart
library(dplyr)
Plot_bar_cellcycle <- function(obj, ymax, title, xlabel, ylabel){
  df <- obj[[]][, c("merlot_clusters", "type")]
  df$merlot_clusters <- as.integer(df$merlot_clusters)
  df <- group_by(df, type, merlot_clusters) %>% summarise (n=n())
  df <- rbind(df, data.frame(type = "Control", merlot_clusters = 3, n = 0))
  df <- rbind(df, data.frame(type = "Control", merlot_clusters = 4, n = 0))
  df <- rbind(df, data.frame(type = "TAM_3weeks", merlot_clusters = 1, n = 0))
  df <- rbind(df, data.frame(type = "TAM_6weeks", merlot_clusters = 1, n = 0))
  df <- rbind(df, data.frame(type = "TAM_9weeks", merlot_clusters = 1, n = 0))
  df$type <- factor(df$type, levels = c("Control", "TAM_3weeks", "TAM_6weeks", "TAM_9weeks"))
  df$merlot_clusters <- as.integer(as.character(df$merlot_clusters))
  pop <- c()
  for(j in 1:length(unique(df$merlot_clusters))){
    c <- df[which(df$merlot_clusters == j),] ; c$n <- c$n / sum(c$n)
    pop <- rbind(pop, c)
  }
  n_groups <- length(unique(df$type))
#  my_colors <- c("green4", "magenta", "cyan", "gold")
  my_colors <- c("grey20", "grey40", "grey60", "grey80")
  p <- ggplot(pop) +
    geom_bar(aes(x = as.factor(merlot_clusters), y = n, fill = type),
           stat="identity", position=position_dodge2(width=0.8), width=0.8) +
    theme_classic(base_size=20, base_family="Helvetica") +
    theme(plot.title=element_text(hjust=0.5)) +
    ylim(0, ymax) + labs(title = title, x = xlabel, y = ylabel) +
    scale_fill_manual(values=my_colors)
  return(p)
}
p <- Plot_bar_cellcycle(obj = mcf7, ymax = 1, title = "MCF-7",
                        xlabel = "MERLoT Cluster", ylabel = "Population ratio")
filename <- "figures/figure_01_0040.png"
ggsave(file = filename, plot = p, dpi = 300, width = 7.5, height = 4)



Plot_bar_cellcycle <- function(obj, ymax, title, xlabel, ylabel){
  df <- obj[[]][, c("merlot_clusters", "Phase")]
  df$merlot_clusters <- as.integer(df$merlot_clusters)
  df <- group_by(df, Phase, merlot_clusters) %>% summarise (n=n())
  df$Phase <- factor(df$Phase, levels = c("G1", "S", "G2M"))
  df$merlot_clusters <- as.integer(as.character(df$merlot_clusters))
  pop <- c()
  for(j in 1:length(unique(df$merlot_clusters))){
    c <- df[which(df$merlot_clusters == j),] ; c$n <- c$n / sum(c$n)
    pop <- rbind(pop, c)
  }
  n_groups <- length(unique(df$Phase))
  my_colors <- c(rainbow(n_groups)[3], rainbow(n_groups)[2], rainbow(n_groups)[1])
  p <- ggplot(pop) +
    geom_bar(aes(x=as.factor(merlot_clusters), y=n, fill=Phase),
           stat="identity", position=position_dodge2(width=0.7), width=0.7) +
    theme_classic(base_size=20, base_family="Helvetica") +
    theme(plot.title=element_text(hjust=0.5)) +
    ylim(0, ymax) + labs(title = title, x = xlabel, y = ylabel) +
    scale_fill_manual(values = my_colors)
  return(p)
}
p <- Plot_bar_cellcycle(obj = mcf7, ymax = 1, title = "MCF-7",
                        xlabel = "MERLoT Cluster", ylabel = "Population ratio")
filename <- "figures/figure_01_0041.png"
ggsave(file = filename, plot = p, dpi = 300, width = 6.5, height = 4)


Plot_bar_ncells_batch2 <- function(obj, ymax, title, xlabel, ylabel)
{
  df <- obj[[]][, c("merlot_clusters", "Phase", "type")]
  df$merlot_clusters <- as.integer(df$merlot_clusters)
  df <- group_by(df, merlot_clusters, Phase, type) %>% summarise (n=n())
  df$Phase <- factor(df$Phase, levels = c("G1", "S", "G2M"))
  df$type <- factor(df$type, levels = c("Control", "TAM_3weeks", "TAM_6weeks", "TAM_9weeks"))
  df$merlot_clusters <- as.integer(as.character(df$merlot_clusters))
  # Complete
  df <- rbind(df, data.frame(merlot_clusters = 1, Phase = "G1", type = "TAM_3weeks", n = 0))
  df <- rbind(df, data.frame(merlot_clusters = 1, Phase = "G1", type = "TAM_6weeks", n = 0))
  df <- rbind(df, data.frame(merlot_clusters = 1, Phase = "G1", type = "TAM_9weeks", n = 0))
  df <- rbind(df, data.frame(merlot_clusters = 1, Phase = "S", type = "TAM_3weeks", n = 0))
  df <- rbind(df, data.frame(merlot_clusters = 1, Phase = "S", type = "TAM_6weeks", n = 0))
  df <- rbind(df, data.frame(merlot_clusters = 1, Phase = "S", type = "TAM_9weeks", n = 0))
  df <- rbind(df, data.frame(merlot_clusters = 1, Phase = "G2M", type = "TAM_3weeks", n = 0))
  df <- rbind(df, data.frame(merlot_clusters = 1, Phase = "G2M", type = "TAM_6weeks", n = 0))
  df <- rbind(df, data.frame(merlot_clusters = 1, Phase = "G2M", type = "TAM_9weeks", n = 0))
  df <- rbind(df, data.frame(merlot_clusters = 2, Phase = "G1", type = "TAM_3weeks", n = 0))
  df <- rbind(df, data.frame(merlot_clusters = 2, Phase = "G1", type = "TAM_6weeks", n = 0))
  df <- rbind(df, data.frame(merlot_clusters = 2, Phase = "G1", type = "TAM_9weeks", n = 0))
  df <- rbind(df, data.frame(merlot_clusters = 3, Phase = "G1", type = "Control", n = 0))
  df <- rbind(df, data.frame(merlot_clusters = 3, Phase = "G2M", type = "Control", n = 0))
  df <- rbind(df, data.frame(merlot_clusters = 3, Phase = "G2M", type = "TAM_3weeks", n = 0))
  df <- rbind(df, data.frame(merlot_clusters = 3, Phase = "G2M", type = "TAM_9weeks", n = 0))
  df <- rbind(df, data.frame(merlot_clusters = 3, Phase = "S", type = "Control", n = 0))
  df <- rbind(df, data.frame(merlot_clusters = 4, Phase = "G1", type = "Control", n = 0))
  df <- rbind(df, data.frame(merlot_clusters = 4, Phase = "S", type = "Control", n = 0))
  df <- rbind(df, data.frame(merlot_clusters = 4, Phase = "G2M", type = "Control", n = 0))
  # Change the name of type
  df[which(df$type == "Control"), ]$type <- "0w"
  df[which(df$type == "TAM_3weeks"), ]$type <- "3w"
  df[which(df$type == "TAM_6weeks"), ]$type <- "6w"
  df[which(df$type == "TAM_9weeks"), ]$type <- "9w"
  df$type <- factor(df$type, levels = c("0w", "3w", "6w", "9w"))
  df$Phase <- factor(df$Phase, levels = c("G1", "S", "G2M"))
  df <- arrange(df, merlot_clusters, Phase, type)
  n_groups <- length(unique(df$Phase))
  my_colors <- c(rainbow(n_groups)[3], rainbow(n_groups)[2], rainbow(n_groups)[1])
  df$r <- NA
  temp <- data.frame()
  for(i in seq_len(length(unique(df$merlot_clusters)))){
    dg <- df[which(df$merlot_clusters == i), ]
    dg$r <- dg$n / sum(dg$n)
    temp <- rbind(temp, dg)
  }
  df$r <- temp$r
  library(ggh4x)
  mycol <- scales::hue_pal()(length(unique(df$merlot_clusters)))[c(3, 5, 4, 1, 2)]
  strip <- strip_themed(background_x = elem_list_rect(fill = mycol))
  p <- ggplot(df) +
    geom_bar(aes(x = type, y = r, fill = Phase),
           stat = "identity", position = position_dodge2(width = 0.75), width = 0.75) +
    facet_grid2(~ as.factor(merlot_clusters), strip = strip) +
#    facet_wrap2(~ as.factor(merlot_clusters), strip = strip) +
    ylim(0, ymax) + labs(title = title, x = xlabel, y = ylabel, fill = "Phase") +
    scale_fill_manual(values = my_colors) +
    theme_classic(base_size = 20, base_family = "Helvetica") +
    theme(plot.title = element_text(hjust = 0.5))
  return(p)
}
p <- Plot_bar_ncells_batch2(obj = mcf7, ymax = 0.5, title = "MCF-7",
                            xlabel = "MERLoT Cluster", ylabel = "Population ratio")
filename <- "figures/figure_01_0042.png"
ggsave(file = filename, plot = p, dpi = 300, width = 12, height = 5)



<br>

# Preparing a figure
mcf7 <- readRDS("backup/01_005_mcf7_seurat.rds")
merlot <- readRDS("robustness_abcdConstraint_C1/merlot_data/01_004_mcf7_merlot_C3_Nyk60.rds")

matl_1 <- list()
matl_2 <- list()
types <- c("Control", "TAM_3weeks", "TAM_6weeks", "TAM_9weeks")
for(i in seq_along(types)){
  cells <- rownames(mcf7[[]][which(mcf7$type == types[i]), ])
  tmp <- as.matrix(mcf7@assays[["RNA"]]@data)
  matl_1[[i]] <- tmp[, cells]
  tmp <- t(merlot[["Pseudotimes_highdim"]][["PseudoExpressionMatrix"]])
  matl_2[[i]] <- tmp[, cells]
}
#--------------------------------------------------
# Random sampling
#--------------------------------------------------
for(i in seq_along(matl)){
  set.seed(1)
  inds <- sample(ncol(matl_1[[i]]), size = 100, replace = FALSE)
  matl_1[[i]] <- matl_1[[i]][, inds]
  matl_2[[i]] <- matl_2[[i]][, inds]
}
tmp <- t(mcf7@assays[["RNA"]]@meta.features[["vst.mean"]])
colnames(tmp) <- seq_len(ncol(tmp))
inds <- as.integer(names(tmp[, order(tmp, decreasing = TRUE)][seq_len(50)]))
for(i in seq_along(matl_1)){
  matl_1[[i]] <- matl_1[[i]][inds, ]
  matl_1[[i]] <- t(scale(t(matl_1[[i]])))
  
  matl_2[[i]] <- matl_2[[i]][inds, ]
  matl_2[[i]] <- t(scale(t(matl_2[[i]])))
}
#--------------------------------------------------
# Compute heatmaps of sign-by-sample matrices.
#--------------------------------------------------
i = 1
  filename <- sprintf("figures/figure_01_%04d.png", i)
  png(file = filename, height = 300, width = 1000, res = 300)
  set.seed(1)
  p <- c()
  res <- hclust(dist(matl_2[[i]]))
  q <- Heatmap(matl_1[[i]], show_row_names = FALSE, show_row_dend = FALSE,
               show_column_names = TRUE, row_order = res$order,
               column_gap = unit(1.5, "mm"),
               column_title = "", name = as.character(i))
  p <- p %v% q
  q <- Heatmap(matl_2[[i]], show_row_names = FALSE, show_row_dend = FALSE,
               show_column_names = FALSE, row_order = res$order)
  p <- p %v% q
  p
  dev.off()

i = 2
  filename <- sprintf("figures/figure_01_%04d.png", i)
  png(file = filename, height = 300, width = 1000, res = 300)
  set.seed(1)
  p <- c()
  res <- hclust(dist(matl_2[[i]]))
  q <- Heatmap(matl_1[[i]], show_row_names = FALSE, show_row_dend = FALSE,
               show_column_names = TRUE, row_order = res$order,
               column_gap = unit(1.5, "mm"),
               column_title = "", name = as.character(i))
  p <- p %v% q
  q <- Heatmap(matl_2[[i]], show_row_names = FALSE, show_row_dend = FALSE,
               show_column_names = FALSE, row_order = res$order)
  p <- p %v% q
  p
  dev.off()

i = 3
  filename <- sprintf("figures/figure_01_%04d.png", i)
  png(file = filename, height = 300, width = 1000, res = 300)
  set.seed(1)
  p <- c()
  res <- hclust(dist(matl_2[[i]]))
  q <- Heatmap(matl_1[[i]], show_row_names = FALSE, show_row_dend = FALSE,
               show_column_names = TRUE, row_order = res$order,
               column_gap = unit(1.5, "mm"),
               column_title = "", name = as.character(i))
  p <- p %v% q
  q <- Heatmap(matl_2[[i]], show_row_names = FALSE, show_row_dend = FALSE,
               show_column_names = FALSE, row_order = res$order)
  p <- p %v% q
  p
  dev.off()

i = 4
  filename <- sprintf("figures/figure_01_%04d.png", i)
  png(file = filename, height = 300, width = 1000, res = 300)
  set.seed(1)
  p <- c()
  res <- hclust(dist(matl_2[[i]]))
  q <- Heatmap(matl_1[[i]], show_row_names = FALSE, show_row_dend = FALSE,
               show_column_names = TRUE, row_order = res$order,
               column_gap = unit(1.5, "mm"),
               column_title = "", name = as.character(i))
  p <- p %v% q
  q <- Heatmap(matl_2[[i]], show_row_names = FALSE, show_row_dend = FALSE,
               show_column_names = FALSE, row_order = res$order)
  p <- p %v% q
  p
  dev.off()



<br>

## Filter unused genes
Make a gene-by-pseudotime matrix.
```{r, eval = FALSE}
#Load the data.
mcf7 <- readRDS("backup/01_005_mcf7_seurat.rds")
merlot <- readRDS("robustness_abcdConstraint_C1/merlot_data/01_004_mcf7_merlot_C3_Nyk60.rds")

genes <- rownames(mcf7)
EmbeddedTree = merlot[["GenesSpaceEmbedding"]]
Pseudotimes = merlot[["Pseudotimes_highdim"]]
inds_node <- c()
for(i in seq_along(Pseudotimes[["Branches"]])){
  branch_i <- Pseudotimes[["Branches"]][[i]]
  aux <- sort(Pseudotimes$Times_yk[branch_i], index.return = T)
  inds_node <- c(inds_node, branch_i[aux$ix])
}
pt_max <- max(merlot[["Pseudotimes_highdim"]][["Times_yk"]])
pt_min <- min(merlot[["Pseudotimes_highdim"]][["Times_yk"]])
ptr <- (Pseudotimes[["Times_yk"]][inds_node] - pt_min) / (pt_max - pt_min)
ptmat <- data.frame(merlot_pseudotime = ptr)
for(i in seq_along(genes)){
#for(i in seq_len(100)){
  GeneName <- genes[i]
  Gene = which(colnames(EmbeddedTree$CellCoords) == GeneName)
  ExpressionGene = EmbeddedTree$Nodes[, Gene]
  ExpressionGeneOriginal = EmbeddedTree$CellCoords[, Gene]
  ptmat <- cbind(ptmat, EmbeddedTree$Nodes[inds_node, Gene])
  colnames(ptmat)[i + 1] <- GeneName
}
ptmat <- ptmat[order(ptmat$merlot_pseudotime), ]
rownames(ptmat) <- seq_len(nrow(ptmat))
ptmat <- t(ptmat)
```

Filter the low-expressing genes using quantile.
```{r, eval = FALSE}
mat_max <- apply(ptmat, 1, max)
ptmat_red <- ptmat[which(mat_max >= 0.1), ]
```

Check the negative values.
```{r, eval = FALSE}
min_vector <- apply(ptmat_red, 1, min)
min(min_vector)
```
```
[1] -0.02426622
```

Add `abs(min_vector) + EPS(0.0001)` to the rows having negative values.
```{r, eval = FALSE}
eps <- 0.0001
pseudotime <- ptmat_red[1, ]
for(i in seq_len(nrow(ptmat_red))){
  if(min_vector[i] < 0){
    ptmat_red[i, ] <- ptmat_red[i, ] - min_vector[i] + eps
  }
}
ptmat_red[1, ] <- pseudotime
```

```{r, eval = FALSE}
mcf7 <- mcf7[rownames(ptmat_red[-1, ]), ]
```

Save the data.
```{r, eval = FALSE}
saveRDS(mcf7, file = "backup/01_006_mcf7_reduced.rds")

#Load the data.
#mcf7 <- readRDS("backup/01_006_mcf7_reduced.rds")
```



#===============================================================================
# Pseudotime analyses
#===============================================================================

<br>

## From cluster 1 to 1 without clusters 2, 3, 4 and 5
mcf7 <- readRDS("backup/01_006_mcf7_reduced.rds")
merlot <- readRDS("backup/01_005_mcf7_merlot_nky60.rds")
mcf7_sub <- mcf7[, which(mcf7$merlot_clusters %in% c(1))]

merlot_sub <- merlot
merlot_sub[["Pseudotimes_highdim"]][["Cells2Branches"]] <- mcf7$merlot_clusters
mat <- merlot_sub[["GenesSpaceEmbedding"]][["CellCoords"]]
inds_cell <- which(rownames(mat) %in% colnames(mcf7_sub))
inds_gene <- which(colnames(mat) %in% rownames(mcf7_sub))

### Modify merlot object.
### GenesSpaceEmbedding
merlot_sub[["GenesSpaceEmbedding"]][["Nodes"]] <-
  merlot[["GenesSpaceEmbedding"]][["Nodes"]][, inds_gene]
merlot_sub[["GenesSpaceEmbedding"]][["Topology"]][["Endpoints"]] <- c(2)
merlot_sub[["GenesSpaceEmbedding"]][["Topology"]][["Branchpoints"]] <- c(5)
merlot_sub[["GenesSpaceEmbedding"]][["CellCoords"]] <-
  merlot[["GenesSpaceEmbedding"]][["CellCoords"]][inds_cell, inds_gene]
merlot_sub[["GenesSpaceEmbedding"]][["Cells2TreeNodes"]] <-
  merlot[["GenesSpaceEmbedding"]][["Cells2TreeNodes"]][inds_cell, ]
merlot_sub[["GenesSpaceEmbedding"]][["Cells2Branches"]] <-
  merlot[["GenesSpaceEmbedding"]][["Cells2Branches"]][inds_cell]
merlot_sub[["GenesSpaceEmbedding"]][["Branches"]][[5]] <- NULL
merlot_sub[["GenesSpaceEmbedding"]][["Branches"]][[4]] <- NULL
merlot_sub[["GenesSpaceEmbedding"]][["Branches"]][[2]] <- NULL
merlot_sub[["GenesSpaceEmbedding"]][["Branches"]][[1]] <- NULL

### Pseudotimes_highdim
merlot_sub[["Pseudotimes_highdim"]][["Times_cells"]] <-
  merlot[["Pseudotimes_highdim"]][["Times_cells"]][inds_cell]
merlot_sub[["Pseudotimes_highdim"]][["Projected_Times_Cells"]] <-
  merlot[["Pseudotimes_highdim"]][["Projected_Times_Cells"]][inds_cell]
merlot_sub[["Pseudotimes_highdim"]][["Branches"]][[5]] <- NULL
merlot_sub[["Pseudotimes_highdim"]][["Branches"]][[4]] <- NULL
merlot_sub[["Pseudotimes_highdim"]][["Branches"]][[2]] <- NULL
merlot_sub[["Pseudotimes_highdim"]][["Branches"]][[1]] <- NULL
merlot_sub[["Pseudotimes_highdim"]][["Cells2TreeNodes"]] <-
  merlot[["Pseudotimes_highdim"]][["Cells2TreeNodes"]][inds_cell, ]
merlot_sub[["Pseudotimes_highdim"]][["Cells2Branches"]] <-
  merlot[["Pseudotimes_highdim"]][["Cells2Branches"]][inds_cell]
merlot_sub[["Pseudotimes_highdim"]][["PseudoExpressionMatrix"]] <-
  merlot[["Pseudotimes_highdim"]][["PseudoExpressionMatrix"]][inds_cell, inds_gene]
# ----------------------------------------
# Show gene expressions as a function of pseudotime
# ----------------------------------------
source("R/myfuncs.R")
genes <- rownames(mcf7_sub)
dirname <- "fig_pt_fromC1toC1_woC2C3C4C5"
dir.create(dirname)
#for(i in seq_along(genes)){
for(i in seq_len(10)){
i = 2634
  filename <- sprintf("%s/%s.png", dirname, genes[i])
  png(file = filename, width = 600, height = 450, res = 100)
  plot_pseudotime_expression(
    GeneName = genes[i], merlot = merlot,
    EmbeddedTree = merlot_sub[["GenesSpaceEmbedding"]],
    Pseudotimes = merlot_sub[["Pseudotimes_highdim"]],
    branch_tags = c(), addlegend = F, range_y = "tree", switch = 5)
  dev.off()
}

#Output the principal tree expression data.
```{r, eval = FALSE}
genes <- rownames(mcf7_sub)
EmbeddedTree = merlot_sub[["GenesSpaceEmbedding"]]
Pseudotimes = merlot_sub[["Pseudotimes_highdim"]]
inds_node <- c()
for(i in seq_along(Pseudotimes[["Branches"]])){
  branch_i <- Pseudotimes[["Branches"]][[i]]
  aux <- sort(Pseudotimes$Times_yk[branch_i], index.return = T)
  inds_node <- c(inds_node, branch_i[aux$ix])
}
pt_max <- max(merlot[["Pseudotimes_highdim"]][["Times_yk"]])
pt_min <- min(merlot[["Pseudotimes_highdim"]][["Times_yk"]])
ptr <- (Pseudotimes[["Times_yk"]][inds_node] - pt_min) / (pt_max - pt_min)
ptmat <- data.frame(merlot_pseudotime = ptr)
for(i in seq_along(genes)){
#for(i in seq_len(100)){
  GeneName <- genes[i]
  Gene = which(colnames(EmbeddedTree$CellCoords) == GeneName)
  ExpressionGene = EmbeddedTree$Nodes[, Gene]
  ExpressionGeneOriginal = EmbeddedTree$CellCoords[, Gene]
  ptmat <- cbind(ptmat, EmbeddedTree$Nodes[inds_node, Gene])
  colnames(ptmat)[i + 1] <- GeneName
}
ptmat <- ptmat[order(ptmat$merlot_pseudotime), ]
rownames(ptmat) <- seq_len(nrow(ptmat))
```

Remove the duplicated rows.
```{r, eval = FALSE}
ptmat_bp <- ptmat[duplicated(ptmat[, seq_len(ncol(ptmat))]), ]
ptmat <- ptmat[!duplicated(ptmat[, seq_len(ncol(ptmat))]), ]
rownames(ptmat) <- seq_len(nrow(ptmat))
colnames(ptmat)[1] <- paste0("#", colnames(ptmat)[1])
```

Check the negative values.
```{r, eval = FALSE}
min_vector <- apply(ptmat, 2, min)
min(min_vector)
```
```
[1] -0.0152289
```

Add `abs(min_vector) + EPS(0.0001)` to the rows having negative values.
```{r, eval = FALSE}
eps <- 0.0001
pseudotime <- ptmat[, 1]
for(i in seq_len(ncol(ptmat))){
  if(min_vector[i] < 0){
    ptmat[, i] <- ptmat[, i] - min_vector[i] + eps
  }
}
ptmat[, 1] <- pseudotime
```

Compute the derivatives.
```{r, eval = FALSE}
i = 1
dt <- ptmat[i + 1, 1] - ptmat[i, 1]
ptmat_dev <- (ptmat[i + 1, ] - ptmat[i, ]) / dt
for(i in 2:(nrow(ptmat)-1)){
  temp <- (ptmat[i + 1, ] - ptmat[i - 1, ]) / (2 * dt)
  ptmat_dev <- rbind(ptmat_dev, temp)
}
i = nrow(ptmat)
ptmat_dev <- rbind(ptmat_dev, (ptmat[i, ] - ptmat[i - 1, ]) / dt)
ptmat_dev[, 1] <- ptmat[, 1]
rownames(ptmat_dev) <- seq_len(nrow(ptmat_dev))
```

Save the data.
```{r, eval = FALSE}
saveRDS(ptmat, file = "backup/01_007_ptmat_fromC1toC1woC2C3C4C5.rds")
saveRDS(ptmat_dev, file = "backup/01_007_ptmat_dev_fromC1toC1woC2C3C4C5.rds")

#Load the data.
ptmat <- readRDS("backup/01_007_ptmat_fromC1toC1woC2C3C4C5.rds")
ptmat_dev <- readRDS("backup/01_007_ptmat_dev_fromC1toC1woC2C3C4C5.rds")
```

Output the data.
```{r, eval = FALSE}
dirname <- "ptmat_fromC1toC1woC2C3C4C5"
dir.create(dirname)
filename <- sprintf("%s.txt", dirname)
write.table(ptmat, file = sprintf("%s/%s", dirname, filename),
            sep = "\t", row.names = FALSE, quote = FALSE)

filename <- sprintf("%s_dev.txt", dirname)
write.table(ptmat_dev, file = sprintf("%s/%s", dirname, filename),
            sep = "\t", row.names = FALSE, quote = FALSE)
```



<br>

# Pseudotime analysis
## From cluster 4 to 4 without clusters 1, 2, 3 and 5
mcf7 <- readRDS("backup/01_006_mcf7_reduced.rds")
merlot <- readRDS("backup/01_005_mcf7_merlot_nky60.rds")
mcf7_sub <- mcf7[, which(mcf7$merlot_clusters %in% c(4))]

merlot_sub <- merlot
merlot_sub[["Pseudotimes_highdim"]][["Cells2Branches"]] <- mcf7$merlot_clusters
mat <- merlot_sub[["GenesSpaceEmbedding"]][["CellCoords"]]
inds_cell <- which(rownames(mat) %in% colnames(mcf7_sub))
inds_gene <- which(colnames(mat) %in% rownames(mcf7_sub))

### Modify merlot object.
### GenesSpaceEmbedding
merlot_sub[["GenesSpaceEmbedding"]][["Nodes"]] <-
  merlot[["GenesSpaceEmbedding"]][["Nodes"]][, inds_gene]
merlot_sub[["GenesSpaceEmbedding"]][["Topology"]][["Endpoints"]] <- c(1)
merlot_sub[["GenesSpaceEmbedding"]][["Topology"]][["Branchpoints"]] <- c(6)
merlot_sub[["GenesSpaceEmbedding"]][["CellCoords"]] <-
  merlot[["GenesSpaceEmbedding"]][["CellCoords"]][inds_cell, inds_gene]
merlot_sub[["GenesSpaceEmbedding"]][["Cells2TreeNodes"]] <-
  merlot[["GenesSpaceEmbedding"]][["Cells2TreeNodes"]][inds_cell, ]
merlot_sub[["GenesSpaceEmbedding"]][["Cells2Branches"]] <-
  merlot[["GenesSpaceEmbedding"]][["Cells2Branches"]][inds_cell]
merlot_sub[["GenesSpaceEmbedding"]][["Branches"]][[5]] <- NULL
merlot_sub[["GenesSpaceEmbedding"]][["Branches"]][[4]] <- NULL
merlot_sub[["GenesSpaceEmbedding"]][["Branches"]][[3]] <- NULL
merlot_sub[["GenesSpaceEmbedding"]][["Branches"]][[2]] <- NULL

### Pseudotimes_highdim
merlot_sub[["Pseudotimes_highdim"]][["Times_cells"]] <-
  merlot[["Pseudotimes_highdim"]][["Times_cells"]][inds_cell]
merlot_sub[["Pseudotimes_highdim"]][["Projected_Times_Cells"]] <-
  merlot[["Pseudotimes_highdim"]][["Projected_Times_Cells"]][inds_cell]
merlot_sub[["Pseudotimes_highdim"]][["Branches"]][[5]] <- NULL
merlot_sub[["Pseudotimes_highdim"]][["Branches"]][[4]] <- NULL
merlot_sub[["Pseudotimes_highdim"]][["Branches"]][[3]] <- NULL
merlot_sub[["Pseudotimes_highdim"]][["Branches"]][[2]] <- NULL
merlot_sub[["Pseudotimes_highdim"]][["Cells2TreeNodes"]] <-
  merlot[["Pseudotimes_highdim"]][["Cells2TreeNodes"]][inds_cell, ]
merlot_sub[["Pseudotimes_highdim"]][["Cells2Branches"]] <-
  merlot[["Pseudotimes_highdim"]][["Cells2Branches"]][inds_cell]
merlot_sub[["Pseudotimes_highdim"]][["PseudoExpressionMatrix"]] <-
  merlot[["Pseudotimes_highdim"]][["PseudoExpressionMatrix"]][inds_cell, inds_gene]
# ----------------------------------------
# Show gene expressions as a function of pseudotime
# ----------------------------------------
source("R/myfuncs.R")
genes <- rownames(mcf7_sub)
dirname <- "fig_pt_fromC4toC4woC1C2C3C5"
dir.create(dirname)
#for(i in seq_along(genes)){
for(i in seq_len(10)){
i = 2634
  filename <- sprintf("%s/%s.png", dirname, genes[i])
  png(file = filename, width = 600, height = 450, res = 100)
  plot_pseudotime_expression(
    GeneName = genes[i], merlot = merlot,
    EmbeddedTree = merlot_sub[["GenesSpaceEmbedding"]],
    Pseudotimes = merlot_sub[["Pseudotimes_highdim"]],
    branch_tags = c(), addlegend = F, range_y = "tree", switch = 12)
  dev.off()
}

Output the principal tree expression data.
```{r, eval = FALSE}
genes <- rownames(mcf7_sub)
EmbeddedTree = merlot_sub[["GenesSpaceEmbedding"]]
Pseudotimes = merlot_sub[["Pseudotimes_highdim"]]
inds_node <- c()
for(i in seq_along(Pseudotimes[["Branches"]])){
  branch_i <- Pseudotimes[["Branches"]][[i]]
  aux <- sort(Pseudotimes$Times_yk[branch_i], index.return = T)
  inds_node <- c(inds_node, branch_i[aux$ix])
}
pt_max <- max(merlot[["Pseudotimes_highdim"]][["Times_yk"]])
pt_min <- min(merlot[["Pseudotimes_highdim"]][["Times_yk"]])
ptr <- (Pseudotimes[["Times_yk"]][inds_node] - pt_min) / (pt_max - pt_min)
ptmat <- data.frame(merlot_pseudotime = ptr)
for(i in seq_along(genes)){
#for(i in seq_len(100)){
  GeneName <- genes[i]
  Gene = which(colnames(EmbeddedTree$CellCoords) == GeneName)
  ExpressionGene = EmbeddedTree$Nodes[, Gene]
  ExpressionGeneOriginal = EmbeddedTree$CellCoords[, Gene]
  ptmat <- cbind(ptmat, EmbeddedTree$Nodes[inds_node, Gene])
  colnames(ptmat)[i + 1] <- GeneName
}
ptmat <- ptmat[order(ptmat$merlot_pseudotime), ]
rownames(ptmat) <- seq_len(nrow(ptmat))
```

Remove the duplicated rows.
```{r, eval = FALSE}
ptmat_bp <- ptmat[duplicated(ptmat[, seq_len(ncol(ptmat))]), ]
ptmat <- ptmat[!duplicated(ptmat[, seq_len(ncol(ptmat))]), ]
rownames(ptmat) <- seq_len(nrow(ptmat))
colnames(ptmat)[1] <- paste0("#", colnames(ptmat)[1])
```

Check the negative values.
```{r, eval = FALSE}
min_vector <- apply(ptmat, 2, min)
min(min_vector)
```
```
[1] -0.02426622
```

Add `abs(min_vector) + EPS(0.0001)` to the rows having negative values.
```{r, eval = FALSE}
eps <- 0.0001
pseudotime <- ptmat[, 1]
for(i in seq_len(ncol(ptmat))){
  if(min_vector[i] < 0){
    ptmat[, i] <- ptmat[, i] - min_vector[i] + eps
  }
}
ptmat[, 1] <- pseudotime
```

Compute the derivatives.
```{r, eval = FALSE}
i = 1
dt <- ptmat[i + 1, 1] - ptmat[i, 1]
ptmat_dev <- (ptmat[i + 1, ] - ptmat[i, ]) / dt
for(i in 2:(nrow(ptmat)-1)){
  temp <- (ptmat[i + 1, ] - ptmat[i - 1, ]) / (2 * dt)
  ptmat_dev <- rbind(ptmat_dev, temp)
}
i = nrow(ptmat)
ptmat_dev <- rbind(ptmat_dev, (ptmat[i, ] - ptmat[i - 1, ]) / dt)
ptmat_dev[, 1] <- ptmat[, 1]
rownames(ptmat_dev) <- seq_len(nrow(ptmat_dev))
```

Save the data.
```{r, eval = FALSE}
saveRDS(ptmat, file = "backup/01_007_ptmat_fromC4toC4woC1C2C3C5.rds")
saveRDS(ptmat_dev, file = "backup/01_007_ptmat_dev_fromC4toC4woC1C2C3C5.rds")

#Load the data.
ptmat <- readRDS("backup/01_007_ptmat_fromC4toC4woC1C2C3C5.rds")
ptmat_dev <- readRDS("backup/01_007_ptmat_dev_fromC4toC4woC1C2C3C5.rds")
```

Output the data.
```{r, eval = FALSE}
dirname <- "ptmat_fromC4toC4woC1C2C3C5"
dir.create(dirname)
filename <- sprintf("%s.txt", dirname)
write.table(ptmat, file = sprintf("%s/%s", dirname, filename),
            sep = "\t", row.names = FALSE, quote = FALSE)

filename <- sprintf("%s_dev.txt", dirname)
write.table(ptmat_dev, file = sprintf("%s/%s", dirname, filename),
            sep = "\t", row.names = FALSE, quote = FALSE)
```



<br>

## From cluster 5 to 5 without clusters 1, 2, 3 and 4
mcf7 <- readRDS("backup/01_006_mcf7_reduced.rds")
merlot <- readRDS("backup/01_004_mcf7_merlot_C3_Nyk60.rds")
mcf7_sub <- mcf7[, which(mcf7$merlot_clusters %in% c(5))]

merlot_sub <- merlot
merlot_sub[["Pseudotimes_highdim"]][["Cells2Branches"]] <- mcf7$merlot_clusters
mat <- merlot_sub[["GenesSpaceEmbedding"]][["CellCoords"]]
inds_cell <- which(rownames(mat) %in% colnames(mcf7_sub))
inds_gene <- which(colnames(mat) %in% rownames(mcf7_sub))

### Modify merlot object.
### GenesSpaceEmbedding
merlot_sub[["GenesSpaceEmbedding"]][["Nodes"]] <-
  merlot[["GenesSpaceEmbedding"]][["Nodes"]][, inds_gene]
merlot_sub[["GenesSpaceEmbedding"]][["Topology"]][["Endpoints"]] <- c(4)
merlot_sub[["GenesSpaceEmbedding"]][["Topology"]][["Branchpoints"]] <- c(6)
merlot_sub[["GenesSpaceEmbedding"]][["CellCoords"]] <-
  merlot[["GenesSpaceEmbedding"]][["CellCoords"]][inds_cell, inds_gene]
merlot_sub[["GenesSpaceEmbedding"]][["Cells2TreeNodes"]] <-
  merlot[["GenesSpaceEmbedding"]][["Cells2TreeNodes"]][inds_cell, ]
merlot_sub[["GenesSpaceEmbedding"]][["Cells2Branches"]] <-
  merlot[["GenesSpaceEmbedding"]][["Cells2Branches"]][inds_cell]
merlot_sub[["GenesSpaceEmbedding"]][["Branches"]][[5]] <- NULL
merlot_sub[["GenesSpaceEmbedding"]][["Branches"]][[4]] <- NULL
merlot_sub[["GenesSpaceEmbedding"]][["Branches"]][[3]] <- NULL
merlot_sub[["GenesSpaceEmbedding"]][["Branches"]][[1]] <- NULL

### Pseudotimes_highdim
merlot_sub[["Pseudotimes_highdim"]][["Times_cells"]] <-
  merlot[["Pseudotimes_highdim"]][["Times_cells"]][inds_cell]
merlot_sub[["Pseudotimes_highdim"]][["Projected_Times_Cells"]] <-
  merlot[["Pseudotimes_highdim"]][["Projected_Times_Cells"]][inds_cell]
merlot_sub[["Pseudotimes_highdim"]][["Branches"]][[5]] <- NULL
merlot_sub[["Pseudotimes_highdim"]][["Branches"]][[4]] <- NULL
merlot_sub[["Pseudotimes_highdim"]][["Branches"]][[3]] <- NULL
merlot_sub[["Pseudotimes_highdim"]][["Branches"]][[1]] <- NULL
merlot_sub[["Pseudotimes_highdim"]][["Cells2TreeNodes"]] <-
  merlot[["Pseudotimes_highdim"]][["Cells2TreeNodes"]][inds_cell, ]
merlot_sub[["Pseudotimes_highdim"]][["Cells2Branches"]] <-
  merlot[["Pseudotimes_highdim"]][["Cells2Branches"]][inds_cell]
merlot_sub[["Pseudotimes_highdim"]][["PseudoExpressionMatrix"]] <-
  merlot[["Pseudotimes_highdim"]][["PseudoExpressionMatrix"]][inds_cell, inds_gene]
# ----------------------------------------
# Show gene expressions as a function of pseudotime
# ----------------------------------------
source("R/myfuncs.R")
genes <- rownames(mcf7_sub)
dirname <- "fig_pt_fromC5toC5woC1C2C3C4"
dir.create(dirname)
#for(i in seq_along(genes)){
for(i in seq_len(10)){
i = 2634
  filename <- sprintf("%s/%s.png", dirname, genes[i])
  png(file = filename, width = 600, height = 450, res = 100)
  plot_pseudotime_expression(
    GeneName = genes[i], merlot = merlot,
    EmbeddedTree = merlot_sub[["GenesSpaceEmbedding"]],
    Pseudotimes = merlot_sub[["Pseudotimes_highdim"]],
    branch_tags = c(), addlegend = F, range_y = "tree", switch = 14)
  dev.off()
}
# ----------------------------------------
# Show derivative of gene expressions as a function of pseudotime
# ----------------------------------------
#i = 2634
source("R/myfuncs.R")
genes <- rownames(mcf7_sub)
dirname <- "fig_pt_fromC5toC5woC1C2C3C4_dev"
dir.create(dirname)
for(i in seq_len(10)){
  filename <- sprintf("%s/%s.png", dirname, genes[i])
  png(file = filename, width = 600, height = 450, res = 100)
  plot_pseudotime_differentiation(
    GeneName = genes[i], merlot = merlot,
    EmbeddedTree = merlot_sub[["GenesSpaceEmbedding"]],
    Pseudotimes = merlot_sub[["Pseudotimes_highdim"]],
    branch_tags = c(), addlegend = F, range_y = "tree", switch = 14)
  dev.off()
}



Output the principal tree expression data.
```{r, eval = FALSE}
genes <- rownames(mcf7_sub)
EmbeddedTree = merlot_sub[["GenesSpaceEmbedding"]]
Pseudotimes = merlot_sub[["Pseudotimes_highdim"]]
inds_node <- c()
for(i in seq_along(Pseudotimes[["Branches"]])){
  branch_i <- Pseudotimes[["Branches"]][[i]]
  aux <- sort(Pseudotimes$Times_yk[branch_i], index.return = T)
  inds_node <- c(inds_node, branch_i[aux$ix])
}
pt_max <- max(merlot[["Pseudotimes_highdim"]][["Times_yk"]])
pt_min <- min(merlot[["Pseudotimes_highdim"]][["Times_yk"]])
ptr <- (Pseudotimes[["Times_yk"]][inds_node] - pt_min) / (pt_max - pt_min)
ptmat <- data.frame(merlot_pseudotime = ptr)
for(i in seq_along(genes)){
#for(i in seq_len(100)){
  GeneName <- genes[i]
  Gene = which(colnames(EmbeddedTree$CellCoords) == GeneName)
  ExpressionGene = EmbeddedTree$Nodes[, Gene]
  ExpressionGeneOriginal = EmbeddedTree$CellCoords[, Gene]
  ptmat <- cbind(ptmat, EmbeddedTree$Nodes[inds_node, Gene])
  colnames(ptmat)[i + 1] <- GeneName
}
ptmat <- ptmat[order(ptmat$merlot_pseudotime), ]
rownames(ptmat) <- seq_len(nrow(ptmat))
```

Remove the duplicated rows.
```{r, eval = FALSE}
ptmat_bp <- ptmat[duplicated(ptmat[, seq_len(ncol(ptmat))]), ]
ptmat <- ptmat[!duplicated(ptmat[, seq_len(ncol(ptmat))]), ]
rownames(ptmat) <- seq_len(nrow(ptmat))
colnames(ptmat)[1] <- paste0("#", colnames(ptmat)[1])
```

Check the negative values.
```{r, eval = FALSE}
min_vector <- apply(ptmat, 2, min)
min(min_vector)
```
```
[1] -0.004252303
```

Add `abs(min_vector) + EPS(0.0001)` to the rows having negative values.
```{r, eval = FALSE}
eps <- 0.0001
pseudotime <- ptmat[, 1]
for(i in seq_len(ncol(ptmat))){
  if(min_vector[i] < 0){
    ptmat[, i] <- ptmat[, i] - min_vector[i] + eps
  }
}
ptmat[, 1] <- pseudotime
```

Compute the derivatives.
```{r, eval = FALSE}
i = 1
dt <- ptmat[i + 1, 1] - ptmat[i, 1]
ptmat_dev <- (ptmat[i + 1, ] - ptmat[i, ]) / dt
for(i in 2:(nrow(ptmat)-1)){
  temp <- (ptmat[i + 1, ] - ptmat[i - 1, ]) / (2 * dt)
  ptmat_dev <- rbind(ptmat_dev, temp)
}
i = nrow(ptmat)
ptmat_dev <- rbind(ptmat_dev, (ptmat[i, ] - ptmat[i - 1, ]) / dt)
ptmat_dev[, 1] <- ptmat[, 1]
rownames(ptmat_dev) <- seq_len(nrow(ptmat_dev))
```

Save the data.
```{r, eval = FALSE}
saveRDS(ptmat, file = "backup/01_007_ptmat_fromC5toC5woC1C2C3C4.rds")
saveRDS(ptmat_dev, file = "backup/01_007_ptmat_dev_fromC5toC5woC1C2C3C4.rds")

#Load the data.
ptmat <- readRDS("backup/01_007_ptmat_fromC5toC5woC1C2C3C4.rds")
ptmat_dev <- readRDS("backup/01_007_ptmat_dev_fromC5toC5woC1C2C3C4.rds")
```

Output the data.
```{r, eval = FALSE}
dirname <- "ptmat_fromC5toC5woC1C2C3C4"
dir.create(dirname)
filename <- sprintf("%s.txt", dirname)
write.table(ptmat, file = sprintf("%s/%s", dirname, filename),
            sep = "\t", row.names = FALSE, quote = FALSE)

filename <- sprintf("%s_dev.txt", dirname)
write.table(ptmat_dev, file = sprintf("%s/%s", dirname, filename),
            sep = "\t", row.names = FALSE, quote = FALSE)
```



<br>

## From cluster 1 to 5 without cluster 2
mcf7 <- readRDS("backup/01_006_mcf7_reduced.rds")
merlot <- readRDS("robustness_abcdConstraint_C1/merlot_data/01_004_mcf7_merlot_C3_Nyk60.rds")
mcf7_sub <- mcf7[, which(mcf7$merlot_clusters %in% c(1, 3, 4, 5))]

merlot_sub <- merlot
merlot_sub[["Pseudotimes_highdim"]][["Cells2Branches"]] <- mcf7$merlot_clusters
mat <- merlot_sub[["GenesSpaceEmbedding"]][["CellCoords"]]
inds_cell <- which(rownames(mat) %in% colnames(mcf7_sub))
inds_gene <- which(colnames(mat) %in% rownames(mcf7_sub))

### Modify merlot object.
### GenesSpaceEmbedding
merlot_sub[["GenesSpaceEmbedding"]][["Nodes"]] <-
  merlot[["GenesSpaceEmbedding"]][["Nodes"]][, inds_gene]
merlot_sub[["GenesSpaceEmbedding"]][["Topology"]][["Endpoints"]] <- c(1, 2, 4)
merlot_sub[["GenesSpaceEmbedding"]][["CellCoords"]] <-
  merlot[["GenesSpaceEmbedding"]][["CellCoords"]][inds_cell, inds_gene]
merlot_sub[["GenesSpaceEmbedding"]][["Cells2TreeNodes"]] <-
  merlot[["GenesSpaceEmbedding"]][["Cells2TreeNodes"]][inds_cell, ]
merlot_sub[["GenesSpaceEmbedding"]][["Cells2Branches"]] <-
  merlot[["GenesSpaceEmbedding"]][["Cells2Branches"]][inds_cell]
merlot_sub[["GenesSpaceEmbedding"]][["Branches"]][[5]] <- NULL

### Pseudotimes_highdim
merlot_sub[["Pseudotimes_highdim"]][["Times_cells"]] <-
  merlot[["Pseudotimes_highdim"]][["Times_cells"]][inds_cell]
merlot_sub[["Pseudotimes_highdim"]][["Projected_Times_Cells"]] <-
  merlot[["Pseudotimes_highdim"]][["Projected_Times_Cells"]][inds_cell]
merlot_sub[["Pseudotimes_highdim"]][["Branches"]][[5]] <- NULL
merlot_sub[["Pseudotimes_highdim"]][["Cells2TreeNodes"]] <-
  merlot[["Pseudotimes_highdim"]][["Cells2TreeNodes"]][inds_cell, ]
merlot_sub[["Pseudotimes_highdim"]][["Cells2Branches"]] <-
  merlot[["Pseudotimes_highdim"]][["Cells2Branches"]][inds_cell]
merlot_sub[["Pseudotimes_highdim"]][["PseudoExpressionMatrix"]] <-
  merlot[["Pseudotimes_highdim"]][["PseudoExpressionMatrix"]][inds_cell, inds_gene]
# ----------------------------------------
# Show gene expressions as a function of pseudotime
# ----------------------------------------
source("R/myfuncs.R")
genes <- rownames(mcf7_sub)
View(t(t(genes)))
dirname <- "fig_pt_fromC1toC4C5woC2_paper"
dir.create(dirname)
# ESR1 (769), GREB1 (5284), XBP1 (865), CREB1 (1890), NFKB1 (1412), FOS (4470),
# RHOA (403), NOTCH2 (2634), MAPK9 (259), RPS6KB1 (1369), TBL1XR1 (4817),
# MAPK14 (1534), JUN (4819)
goi <- c(769, 5284, 865, 1890, 1412, 4470, 403, 2634, 259, 1369, 4817, 1534,
         4819, 2234)
for(i in seq_along(goi)){
  filename <- sprintf("%s/%s.png", dirname, genes[goi[i]])
#  png(file = filename, width = 600, height = 450, res = 100)
  png(file = filename, width = 1250, height = 1200, res = 300)
  plot_pseudotime_expression(
    GeneName = genes[goi[i]], merlot = merlot,
    EmbeddedTree = merlot_sub[["GenesSpaceEmbedding"]],
    Pseudotimes = merlot_sub[["Pseudotimes_highdim"]],
    branch_tags = c(), addlegend = F, range_y = "tree", switch = 1)
  dev.off()
}



<br>

## From cluster 1 to 5 using all branches
mcf7 <- readRDS("backup/01_006_mcf7_reduced.rds")
merlot <- readRDS("robustness_abcdConstraint_C1/merlot_data/01_004_mcf7_merlot_C3_Nyk60.rds")
# ----------------------------------------
# Show gene expressions as a function of pseudotime
# ----------------------------------------
source("R/myfuncs.R")
genes <- rownames(mcf7_sub)
View(t(t(genes)))
dirname <- "fig_pt_fromC1toC5usingAll_paper"
dir.create(dirname)
goi <- c(1543, 1184, 396)
for(i in seq_along(goi)){
  filename <- sprintf("%s/%s.png", dirname, genes[goi[i]])
#  png(file = filename, width = 600, height = 450, res = 100)
  png(file = filename, width = 1250, height = 1200, res = 300)
  plot_pseudotime_expression(
    GeneName = genes[goi[i]], merlot = merlot,
    EmbeddedTree = merlot[["GenesSpaceEmbedding"]],
    Pseudotimes = merlot[["Pseudotimes_highdim"]],
    branch_tags = c(), addlegend = F, range_y = "tree", switch = 16)
  dev.off()
}



<br>

#===============================================================================
# Mathematical modeling
#===============================================================================
#-------------------------------------------------------------------------------
# Manual selection
#-------------------------------------------------------------------------------
mcf7 <- readRDS("backup/01_006_mcf7_reduced.rds")
markers <- rownames(mcf7)

#Load the data for C1
ptmat <- readRDS("backup/01_007_ptmat_fromC1toC1woC2C3C4C5.rds")
ptmat_dev <- readRDS("backup/01_007_ptmat_dev_fromC1toC1woC2C3C4C5.rds")
ptmat_red <- ptmat[, c("#merlot_pseudotime", markers)]
ptmat_dev_red <- ptmat_dev[, c("#merlot_pseudotime", markers)]

Save the data.
```{r, eval = FALSE}
type <- "fromC1toC1woC2C3C4C5"
saveRDS(ptmat_red, file = sprintf("backup/01_008_ptmat_red_%s.rds", type))
saveRDS(ptmat_dev_red, file = sprintf("backup/01_008_ptmat_dev_red_%s.rds", type))

#Load the data.
ptmat_red <- readRDS(sprintf("backup/01_008_ptmat_red_%s.rds", type))
ptmat_dev_red <- readRDS(sprintf("backup/01_008_ptmat_dev_red_%s.rds", type))
```

Output the data.
```{r, eval = FALSE}
type <- "fromC1toC1woC2C3C4C5"
dirname <- paste0("ptmat_genes_", type)
dir.create(dirname)

ncols <- ncol(ptmat_dev_red)
for(j in seq_len(ncols-1)){
  gene <- colnames(ptmat_dev_red)[j + 1]
  filename <- sprintf("%s.txt", gene)
  mat <- ptmat_red[, c(1, j + 1)]
  mat <- cbind(mat, ptmat_dev_red[, j + 1])
  colnames(mat)[3] <- paste0(gene, "_dxdt")
  write.table(mat, file = sprintf("%s/%s", dirname, filename),
              sep = "\t", row.names = FALSE, quote = FALSE)
}
```

Make a gnuplot script.
```{r, eval = FALSE}
type <- "fromC1toC1woC2C3C4C5"
dirname <- paste0("ptmat_genes_", type)
input_dirname <- paste0("gnuplot_input_", type)
output_dirname <- paste0("gnuplot_output_", type)
param_candi <- sprintf("param_candi_%s.txt", type)
dir.create(input_dirname)
dir.create(output_dirname)

ncols <- ncol(ptmat_dev_red)
for(j in seq_len(ncols-1)){
  gene <- colnames(ptmat_dev_red)[j + 1]
  infile <- sprintf("../%s/%s", dirname, sprintf("%s.txt", gene))
  f <- list()
  f[[1]] <- "reset"
  f[[2]] <- "f(x, y) = a**2 * y**(k**2) / (y**(k**2) + K**2) * (x**(m**2) / (x**(m**2) + M**2)) - b**2 * y**(l**2) / (y**(l**2) + L**2) * (x**(n**2) / (x**(n**2) + N**2)) + c**2 - d**2 * y"
  f[[3]] <- sprintf("stats \"%s\" u 1 every ::0::0 nooutput", infile)
  f[[4]] <- "x0 = STATS_max"
  f[[5]] <- sprintf("stats \"%s\" u 2 every ::0::0 nooutput", infile)
  f[[6]] <- "y0 = STATS_max"
  f[[7]] <- sprintf("stats \"%s\" u 3 every ::0::0 nooutput", infile)
  f[[8]] <- "dxdt0 = STATS_max"
  f[[9]] <- sprintf("stats \"%s\" u 1 every ::0::0 nooutput", param_candi)
  f[[10]] <- "a = STATS_max"
  f[[11]] <- sprintf("stats \"%s\" u 2 every ::0::0 nooutput", param_candi)
  f[[12]] <- "b = STATS_max"
  f[[13]] <- sprintf("stats \"%s\" u 3 every ::0::0 nooutput", param_candi)
  f[[14]] <- "c = STATS_max"
  f[[15]] <- "d = (c**2 - dxdt0) / y0"  # y0 ne 0
  f[[16]] <- sprintf("stats \"%s\" u 4 every ::0::0 nooutput", param_candi)
  f[[17]] <- "k = STATS_max"
  f[[18]] <- sprintf("stats \"%s\" u 5 every ::0::0 nooutput", param_candi)
  f[[19]] <- "K = STATS_max"
  f[[20]] <- sprintf("stats \"%s\" u 6 every ::0::0 nooutput", param_candi)
  f[[21]] <- "l = STATS_max"
  f[[22]] <- sprintf("stats \"%s\" u 7 every ::0::0 nooutput", param_candi)
  f[[23]] <- "L = STATS_max"
  f[[24]] <- sprintf("stats \"%s\" u 8 every ::0::0 nooutput", param_candi)
  f[[25]] <- "m = STATS_max"
  f[[26]] <- sprintf("stats \"%s\" u 9 every ::0::0 nooutput", param_candi)
  f[[27]] <- "M = STATS_max"
  f[[28]] <- sprintf("stats \"%s\" u 10 every ::0::0 nooutput", param_candi)
  f[[29]] <- "n = STATS_max"
  f[[30]] <- sprintf("stats \"%s\" u 11 every ::0::0 nooutput", param_candi)
  f[[31]] <- "N = STATS_max"
  f[[32]] <- "set fit quiet"
  f[[33]] <- paste0("fit f(x, y) '", infile,
                    "' u 1:2:3 via a, b, c, d, k, K, l, L, m, M, n, N")
  f[[34]] <- sprintf("#unset key; set title font \"Arial,20\"; set title \"%05d %s\"", j, gene)
  f[[35]] <- sprintf("#plot \"%s\" u 2:3 w p ps 1.5 pt 7, \"\" u 2:(f($1, $2)) w l lw 2", infile)
  f[[36]] <- "#pause 0.5"
  f[[37]] <- "abcd = sqrt(a**4 + b**4 + c**4 + d**4)"
  f[[38]] <- paste0("set print \"", sprintf("../%s/%s.txt",
                                            output_dirname, gene), "\"")
  f[[39]] <- "print \"#i t0 x0 a b c d k K l L m M n N abcd FIT_STDFIT\""
  f[[40]] <- "unset print"
  f[[41]] <- paste0("set print \"", sprintf("../%s/%s.txt",
                                            output_dirname, gene), "\" append")
  f[[42]] <- sprintf("print %d, x0, y0, a, b, c, d, k, K, l, L, m, M, n, N, abcd, FIT_STDFIT", j)
  f[[43]] <- "unset print"
  outfile <- sprintf("%s/%s", input_dirname, sprintf("%s.txt", gene))
  for(k in seq_along(f)){
    write.table(f[[k]], file = outfile, col.names = FALSE, row.names = FALSE,
                append = (k != 1), quote = FALSE)
  }
  f <- list()
  f[[1]] <- "#items[n]=\"a b c k K l L m M n N\""
  cnt <- 2
  for(ii in c(1, 0.1)){
    for(jj in c(1, 0.1)){
      for(kk in c(1, 0.1)){
        for(ll in c(1, 0.1)){
          f[[cnt]] <- sprintf("items[%d]=\"%2.1f %2.1f %2.1f %2.1f %2.1f %2.1f %2.1f %2.1f %2.1f %2.1f %2.1f\"",
                              cnt - 2, 1.0, 1.0, 1.0, ii, jj, ii, jj, kk, ll, kk, ll)
          cnt <- cnt + 1
        }
      }
    }
  }
  kkk = cnt
  f[[kkk]] <- ""
  f[[kkk+1]] <- "i=-1"
  f[[kkk+2]] <- "for item in \"${items[@]}\" ; do"
  f[[kkk+3]] <- sprintf("    if [ -s ../%s/%s.txt ]; then", output_dirname, gene)
  f[[kkk+4]] <- "        i=$((i+1))"
  f[[kkk+5]] <- sprintf("        data[$i]=`cat ../%s/%s.txt | tail -n 1`",
                        output_dirname, gene)
  f[[kkk+6]] <- sprintf("        rm ../%s/%s.txt", output_dirname, gene)
  f[[kkk+7]] <- sprintf("        echo \"${item}\" > %s", param_candi)
  f[[kkk+8]] <- sprintf("        gnuplot %s.txt", gene)
  f[[kkk+9]] <- "    else"
  f[[kkk+10]] <- sprintf("        echo \"${item}\" > %s", param_candi)
  f[[kkk+11]] <- sprintf("        gnuplot %s.txt", gene)
  f[[kkk+12]] <- "    fi"
  f[[kkk+13]] <- "done"
  f[[kkk+14]] <- ""
  f[[kkk+15]] <- "if [ `echo ${#data[@]}` = 0 ]; then"
  f[[kkk+16]] <- sprintf("    echo %s >> ../%s/miss.txt", gene, output_dirname)
  f[[kkk+17]] <- "else"
  f[[kkk+18]] <- sprintf("    echo \"#i t0 x0 a b c d k K l L m M n N abcd FIT_STDFIT\" > ../%s/%s.txt", output_dirname, gene)
  f[[kkk+19]] <- "    IFS=$'\\n'"
  f[[kkk+20]] <- sprintf("    echo \"${data[*]}\" >> ../%s/%s.txt", output_dirname, gene)
  f[[kkk+21]] <- "fi"
  outfile <- sprintf("%s/%s", input_dirname, sprintf("%s_all.sh", gene))
  for(k in seq_along(f)){
    write.table(f[[k]], file = outfile, col.names = FALSE, row.names = FALSE,
                append = (k != 1), quote = FALSE)
  }
}

gfile <- sprintf("%s/all.sh", input_dirname)
f <- sprintf("rm ../%s/miss.txt", output_dirname)
write.table(f, file = gfile, col.names = FALSE, row.names = FALSE,
            append = FALSE, quote = FALSE)
for(j in seq_len(ncols-1)){
  gene <- colnames(ptmat_dev_red)[j + 1]
  f <- sprintf("bash %s", sprintf("%s_all.sh", gene))
  gfile <- sprintf("%s/all.sh", input_dirname)
  write.table(f, file = gfile, col.names = FALSE, row.names = FALSE,
              append = TRUE, quote = FALSE)
}
```



#Load the data for C4
type <- "fromC4toC4woC1C2C3C5"
ptmat <- readRDS(sprintf("backup/01_007_ptmat_%s.rds", type))
ptmat_dev <- readRDS(sprintf("backup/01_007_ptmat_dev_%s.rds", type))
ptmat_red <- ptmat[, c("#merlot_pseudotime", markers)]
ptmat_dev_red <- ptmat_dev[, c("#merlot_pseudotime", markers)]

Save the data.
```{r, eval = FALSE}
type <- "fromC4toC4woC1C2C3C5"
saveRDS(ptmat_red, file = sprintf("backup/01_008_ptmat_red_%s.rds", type))
saveRDS(ptmat_dev_red, file = sprintf("backup/01_008_ptmat_dev_red_%s.rds", type))

#Load the data.
ptmat_red <- readRDS(sprintf("backup/01_008_ptmat_red_%s.rds", type))
ptmat_dev_red <- readRDS(sprintf("backup/01_008_ptmat_dev_red_%s.rds", type))
```

Output the data.
```{r, eval = FALSE}
type <- "fromC4toC4woC1C2C3C5"
dirname <- paste0("ptmat_genes_", type)
dir.create(dirname)

ncols <- ncol(ptmat_dev_red)
for(j in seq_len(ncols-1)){
  gene <- colnames(ptmat_dev_red)[j + 1]
  filename <- sprintf("%s.txt", gene)
  mat <- ptmat_red[, c(1, j + 1)]
  mat <- cbind(mat, ptmat_dev_red[, j + 1])
  colnames(mat)[3] <- paste0(gene, "_dxdt")
  write.table(mat, file = sprintf("%s/%s", dirname, filename),
              sep = "\t", row.names = FALSE, quote = FALSE)
}
```

Make a gnuplot script.
```{r, eval = FALSE}
type <- "fromC4toC4woC1C2C3C5"
dirname <- paste0("ptmat_genes_", type)
input_dirname <- paste0("gnuplot_input_", type)
output_dirname <- paste0("gnuplot_output_", type)
param_candi <- sprintf("param_candi_%s.txt", type)
dir.create(input_dirname)
dir.create(output_dirname)

ncols <- ncol(ptmat_dev_red)
for(j in seq_len(ncols-1)){
  gene <- colnames(ptmat_dev_red)[j + 1]
  infile <- sprintf("../%s/%s", dirname, sprintf("%s.txt", gene))
  f <- list()
  f[[1]] <- "reset"
  f[[2]] <- "f(x, y) = a**2 * y**(k**2) / (y**(k**2) + K**2) * (x**(m**2) / (x**(m**2) + M**2)) - b**2 * y**(l**2) / (y**(l**2) + L**2) * (x**(n**2) / (x**(n**2) + N**2)) + c**2 - d**2 * y"
  f[[3]] <- sprintf("stats \"%s\" u 1 every ::0::0 nooutput", infile)
  f[[4]] <- "x0 = STATS_max"
  f[[5]] <- sprintf("stats \"%s\" u 2 every ::0::0 nooutput", infile)
  f[[6]] <- "y0 = STATS_max"
  f[[7]] <- sprintf("stats \"%s\" u 3 every ::0::0 nooutput", infile)
  f[[8]] <- "dxdt0 = STATS_max"
  f[[9]] <- sprintf("stats \"%s\" u 1 every ::0::0 nooutput", param_candi)
  f[[10]] <- "a = STATS_max"
  f[[11]] <- sprintf("stats \"%s\" u 2 every ::0::0 nooutput", param_candi)
  f[[12]] <- "b = STATS_max"
  f[[13]] <- sprintf("stats \"%s\" u 3 every ::0::0 nooutput", param_candi)
  f[[14]] <- "c = STATS_max"
  f[[15]] <- "d = (c**2 - dxdt0) / y0"  # y0 ne 0
  f[[16]] <- sprintf("stats \"%s\" u 4 every ::0::0 nooutput", param_candi)
  f[[17]] <- "k = STATS_max"
  f[[18]] <- sprintf("stats \"%s\" u 5 every ::0::0 nooutput", param_candi)
  f[[19]] <- "K = STATS_max"
  f[[20]] <- sprintf("stats \"%s\" u 6 every ::0::0 nooutput", param_candi)
  f[[21]] <- "l = STATS_max"
  f[[22]] <- sprintf("stats \"%s\" u 7 every ::0::0 nooutput", param_candi)
  f[[23]] <- "L = STATS_max"
  f[[24]] <- sprintf("stats \"%s\" u 8 every ::0::0 nooutput", param_candi)
  f[[25]] <- "m = STATS_max"
  f[[26]] <- sprintf("stats \"%s\" u 9 every ::0::0 nooutput", param_candi)
  f[[27]] <- "M = STATS_max"
  f[[28]] <- sprintf("stats \"%s\" u 10 every ::0::0 nooutput", param_candi)
  f[[29]] <- "n = STATS_max"
  f[[30]] <- sprintf("stats \"%s\" u 11 every ::0::0 nooutput", param_candi)
  f[[31]] <- "N = STATS_max"
  f[[32]] <- "set fit quiet"
  f[[33]] <- paste0("fit f(x, y) '", infile,
                    "' u 1:2:3 via a, b, c, d, k, K, l, L, m, M, n, N")
  f[[34]] <- sprintf("#unset key; set title font \"Arial,20\"; set title \"%05d %s\"", j, gene)
  f[[35]] <- sprintf("#plot \"%s\" u 2:3 w p ps 1.5 pt 7, \"\" u 2:(f($1, $2)) w l lw 2", infile)
  f[[36]] <- "#pause 0.5"
  f[[37]] <- "abcd = sqrt(a**4 + b**4 + c**4 + d**4)"
  f[[38]] <- paste0("set print \"", sprintf("../%s/%s.txt",
                                            output_dirname, gene), "\"")
  f[[39]] <- "print \"#i t0 x0 a b c d k K l L m M n N abcd FIT_STDFIT\""
  f[[40]] <- "unset print"
  f[[41]] <- paste0("set print \"", sprintf("../%s/%s.txt",
                                            output_dirname, gene), "\" append")
  f[[42]] <- sprintf("print %d, x0, y0, a, b, c, d, k, K, l, L, m, M, n, N, abcd, FIT_STDFIT", j)
  f[[43]] <- "unset print"
  outfile <- sprintf("%s/%s", input_dirname, sprintf("%s.txt", gene))
  for(k in seq_along(f)){
    write.table(f[[k]], file = outfile, col.names = FALSE, row.names = FALSE,
                append = (k != 1), quote = FALSE)
  }
  f <- list()
  f[[1]] <- "#items[n]=\"a b c k K l L m M n N\""
  cnt <- 2
  for(ii in c(1, 0.1)){
    for(jj in c(1, 0.1)){
      for(kk in c(1, 0.1)){
        for(ll in c(1, 0.1)){
          f[[cnt]] <- sprintf("items[%d]=\"%2.1f %2.1f %2.1f %2.1f %2.1f %2.1f %2.1f %2.1f %2.1f %2.1f %2.1f\"",
                              cnt - 2, 1.0, 1.0, 1.0, ii, jj, ii, jj, kk, ll, kk, ll)
          cnt <- cnt + 1
        }
      }
    }
  }
  kkk = cnt
  f[[kkk]] <- ""
  f[[kkk+1]] <- "i=-1"
  f[[kkk+2]] <- "for item in \"${items[@]}\" ; do"
  f[[kkk+3]] <- sprintf("    if [ -s ../%s/%s.txt ]; then", output_dirname, gene)
  f[[kkk+4]] <- "        i=$((i+1))"
  f[[kkk+5]] <- sprintf("        data[$i]=`cat ../%s/%s.txt | tail -n 1`",
                        output_dirname, gene)
  f[[kkk+6]] <- sprintf("        rm ../%s/%s.txt", output_dirname, gene)
  f[[kkk+7]] <- sprintf("        echo \"${item}\" > %s", param_candi)
  f[[kkk+8]] <- sprintf("        gnuplot %s.txt", gene)
  f[[kkk+9]] <- "    else"
  f[[kkk+10]] <- sprintf("        echo \"${item}\" > %s", param_candi)
  f[[kkk+11]] <- sprintf("        gnuplot %s.txt", gene)
  f[[kkk+12]] <- "    fi"
  f[[kkk+13]] <- "done"
  f[[kkk+14]] <- ""
  f[[kkk+15]] <- "if [ `echo ${#data[@]}` = 0 ]; then"
  f[[kkk+16]] <- sprintf("    echo %s >> ../%s/miss.txt", gene, output_dirname)
  f[[kkk+17]] <- "else"
  f[[kkk+18]] <- sprintf("    echo \"#i t0 x0 a b c d k K l L m M n N abcd FIT_STDFIT\" > ../%s/%s.txt", output_dirname, gene)
  f[[kkk+19]] <- "    IFS=$'\\n'"
  f[[kkk+20]] <- sprintf("    echo \"${data[*]}\" >> ../%s/%s.txt", output_dirname, gene)
  f[[kkk+21]] <- "fi"
  outfile <- sprintf("%s/%s", input_dirname, sprintf("%s_all.sh", gene))
  for(k in seq_along(f)){
    write.table(f[[k]], file = outfile, col.names = FALSE, row.names = FALSE,
                append = (k != 1), quote = FALSE)
  }
}

gfile <- sprintf("%s/all.sh", input_dirname)
f <- sprintf("rm ../%s/miss.txt", output_dirname)
write.table(f, file = gfile, col.names = FALSE, row.names = FALSE,
            append = FALSE, quote = FALSE)
for(j in seq_len(ncols-1)){
  gene <- colnames(ptmat_dev_red)[j + 1]
  f <- sprintf("bash %s", sprintf("%s_all.sh", gene))
  gfile <- sprintf("%s/all.sh", input_dirname)
  write.table(f, file = gfile, col.names = FALSE, row.names = FALSE,
              append = TRUE, quote = FALSE)
}
```



#Load the data for C5
mcf7 <- readRDS("backup/01_006_mcf7_reduced.rds")
markers <- rownames(mcf7)
type <- "fromC5toC5woC1C2C3C4"
ptmat <- readRDS(sprintf("backup/01_007_ptmat_%s.rds", type))
ptmat_dev <- readRDS(sprintf("backup/01_007_ptmat_dev_%s.rds", type))
ptmat_red <- ptmat[, c("#merlot_pseudotime", markers)]
ptmat_dev_red <- ptmat_dev[, c("#merlot_pseudotime", markers)]

ptmat_red <- ptmat_red[, c("#merlot_pseudotime", "RPS6KB1", "MAPK9")]
ptmat_dev_red <- ptmat_dev_red[, c("#merlot_pseudotime", "RPS6KB1", "MAPK9")]

Save the data.
```{r, eval = FALSE}
type <- "fromC5toC5woC1C2C3C4"
saveRDS(ptmat_red, file = sprintf("backup/01_008_ptmat_red_%s.rds", type))
saveRDS(ptmat_dev_red, file = sprintf("backup/01_008_ptmat_dev_red_%s.rds", type))

#Load the data.
ptmat_red <- readRDS(sprintf("backup/01_008_ptmat_red_%s.rds", type))
ptmat_dev_red <- readRDS(sprintf("backup/01_008_ptmat_dev_red_%s.rds", type))
```

Output the data.
```{r, eval = FALSE}
type <- "fromC5toC5woC1C2C3C4"
dirname <- paste0("ptmat_genes_", type)
dir.create(dirname)

ncols <- ncol(ptmat_dev_red)
for(j in seq_len(ncols-1)){
  gene <- colnames(ptmat_dev_red)[j + 1]
  filename <- sprintf("%s.txt", gene)
  mat <- ptmat_red[, c(1, j + 1)]
  mat <- cbind(mat, ptmat_dev_red[, j + 1])
  colnames(mat)[3] <- paste0(gene, "_dxdt")
  write.table(mat, file = sprintf("%s/%s", dirname, filename),
              sep = "\t", row.names = FALSE, quote = FALSE)
}
```

Make a gnuplot script.
```{r, eval = FALSE}
type <- "fromC1toC1woC2C3C4C5"
dirname <- paste0("ptmat_genes_", type)
input_dirname <- paste0("gnuplot_input_", type)
output_dirname <- paste0("gnuplot_output_", type)
param_candi <- sprintf("param_candi_%s.txt", type)
dir.create(input_dirname)
dir.create(output_dirname)

ncols <- ncol(ptmat_dev_red)
for(j in seq_len(ncols-1)){
  gene <- colnames(ptmat_dev_red)[j + 1]
  infile <- sprintf("../%s/%s", dirname, sprintf("%s.txt", gene))
  f <- list()
  f[[1]] <- "reset"
  f[[2]] <- "f(x, y) = a**2 * y**(k**2) / (y**(k**2) + K**2) * (x**(m**2) / (x**(m**2) + M**2)) - b**2 * y**(l**2) / (y**(l**2) + L**2) * (x**(n**2) / (x**(n**2) + N**2)) + c**2 - d**2 * y"
  f[[3]] <- sprintf("stats \"%s\" u 1 every ::0::0 nooutput", infile)
  f[[4]] <- "x0 = STATS_max"
  f[[5]] <- sprintf("stats \"%s\" u 2 every ::0::0 nooutput", infile)
  f[[6]] <- "y0 = STATS_max"
  f[[7]] <- sprintf("stats \"%s\" u 3 every ::0::0 nooutput", infile)
  f[[8]] <- "dxdt0 = STATS_max"
  f[[9]] <- sprintf("stats \"%s\" u 1 every ::0::0 nooutput", param_candi)
  f[[10]] <- "a = STATS_max"
  f[[11]] <- sprintf("stats \"%s\" u 2 every ::0::0 nooutput", param_candi)
  f[[12]] <- "b = STATS_max"
  f[[13]] <- sprintf("stats \"%s\" u 3 every ::0::0 nooutput", param_candi)
  f[[14]] <- "c = STATS_max"
  f[[15]] <- "d = (c**2 - dxdt0) / y0"  # y0 ne 0
  f[[16]] <- sprintf("stats \"%s\" u 4 every ::0::0 nooutput", param_candi)
  f[[17]] <- "k = STATS_max"
  f[[18]] <- sprintf("stats \"%s\" u 5 every ::0::0 nooutput", param_candi)
  f[[19]] <- "K = STATS_max"
  f[[20]] <- sprintf("stats \"%s\" u 6 every ::0::0 nooutput", param_candi)
  f[[21]] <- "l = STATS_max"
  f[[22]] <- sprintf("stats \"%s\" u 7 every ::0::0 nooutput", param_candi)
  f[[23]] <- "L = STATS_max"
  f[[24]] <- sprintf("stats \"%s\" u 8 every ::0::0 nooutput", param_candi)
  f[[25]] <- "m = STATS_max"
  f[[26]] <- sprintf("stats \"%s\" u 9 every ::0::0 nooutput", param_candi)
  f[[27]] <- "M = STATS_max"
  f[[28]] <- sprintf("stats \"%s\" u 10 every ::0::0 nooutput", param_candi)
  f[[29]] <- "n = STATS_max"
  f[[30]] <- sprintf("stats \"%s\" u 11 every ::0::0 nooutput", param_candi)
  f[[31]] <- "N = STATS_max"
  f[[32]] <- "set fit quiet"
  f[[33]] <- paste0("fit f(x, y) '", infile,
                    "' u 1:2:3 via a, b, c, d, k, K, l, L, m, M, n, N")
  f[[34]] <- sprintf("#unset key; set title font \"Arial,20\"; set title \"%05d %s\"", j, gene)
  f[[35]] <- sprintf("#plot \"%s\" u 2:3 w p ps 1.5 pt 7, \"\" u 2:(f($1, $2)) w l lw 2", infile)
  f[[36]] <- "#pause 0.5"
  f[[37]] <- "abcd = sqrt(a**4 + b**4 + c**4 + d**4)"
  f[[38]] <- paste0("set print \"", sprintf("../%s/%s.txt",
                                            output_dirname, gene), "\"")
  f[[39]] <- "print \"#i t0 x0 a b c d k K l L m M n N abcd FIT_STDFIT\""
  f[[40]] <- "unset print"
  f[[41]] <- paste0("set print \"", sprintf("../%s/%s.txt",
                                            output_dirname, gene), "\" append")
  f[[42]] <- sprintf("print %d, x0, y0, a, b, c, d, k, K, l, L, m, M, n, N, abcd, FIT_STDFIT", j)
  f[[43]] <- "unset print"
  outfile <- sprintf("%s/%s", input_dirname, sprintf("%s.txt", gene))
  for(k in seq_along(f)){
    write.table(f[[k]], file = outfile, col.names = FALSE, row.names = FALSE,
                append = (k != 1), quote = FALSE)
  }
  f <- list()
  f[[1]] <- "#items[n]=\"a b c k K l L m M n N\""
  cnt <- 2
  for(ii in c(1, 0.1)){
    for(jj in c(1, 0.1)){
      for(kk in c(1, 0.1)){
        for(ll in c(1, 0.1)){
          f[[cnt]] <- sprintf("items[%d]=\"%2.1f %2.1f %2.1f %2.1f %2.1f %2.1f %2.1f %2.1f %2.1f %2.1f %2.1f\"",
                              cnt - 2, 1.0, 1.0, 1.0, ii, jj, ii, jj, kk, ll, kk, ll)
          cnt <- cnt + 1
        }
      }
    }
  }
  kkk = cnt
  f[[kkk]] <- ""
  f[[kkk+1]] <- "i=-1"
  f[[kkk+2]] <- "for item in \"${items[@]}\" ; do"
  f[[kkk+3]] <- sprintf("    if [ -s ../%s/%s.txt ]; then", output_dirname, gene)
  f[[kkk+4]] <- "        i=$((i+1))"
  f[[kkk+5]] <- sprintf("        data[$i]=`cat ../%s/%s.txt | tail -n 1`",
                        output_dirname, gene)
  f[[kkk+6]] <- sprintf("        rm ../%s/%s.txt", output_dirname, gene)
  f[[kkk+7]] <- sprintf("        echo \"${item}\" > %s", param_candi)
  f[[kkk+8]] <- sprintf("        gnuplot %s.txt", gene)
  f[[kkk+9]] <- "    else"
  f[[kkk+10]] <- sprintf("        echo \"${item}\" > %s", param_candi)
  f[[kkk+11]] <- sprintf("        gnuplot %s.txt", gene)
  f[[kkk+12]] <- "    fi"
  f[[kkk+13]] <- "done"
  f[[kkk+14]] <- ""
  f[[kkk+15]] <- "if [ `echo ${#data[@]}` = 0 ]; then"
  f[[kkk+16]] <- sprintf("    echo \"#i t0 x0 a b c d k K l L m M n N abcd FIT_STDFIT\" > ../%s/%s.txt", output_dirname, gene)
  f[[kkk+17]] <- sprintf("    echo %s >> ../%s/miss.txt", gene, output_dirname)
  f[[kkk+18]] <- "else"
  f[[kkk+19]] <- sprintf("    echo \"#i t0 x0 a b c d k K l L m M n N abcd FIT_STDFIT\" > ../%s/%s.txt", output_dirname, gene)
  f[[kkk+20]] <- "    IFS=$'\\n'"
  f[[kkk+21]] <- sprintf("    echo \"${data[*]}\" >> ../%s/%s.txt", output_dirname, gene)
  f[[kkk+22]] <- "fi"
  outfile <- sprintf("%s/%s", input_dirname, sprintf("%s_all.sh", gene))
  for(k in seq_along(f)){
    write.table(f[[k]], file = outfile, col.names = FALSE, row.names = FALSE,
                append = (k != 1), quote = FALSE)
  }
}

gfile <- sprintf("%s/all.sh", input_dirname)
f <- sprintf("rm ../%s/miss.txt", output_dirname)
write.table(f, file = gfile, col.names = FALSE, row.names = FALSE,
            append = FALSE, quote = FALSE)
for(j in seq_len(ncols-1)){
  gene <- colnames(ptmat_dev_red)[j + 1]
  f <- sprintf("bash %s", sprintf("%s_all.sh", gene))
  gfile <- sprintf("%s/all.sh", input_dirname)
  write.table(f, file = gfile, col.names = FALSE, row.names = FALSE,
              append = TRUE, quote = FALSE)
}
```



<br>

# Summarize
```{r, eval = FALSE}
types <- c("fromC1toC1woC2C3C4C5", "fromC4toC4woC1C2C3C5", "fromC5toC5woC1C2C3C4")
output_dirname <- c()
for(i in seq_along(types)){
  output_dirname <- c(output_dirname, paste0("gnuplot_output_", types[i], "_result"))
  dir.create(output_dirname[i])
}
ptmat_devs_red <- list()
indirs <- list()
for(i in seq_along(types)){
  ptmat_devs_red[[i]] <- readRDS(sprintf("backup/01_008_ptmat_dev_red_%s.rds", types[i]))
  indirs[[i]] <- sprintf("gnuplot_output_%s", types[i])
}

for(i in seq_along(indirs)){
  ncols <- ncol(ptmat_devs_red[[i]])
  f <- "#i t0 x0 a b c d k K l L m M n N abcd FIT_STDFIT"
  outfile <- sprintf("%s/%s", output_dirname[i], sprintf("%s_all.txt", indirs[[i]]))
  write.table(f, file = outfile, col.names = FALSE, row.names = FALSE,
              append = FALSE, quote = FALSE)
  outfile <- sprintf("%s/%s", output_dirname[i], sprintf("%s_all.sh", indirs[[i]]))
  write.table(f, file = outfile, col.names = FALSE, row.names = FALSE,
              append = FALSE, quote = FALSE)
  for(j in 2:ncols){
    ############################################################################
    ### Exception handling #####################################################
    if(i == 1 & j == 1535){
      gene <- colnames(ptmat_devs_red[[i]])[j]
      infile <- sprintf("%s/%s", indirs[[i]], sprintf("%s.txt", gene))
      data <- read.table(infile, skip = 0, header = TRUE, comment.char = "",
                         check.names = FALSE, fill = TRUE)
      #---------------------------------------------------------------------------
      # Choose appropriate parameters.
      #---------------------------------------------------------------------------
      cand <- data[14, ]
      f <- sprintf("echo %s >> %s", paste(cand, collapse = " "),
                   sprintf("%s_all.txt", indirs[[i]]))
      write.table(f, file = outfile, col.names = FALSE, row.names = FALSE,
                  append = TRUE, quote = FALSE)
      next
    }
    ############################################################################
    gene <- colnames(ptmat_devs_red[[i]])[j]
    infile <- sprintf("%s/%s", indirs[[i]], sprintf("%s.txt", gene))
    data <- read.table(infile, skip = 0, header = TRUE, comment.char = "",
                       check.names = FALSE, fill = TRUE)
    #---------------------------------------------------------------------------
    # Choose appropriate parameters.
    #---------------------------------------------------------------------------
    err_th <- 0.01
    n = nrow(data[!is.na(data$FIT_STDFIT), ])
    if(n == 0){
      cand <- data[1, ]
      cand[, ncol(cand)] <- 9999
      f <- sprintf("echo %s >> %s", paste(cand, collapse = " "),
                   sprintf("%s_all.txt", indirs[[i]]))
      write.table(f, file = outfile, col.names = FALSE, row.names = FALSE,
                  append = TRUE, quote = FALSE)
    }else if(n == 1){
      cand <- data[!is.na(data$FIT_STDFIT), ]
      f <- sprintf("echo %s >> %s", paste(cand, collapse = " "),
                   sprintf("%s_all.txt", indirs[[i]]))
      write.table(f, file = outfile, col.names = FALSE, row.names = FALSE,
                  append = TRUE, quote = FALSE)
    }else{
      cand <- data[which(data$FIT_STDFIT < err_th), ]
      if(nrow(cand) > 0){
        cand <- cand[which(cand$abcd == min(cand$abcd, na.rm = TRUE)), ]
      }else{
        inds <- which(data$FIT_STDFIT == min(data$FIT_STDFIT, na.rm = TRUE))
        cand1 <- data[inds, ][1, ]
        cand2 <- data[-inds, ]
        if(nrow(cand2) > 0){
          inds <- which(cand2$FIT_STDFIT == min(cand2$FIT_STDFIT, na.rm = TRUE))
          cand2 <- cand2[inds, ][1, ]
          if(abs(cand1$FIT_STDFIT - cand2$FIT_STDFIT) < err_th){
            cand <- rbind(cand1, cand2)
            cand <- cand[which(cand$abcd == min(cand$abcd, na.rm = TRUE)), ][1, ]
          }else{
            cand <- cand1
          }
        }else{
          cand <- cand1
        }
      }
      f <- sprintf("echo %s >> %s", paste(cand, collapse = " "),
                   sprintf("%s_all.txt", indirs[[i]]))
      write.table(f, file = outfile, col.names = FALSE, row.names = FALSE,
                  append = TRUE, quote = FALSE)
    }
  }
}
```

```{bash, eval = FALSE}
cd gnuplot_output_fromC1toC1woC2C3C4C5_result
bash gnuplot_output_fromC1toC1woC2C3C4C5_all.sh
cd ..
cd gnuplot_output_fromC4toC4woC1C2C3C5_result
bash gnuplot_output_fromC4toC4woC1C2C3C5_all.sh
cd ..
cd gnuplot_output_fromC5toC5woC1C2C3C4_result
bash gnuplot_output_fromC5toC5woC1C2C3C4_all.sh
cd ..
```

```{r, eval = FALSE}
param_ptmat_dev <- list()
for(i in seq_along(indirs)){
  infile <- sprintf("%s/%s_all.txt", output_dirname[i], indirs[[i]])
  data <- read.table(infile, skip = 0, header = TRUE, comment.char = "",
                     check.names = FALSE)
  data[, 1] <- colnames(ptmat_devs_red[[i]])[-1]
  colnames(data)[1] <- "gene"
  param_ptmat_dev[[i]] <- data
  rownames(param_ptmat_dev[[i]]) <- seq_len(nrow(param_ptmat_dev[[i]]))
}
names(param_ptmat_dev) <- c("fromC1toC1woC2C3C4C5", "fromC4toC4woC1C2C3C5", 
                            "fromC5toC5woC1C2C3C4")
```

Save the data.
```{r, eval = FALSE}
saveRDS(param_ptmat_dev, file = "backup/01_009_param_ptmat_dev.rds")

#Load the data.
#param_ptmat_dev <- readRDS("backup/01_009_param_ptmat_dev.rds")
```



## Re-interpret the parameters
for(i in seq_along(param_ptmat_dev)){
  param_ptmat_dev[[i]]$a <- param_ptmat_dev[[i]]$a^2
  param_ptmat_dev[[i]]$b <- param_ptmat_dev[[i]]$b^2
  param_ptmat_dev[[i]]$c <- param_ptmat_dev[[i]]$c^2
  param_ptmat_dev[[i]]$d <- param_ptmat_dev[[i]]$d^2
  param_ptmat_dev[[i]]$k <- param_ptmat_dev[[i]]$k^2
  param_ptmat_dev[[i]]$K <- param_ptmat_dev[[i]]$K^2
  param_ptmat_dev[[i]]$l <- param_ptmat_dev[[i]]$l^2
  param_ptmat_dev[[i]]$L <- param_ptmat_dev[[i]]$L^2
  param_ptmat_dev[[i]]$m <- param_ptmat_dev[[i]]$m^2
  param_ptmat_dev[[i]]$M <- param_ptmat_dev[[i]]$M^2
  param_ptmat_dev[[i]]$n <- param_ptmat_dev[[i]]$n^2
  param_ptmat_dev[[i]]$N <- param_ptmat_dev[[i]]$N^2
}

Save the fixed data.
```{r, eval = FALSE}
saveRDS(param_ptmat_dev, file = "backup/01_010_param_ptmat_dev_reinterpret.rds")

#Load the data.
#param_ptmat_dev <- readRDS("backup/01_010_param_ptmat_dev_reinterpret.rds")
```



## Remove the genes with FIT_STDFIT >= 0.10 at least one dataset.
ng_genes <- c()
for(i in seq_along(param_ptmat_dev)){
  ng_genes <- c(ng_genes,
    param_ptmat_dev[[i]][which(param_ptmat_dev[[i]]$FIT_STDFIT >= 0.10), ]$gene)
}
ng_genes <- paste(ng_genes, collapse = "|")
ng_inds <- c()
for(i in seq_along(param_ptmat_dev)){
  ng_inds <- grep(ng_genes, param_ptmat_dev[[i]]$gene)
  param_ptmat_dev[[i]] <- param_ptmat_dev[[i]][-ng_inds, ]
  rownames(param_ptmat_dev[[i]]) <- seq_len(nrow(param_ptmat_dev[[i]]))
}

Save the fixed data.
```{r, eval = FALSE}
saveRDS(param_ptmat_dev, file = "backup/01_011_param_reinterpret_gte0.1.rds")

#Load the data.
#param_ptmat_dev <- readRDS("backup/01_011_param_reinterpret_gte0.1.rds")
```



## Visualize the estimation results.
```{r, eval = FALSE}
for(i in seq_along(param_ptmat_dev)){
  type <- names(param_ptmat_dev)[i]
  data <- param_ptmat_dev[[i]]
  df <- as.data.frame(data[, -1])
  # a
  title <- names(param_ptmat_dev)[i]
  p <- ggplot() + geom_histogram(aes(x = df$a), color = "grey35", fill = "grey35") +
    scale_x_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),
                  labels = scales::trans_format("log10",
                                                scales::math_format(10^.x))) +
    labs(title = title, x = "Estimated value of a", y = "Frequency") +
    theme_classic(base_size = 20) +
    theme(plot.title = element_text(hjust = 0.5, size = 20))
  filename <- sprintf("figures/figure_01_%04d_%s_a.png", 100, type)
  ggsave(file = filename, plot = p, dpi = 100, width = 7, height = 5)

  # b
  title <- names(param_ptmat_dev)[i]
  p <- ggplot() + geom_histogram(aes(x = df$b), color = "grey35", fill = "grey35") +
    scale_x_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),
                  labels = scales::trans_format("log10",
                                                scales::math_format(10^.x))) +
    labs(title = title, x = "Estimated value of b", y = "Frequency") +
    theme_classic(base_size = 20) +
    theme(plot.title = element_text(hjust = 0.5, size = 20))
  filename <- sprintf("figures/figure_01_%04d_%s_b.png", 100, type)
  ggsave(file = filename, plot = p, dpi = 100, width = 7, height = 5)
  
  # c
  title <- names(param_ptmat_dev)[i]
  p <- ggplot() + geom_histogram(aes(x = df$c), color = "grey35", fill = "grey35") +
    scale_x_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),
                  labels = scales::trans_format("log10",
                                                scales::math_format(10^.x))) +
    labs(title = title, x = "Estimated value of c", y = "Frequency") +
    theme_classic(base_size = 20) +
    theme(plot.title = element_text(hjust = 0.5, size = 20))
  filename <- sprintf("figures/figure_01_%04d_%s_c.png", 100, type)
  ggsave(file = filename, plot = p, dpi = 100, width = 7, height = 5)
  
  # d
  title <- names(param_ptmat_dev)[i]
  p <- ggplot() + geom_histogram(aes(x = df$d), color = "grey35", fill = "grey35") +
    scale_x_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),
                  labels = scales::trans_format("log10",
                                                scales::math_format(10^.x))) +
    labs(title = title, x = "Estimated value of d", y = "Frequency") +
    theme_classic(base_size = 20) +
    theme(plot.title = element_text(hjust = 0.5, size = 20))
  filename <- sprintf("figures/figure_01_%04d_%s_d.png", 100, type)
  ggsave(file = filename, plot = p, dpi = 100, width = 7, height = 5)
  
  # k
  title <- names(param_ptmat_dev)[i]
  p <- ggplot() + geom_histogram(aes(x = df$k), color = "grey35", fill = "grey35") +
    scale_x_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),
                  labels = scales::trans_format("log10",
                                                scales::math_format(10^.x))) +
    labs(title = title, x = "Estimated value of k", y = "Frequency") +
    theme_classic(base_size = 20) +
    theme(plot.title = element_text(hjust = 0.5, size = 20))
  filename <- sprintf("figures/figure_01_%04d_%s_k.png", 100, type)
  ggsave(file = filename, plot = p, dpi = 100, width = 7, height = 5)
  
  # K
  title <- names(param_ptmat_dev)[i]
  p <- ggplot() + geom_histogram(aes(x = df$K), color = "grey35", fill = "grey35") +
    scale_x_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),
                  labels = scales::trans_format("log10",
                                                scales::math_format(10^.x))) +
    labs(title = title, x = "Estimated value of K", y = "Frequency") +
    theme_classic(base_size = 20) +
    theme(plot.title = element_text(hjust = 0.5, size = 20))
  filename <- sprintf("figures/figure_01_%04d_%s_KK.png", 100, type)
  ggsave(file = filename, plot = p, dpi = 100, width = 7, height = 5)
  
  # l
  title <- names(param_ptmat_dev)[i]
  p <- ggplot() + geom_histogram(aes(x = df$l), color = "grey35", fill = "grey35") +
    scale_x_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),
                  labels = scales::trans_format("log10",
                                                scales::math_format(10^.x))) +
    labs(title = title, x = "Estimated value of l", y = "Frequency") +
    theme_classic(base_size = 20) +
    theme(plot.title = element_text(hjust = 0.5, size = 20))
  filename <- sprintf("figures/figure_01_%04d_%s_l.png", 100, type)
  ggsave(file = filename, plot = p, dpi = 100, width = 7, height = 5)
  
  # L
  title <- names(param_ptmat_dev)[i]
  p <- ggplot() + geom_histogram(aes(x = df$L), color = "grey35", fill = "grey35") +
    scale_x_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),
                  labels = scales::trans_format("log10",
                                                scales::math_format(10^.x))) +
    labs(title = title, x = "Estimated value of L", y = "Frequency") +
    theme_classic(base_size = 20) +
    theme(plot.title = element_text(hjust = 0.5, size = 20))
  filename <- sprintf("figures/figure_01_%04d_%s_LL.png", 100, type)
  ggsave(file = filename, plot = p, dpi = 100, width = 7, height = 5)

  # m
  title <- names(param_ptmat_dev)[i]
  p <- ggplot() + geom_histogram(aes(x = df$m), color = "grey35", fill = "grey35") +
    scale_x_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),
                  labels = scales::trans_format("log10",
                                                scales::math_format(10^.x))) +
    labs(title = title, x = "Estimated value of m", y = "Frequency") +
    theme_classic(base_size = 20) +
    theme(plot.title = element_text(hjust = 0.5, size = 20))
  filename <- sprintf("figures/figure_01_%04d_%s_m.png", 100, type)
  ggsave(file = filename, plot = p, dpi = 100, width = 7, height = 5)
  
  # M
  title <- names(param_ptmat_dev)[i]
  p <- ggplot() + geom_histogram(aes(x = df$M), color = "grey35", fill = "grey35") +
    scale_x_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),
                  labels = scales::trans_format("log10",
                                                scales::math_format(10^.x))) +
    labs(title = title, x = "Estimated value of M", y = "Frequency") +
    theme_classic(base_size = 20) +
    theme(plot.title = element_text(hjust = 0.5, size = 20))
  filename <- sprintf("figures/figure_01_%04d_%s_MM.png", 100, type)
  ggsave(file = filename, plot = p, dpi = 100, width = 7, height = 5)
  
  # n
  title <- names(param_ptmat_dev)[i]
  p <- ggplot() + geom_histogram(aes(x = df$n), color = "grey35", fill = "grey35") +
    scale_x_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),
                  labels = scales::trans_format("log10",
                                                scales::math_format(10^.x))) +
    labs(title = title, x = "Estimated value of n", y = "Frequency") +
    theme_classic(base_size = 20) +
    theme(plot.title = element_text(hjust = 0.5, size = 20))
  filename <- sprintf("figures/figure_01_%04d_%s_n.png", 100, type)
  ggsave(file = filename, plot = p, dpi = 100, width = 7, height = 5)
  
  # N
  title <- names(param_ptmat_dev)[i]
  p <- ggplot() + geom_histogram(aes(x = df$N), color = "grey35", fill = "grey35") +
    scale_x_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),
                  labels = scales::trans_format("log10",
                                                scales::math_format(10^.x))) +
    labs(title = title, x = "Estimated value of N", y = "Frequency") +
    theme_classic(base_size = 20) +
    theme(plot.title = element_text(hjust = 0.5, size = 20))
  filename <- sprintf("figures/figure_01_%04d_%s_NN.png", 100, type)
  ggsave(file = filename, plot = p, dpi = 100, width = 7, height = 5)
  
  # FIT_STDFIT
  title <- names(param_ptmat_dev)[i]
  p <- ggplot() + geom_histogram(aes(x = df$FIT_STDFIT), color = "grey35", fill = "grey35") +
    labs(title = title, x = "Estimated value of FIT_STDFIT", y = "Frequency") +
    theme_classic(base_size = 20) +
    theme(plot.title = element_text(hjust = 0.5, size = 20))
  filename <- sprintf("figures/figure_01_%04d_%s_error.png", 100, type)
  ggsave(file = filename, plot = p, dpi = 300, width = 5, height = 5)
}
```

Compare the input and inferred data.
```{r, eval = FALSE}
f <- function(t, x, a, b, c, d, k, K, l, L, m, M, n, N){
  dxdt <- a * x**k / (x**k + K) * t**m / (t**m + M) - b * x**l / (x**l + L) * t**n / (t**n + N) + c - d * x
  return(dxdt)
}
#
# For derivative of timecourse
#
type <- "fromC1toC1woC2C3C4C5"
type2 <- "C1"
gene <- "MTCO1P12"
data <- read.table(sprintf("ptmat_genes_%s/%s.txt", type, gene),
                   skip = 0, header = TRUE, comment.char = "", check.names = FALSE)
a <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$a
b <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$b
c <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$c
d <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$d
k <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$k
K <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$K
l <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$l
L <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$L
m <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$m
M <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$M
n <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$n
N <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$N

res <- data.frame(t = data[, 1], x = NA)
for(i in seq_len(nrow(data))){
  res$x[i] <- f(data[i, 1], data[i, 2], a, b, c, d, k, K, l, L, m, M, n, N)
}
title <- paste0(gene, " (", type, ")")
p <- ggplot() +
  geom_point(aes(x = data[, 2], y = data[, 3]),
             size = 5, shape = 21, stroke = 2, color = "black", fill = "grey80") +
  geom_path(aes(x = data[, 2], y = res$x),
            size = 2, color = scales::hue_pal()(5)[3]) +
  labs(title = title, x = "x", y = "dx/dt") +
  theme_classic(base_size = 20) +
  theme(plot.title = element_text(hjust = 0.5, size = 20))
filename <- sprintf("figures/figure_01_0105_%s_%s_x_vs_dxdt.png", type2, gene)
ggsave(file = filename, plot = p, dpi = 300, width = 5.2, height = 5)

#
# For derivative of timecourse
#
title <- paste0(gene, " (", type, ")")
p <- ggplot() +
  geom_point(aes(x = data[, 1], y = data[, 3]),
             size = 5, shape = 21, stroke = 2, color = "black", fill = "grey80") +
  geom_path(aes(x = res$t, y = res$x), size = 1.5, color = "red") +
  labs(title = title, x = "Pseudotime", y = "Time derivative of expr level") +
  theme_classic(base_size = 20) +
  theme(plot.title = element_text(hjust = 0.5, size = 20))
filename <- sprintf("figures/figure_01_0105_%s_%s_dxdt.png", type2, gene)
ggsave(file = filename, plot = p, dpi = 300, width = 5, height = 5)

#
# For timecourse
#
res <- data.frame(t = data[, 1], x = NA)
i <- 1
res$x[i] <- data[i, 2]
for(i in 2:(nrow(data)-1)){
  res$x[i] <- res$x[i-1] + (res$t[i] - res$t[i-1]) * f(data[i-1, 1], data[i-1, 2], a, b, c, d, k, K, l, L, m, M, n, N)
}
i <- nrow(data)
res$x[i] <- res$x[i-1] + (res$t[i] - res$t[i-1]) * f(data[i-1, 1], data[i-1, 2], a, b, c, d, k, K, l, L, m, M, n, N)

title <- paste0(gene, " (", type, ")")
p <- ggplot() +
  geom_point(aes(x = data[, 1], y = data[, 2]),
             size = 5, shape = 21, stroke = 2, color = "black", fill = "grey80") +
  geom_path(aes(x = res$t, y = res$x), size = 2, color = scales::hue_pal()(5)[3]) +
  labs(title = title, x = "Pseudotime", y = "Expression") +
  theme_classic(base_size = 20) +
  theme(plot.title = element_text(hjust = 0.5, size = 20))
filename <- sprintf("figures/figure_01_0105_%s_%s_x.png", type2, gene)
ggsave(file = filename, plot = p, dpi = 300, width = 5, height = 5)
```



## Further check the estimated result.
f <- function(t, x, a, k, K, m, M){
  dxdt <- a * x**k / (x**k + K) * t**m / (t**m + M)
  return(dxdt)
}
fx <- function(t, x, a, k, K, m, M){
  dxdt <- x**k / (x**k + K)
  return(dxdt)
}
ft <- function(t, x, a, k, K, m, M){
  dxdt <- t**m / (t**m + M)
  return(dxdt)
}

g <- function(t, x, b, l, L, n, N){
  dxdt <- - b * x**l / (x**l + L) * t**n / (t**n + N)
  return(dxdt)
}
gx <- function(t, x, b, l, L, n, N){
  dxdt <- x**l / (x**l + L)
  return(dxdt)
}
gt <- function(t, x, b, l, L, n, N){
  dxdt <- t**n / (t**n + N)
  return(dxdt)
}

param_ptmat_dev <- readRDS("backup/01_011_param_reinterpret_gte0.1.rds")
names <- c("fromC1toC1woC2C3C4C5", "fromC4toC4woC1C2C3C5", "fromC5toC5woC1C2C3C4")

for(i in seq_along(param_ptmat_dev)){
  type <- names(param_ptmat_dev)[i]
  ptmat_red <- readRDS(sprintf("backup/01_008_ptmat_red_%s.rds", type))
  output_dirname <- paste0("check_param_", type)
  dir.create(output_dirname)
  nrow <- nrow(param_ptmat_dev[[type]])

  output_dirname2 <- sprintf("%s/act_inh", output_dirname)
  dir.create(output_dirname2)

  output_dirname3 <- sprintf("%s/act_feedback_time", output_dirname)
  dir.create(output_dirname3)

  output_dirname4 <- sprintf("%s/inh_feedback_time", output_dirname)
  dir.create(output_dirname4)

  for(j in seq_len(nrow)){
    gene <- param_ptmat_dev[[type]]$gene[j]
    t0 <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$t0
    a <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$a
    b <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$b
    c <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$c
    d <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$d
    k <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$k
    K <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$K
    l <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$l
    L <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$L
    m <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$m
    M <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$M
    n <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$n
    N <- param_ptmat_dev[[type]][which(param_ptmat_dev[[type]]$gene == gene), ]$N
    df <- data.frame(t = ptmat_red[, 1],
                     F = f(ptmat_red[, 1], ptmat_red[, gene], a, k, K, m, M),
                     lb = "Activation")
    dg <- data.frame(t = ptmat_red[, 1],
                     F = g(ptmat_red[, 1], ptmat_red[, gene], b, l, L, n, N),
                     lb = "Inhibition")
    df <- rbind(df, dg)
    title <- paste0(gene, " (", type, ")")
    p <- ggplot() +
      geom_point(aes(x = df[, 1], y = df[, 2], fill = df[, 3]),
                 size = 3, shape = 21, stroke = 1, color = "black") +
      labs(title = title, x = "Pseudotime", y = "Contribution",
           fill = "") +
      theme_classic(base_size = 20) +
      theme(plot.title = element_text(hjust = 0.5, size = 16)) +
      scale_fill_brewer(palette = "Set1") +
      guides(colour = guide_legend(override.aes = list(size = 4)))
    filename <- sprintf("%s/%s.png", output_dirname2, gene)
    ggsave(file = filename, plot = p, dpi = 100, width = 7, height = 5)
#
    df <- data.frame(t = ptmat_red[, 1],
                     F = fx(ptmat_red[, 1], ptmat_red[, j + 1], a, k, K, m, M),
                     lb = "Feedback")
    dg <- data.frame(t = ptmat_red[, 1],
                     F = ft(ptmat_red[, 1], ptmat_red[, j + 1], a, k, K, m, M),
                     lb = "Non-feedback")
    df <- rbind(dg, df)
    title <- paste0(gene, " (", type, ")")
    p <- ggplot() +
      geom_point(aes(x = df[, 1], y = df[, 2], fill = df[, 3]),
                 size = 3, shape = 21, stroke = 1, color = "black") +
      labs(title = title, x = "Pseudotime", y = "Activation", fill = "") +
      theme_classic(base_size = 20) +
      theme(plot.title = element_text(hjust = 0.5, size = 16)) +
      scale_fill_manual(values = c("Feedback" = "red", "Non-feedback" = "yellow")) +
      guides(colour = guide_legend(override.aes = list(size = 4)))
    filename <- sprintf("%s/%s.png", output_dirname3, gene)
    ggsave(file = filename, plot = p, dpi = 100, width = 7.7, height = 5)
#
    df <- data.frame(t = ptmat_red[, 1],
                     F = gx(ptmat_red[, 1], ptmat_red[, j + 1], b, l, L, n, N),
                     lb = "Feedback")
    dg <- data.frame(t = ptmat_red[, 1],
                     F = gt(ptmat_red[, 1], ptmat_red[, j + 1], b, l, L, n, N),
                     lb = "Non-feedback")
    df <- rbind(dg, df)
    title <- paste0(gene, " (", type, ")")
    p <- ggplot() +
      geom_point(aes(x = df[, 1], y = df[, 2], fill = df[, 3]),
                 size = 3, shape = 21, stroke = 1, color = "black") +
      labs(title = title, x = "Pseudotime", y = "Inhibition", fill = "") +
      theme_classic(base_size = 20) +
      theme(plot.title = element_text(hjust = 0.5, size = 16)) +
      scale_fill_manual(values = c("Feedback" = "cyan", "Non-feedback" = "yellow")) +
      guides(colour = guide_legend(override.aes = list(size = 4)))
    filename <- sprintf("%s/%s.png", output_dirname4, gene)
    ggsave(file = filename, plot = p, dpi = 100, width = 7.7, height = 5)
  }
}



## Bifurcation analysis
#Load data
param_ptmat_dev <- readRDS("backup/01_010_param_ptmat_dev_reinterpret.rds")

Make a gnuplot script.
```{r, eval = FALSE}
types <- c("fromC1toC1woC2C3C4C5", "fromC4toC4woC1C2C3C5", "fromC5toC5woC1C2C3C4")
for(i in seq_along(types)){
  dirname <- paste0("bifurcation_", types[i])
  dir.create(dirname)

  param_mat <- param_ptmat_dev[[types[i]]]
  nrow <- nrow(param_mat)
  for(j in seq_len(nrow)){
    f <- list()
    f[[1]] <- "reset"
    f[[2]] <- "f(x) = a * x**k / (x**k + K) - b * x**l / (x**l + L) + c - d * x"
    f[[3]] <- "\n"
    f[[4]] <- sprintf("a = %s; b = %s; c = %s; d = %s; k = %s; K = %s; l = %s; L = %s",
                      param_mat[j, ]$a, param_mat[j, ]$b, param_mat[j, ]$c,
                      param_mat[j, ]$d, param_mat[j, ]$k, param_mat[j, ]$K,
                      param_mat[j, ]$l, param_mat[j, ]$L)
    f[[5]] <- sprintf("set title font \"Arial,20\"; set title \"%05d %s\"",
                      j, param_mat[j, ]$gene)
    f[[6]] <- sprintf("plot [0:] 0 lt -1, f(x) lw 2")
    f[[7]] <- "pause -1"
    outfile <- sprintf("%s/%s", dirname, sprintf("%s.txt", param_mat[j, ]$gene))
    for(k in seq_along(f)){
      write.table(f[[k]], file = outfile, col.names = FALSE, row.names = FALSE,
                  append = (k != 1), quote = FALSE)
    }
    f <- list()
    f[[1]] <- sprintf("gnuplot %s.txt", param_mat[j, ]$gene)
    outfile <- sprintf("%s/%s", dirname, sprintf("%s_all.sh", param_mat[j, ]$gene))
    for(k in seq_along(f)){
      write.table(f[[k]], file = outfile, col.names = FALSE, row.names = FALSE,
                  append = (k != 1), quote = FALSE)
    }
  }
}

for(i in seq_along(types)){
  dirname <- paste0("bifurcation_", types[i])
  gfile <- sprintf("%s/all.sh", dirname)
  param_mat <- param_ptmat_dev[[types[i]]]
  nrow <- nrow(param_mat)
  for(j in seq_len(nrow)){
    f <- sprintf("bash %s", sprintf("%s_all.sh", param_mat[j, ]$gene))
    gfile <- sprintf("%s/all.sh", dirname)
    write.table(f, file = gfile, col.names = FALSE, row.names = FALSE,
                append = (j != 1), quote = FALSE)
  }
}
```



# Investigate multi-stability
param_ptmat_dev <- readRDS("backup/01_010_param_ptmat_dev_reinterpret.rds")
types <- c("fromC1toC1woC2C3C4C5", "fromC4toC4woC1C2C3C5", "fromC5toC5woC1C2C3C4")
data <- list()
for(i in seq_along(types)){
  data[[i]] <- read.table(sprintf("C/01_outgoing/%s.txt", types[i]),
                          skip = 0, header = TRUE, comment.char = "",
                          check.names = FALSE)
  colnames(data[[i]]) <- c("i", "NSP")
  param_ptmat_dev[[i]]$NSP <- data[[i]]$NSP
}

## Remove the genes with FIT_STDFIT >= 0.10 at least one dataset.
ng_genes <- c()
for(i in seq_along(param_ptmat_dev)){
  ng_genes <- c(ng_genes,
    param_ptmat_dev[[i]][which(param_ptmat_dev[[i]]$FIT_STDFIT >= 0.10), ]$gene)
}
ng_genes <- paste(ng_genes, collapse = "|")
ng_inds <- c()
for(i in seq_along(param_ptmat_dev)){
  ng_inds <- grep(ng_genes, param_ptmat_dev[[i]]$gene)
  param_ptmat_dev[[i]] <- param_ptmat_dev[[i]][-ng_inds, ]
  rownames(param_ptmat_dev[[i]]) <- seq_len(nrow(param_ptmat_dev[[i]]))
}
#-------------------------------------------------------------------------------

df <- list()
for(i in seq_along(types)){
  df[[i]] <- param_ptmat_dev[[i]]
  df[[i]] <- df[[i]][which(df[[i]]$NSP >= 2), ]
  print(nrow(df[[i]]))
}
```
[1] 670
[1] 561
[1] 690
```

df <- param_ptmat_dev[["fromC1toC1woC2C3C4C5"]]
dg <- param_ptmat_dev[["fromC4toC4woC1C2C3C5"]]
dh <- param_ptmat_dev[["fromC5toC5woC1C2C3C4"]]

df <- df[which(df$NSP >= 2), ]
dg <- dg[which(dg$NSP >= 2), ]
dh <- dh[which(dh$NSP >= 2), ]

library(gplots)
data <- list(C1 = df$gene, C4 = dg$gene, C5 = dh$gene)
filename <- "figures/figure_01_0110.png"
png(file = filename, width = 1000, height = 1000, res = 200)
venn(data)
dev.off()

Save the data.
```{r, eval = FALSE}
saveRDS(param_ptmat_dev, file = "backup/01_012_param_bifurcation_reinterpret.rds")

#Load the data.
#param_ptmat_dev <- readRDS("backup/01_012_param_bifurcation_reinterpret.rds")
```



<br>

#===============================================================================
# Schematic modeling (from C1 to C5)
#===============================================================================
mcf7 <- readRDS("backup/01_005_mcf7_seurat.rds")
#View(mcf7@misc[["FindAllMarkers"]])
pval_adj_th <- 1e-10
degs <- mcf7@misc[["FindAllMarkers"]][which(mcf7@misc[["FindAllMarkers"]]$p_val_adj <= pval_adj_th), ]

# Search as "<gene> estrogen receptor breast cancer ---------------------------"

selected_genes <- list()
selected_genes[[1]] <- c("GREB1", "AREG", "NUPR1", "NME1", "HDAC2", "FOS")
# We found that NME1 expression was negatively associated with EMT markers in many human cancers and was reduced in human breast tumor cell lines

selected_genes[[2]] <- c("GSTM3")
# Estrogen induces expression of BCAS3, a novel estrogen receptor-alpha coactivator...
# GSTM3: Induce AP.1 https://pubmed.ncbi.nlm.nih.gov/30054830/

selected_genes[[3]] <- c("NRAS")

#-------------------------------------------------------------------------------
# Manual selection
#-------------------------------------------------------------------------------
selected_genes[[4]] <- c("CREB1", "TBL1X", "YY1",
                         "AKT1", "MTOR",                    # Akt/MTOR
                         "RAF1", "MAPK1",                   # MAPK
                         "JUN", "JUNB", "FOSB",             # AP.1
                         "RELA", "NFKB1",                   # NFKB
                         "NOTCH2", "JAG1", "CD44",          # NOTCH and EMT
                         "ESR1", "FOXA1", "MYC")
markers <- unique(degs$gene)
markers <- markers[!grepl("^MT-", markers)]
markers <- intersect(markers, unlist(selected_genes))
markers <- c(markers, unlist(selected_genes))
markers <- unique(markers)
#View(t(t(markers)))

#-------------------------------------------------------------------------------
# From C1 to C5
#-------------------------------------------------------------------------------
ptmat_dev <- readRDS("backup/01_007_ptmat_dev_fromC1toC5woC2C4.rds")
ptmat_dev_scaled <- ptmat_dev
ptmat_dev_scaled <- ptmat_dev_scaled[, c("#merlot_pseudotime", markers)]
for(j in 2:(ncol(ptmat_dev_scaled))){
  if((length(ptmat_dev_scaled[, j][ptmat_dev_scaled[, j] >= 0]) >= 1) &
     (length(ptmat_dev_scaled[, j][ptmat_dev_scaled[, j] < 0]) >= 1)){
    max_po <- max(ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j])
    min_po <- min(ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j])
    max_ne <- max(ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j])
    min_ne <- min(ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j])
    if(max_po != min_po){
      ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] <-
        (ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] - min_po) / (max_po - min_po)
    }else{
      ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] <- 0
    }
    if(max_ne != min_ne){
      ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j] <-
        -(max_ne - ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j]) / (max_ne - min_ne)
    }else{
      ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j] <- 0
    }
  }else if(length(ptmat_dev_scaled[, j][ptmat_dev_scaled[, j] >= 0]) >= 1){
    max_po <- max(ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j])
    min_po <- min(ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j])
    if(max_po != min_po){
      ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] <-
        (ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] - min_po) / (max_po - min_po)
    }else{
      ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] <- 0
    }
  }else if(length(ptmat_dev_scaled[, j][ptmat_dev_scaled[, j] < 0]) >= 1){
    max_ne <- max(ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j])
    min_ne <- min(ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j])
    if(max_ne != min_ne){
      ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j] <-
        -(max_ne - ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j]) / (max_ne - min_ne)
    }else{
      ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j] <- 0
    }
  }
#  ptmat_dev_scaled[, j][which(abs(ptmat_dev_scaled[, j]) != 1)] <- 0
}

mat <- t(ptmat_dev_scaled[, -1])
colnames(mat) <- ptmat_dev_scaled[, 1]
hres <- hclust(amap::Dist(mat, method = "pearson"))
#hres <- hclust(amap::Dist(mat, method = "pearson"), method = "ward.D2")
split = data.frame(cutree(hres, k = 10))
colnames(split) <- "class"
split[which(split$class == 10), ] <- 11
split[which(split$class == 1), ] <- 22
split[which(split$class == 2), ] <- 33
split[which(split$class == 5), ] <- 44
split[which(split$class == 7), ] <- 55
split[which(split$class == 3), ] <- 66
split[which(split$class == 9), ] <- 77
split[which(split$class == 4), ] <- 88
split[which(split$class == 8), ] <- 99
split[which(split$class == 6), ] <- 100

split[which(split$class == 11), ] <- 1
split[which(split$class == 22), ] <- 2
split[which(split$class == 33), ] <- 3
split[which(split$class == 44), ] <- 4
split[which(split$class == 55), ] <- 5
split[which(split$class == 66), ] <- 6
split[which(split$class == 77), ] <- 7
split[which(split$class == 88), ] <- 8
split[which(split$class == 99), ] <- 9
split[which(split$class == 100), ] <- 10
row_colors <- c(rep("red", length(c(selected_genes[[1]], selected_genes[[2]],
                                      selected_genes[[3]]))),
                rep("black", length(c(selected_genes[[4]]))))
col_colors <- rep("black", 37)
fontsize <- c(rep(c(12, rep(0, 8)), 4), 12)
## Annotation ##################################################################
param_ptmat_dev <- readRDS("backup/01_012_param_bifurcation_reinterpret.rds")
param_others <- readRDS("individual_genes/backup/01_012_param_bifurcation_reinterpret.rds")
para <- list()
anno_row <- c()
for(i in seq_along(param_ptmat_dev)){
  para[[i]] <- rbind(param_ptmat_dev[[i]], param_others[[i]])
  para[[i]] <- para[[i]][which(para[[i]]$gene %in% markers), ]
  if(is.element(0, para[[i]]$NSP)){
    para[[i]][which(para[[i]]$NSP == 0), ]$NSP <- 1
  }
  anno_row <- cbind(anno_row, para[[i]]$NSP)
}
rownames(anno_row) <- para[[1]]$gene
colnames(anno_row) <- c("NSP_C1", "NSP_C2", "NSP_C4", "NSP_C5")
anno_row <- anno_row[rownames(mat), ]
anno_row <- as.data.frame(anno_row)
mypal <- paste(c("#000000", palette()[c(-1, -8)], "#9e9e9e"), "35", sep = "")
c1_1 <- mypal[3]
c1_2 <- scales::hue_pal()(5)[3]
c2_1 <- mypal[6]
c2_2 <- scales::hue_pal()(5)[5]
c4_1 <- mypal[2]
c4_2 <- scales::hue_pal()(5)[1]
c5_1 <- mypal[7]
c5_2 <- scales::hue_pal()(5)[2]
row_ha = rowAnnotation(NSP_C1 = anno_row$NSP_C1,
#                       NSP_C2 = anno_row$NSP_C2,
#                       NSP_C4 = anno_row$NSP_C4,
                       NSP_C5 = anno_row$NSP_C5,
                       col = list(NSP_C1 = c("1" = c1_1, "2" = c1_2),
#                                  NSP_C2 = c("1" = c2_1, "2" = c2_2),
#                                  NSP_C4 = c("1" = c4_1, "2" = c4_2),
                                  NSP_C5 = c("1" = c5_1, "2" = c5_2)),
                       annotation_name_gp= gpar(fontsize = 12))
################################################################################
filename <- "figures/figure_01_0500.png"
png(file = filename, height = 1800, width = 1500, res = 300)
Heatmap(mat, name = "d[Expr]", right_annotation = row_ha,
        row_names_gp = gpar(col = row_colors, fontsize = 12),
        column_names_gp = gpar(col = col_colors, fontsize = fontsize),
        column_names_rot = 90,
        row_split = factor(split$class, levels = unique(sort(split$class))),
        show_row_names = TRUE, show_column_dend = FALSE,
        column_order = colnames(mat), row_gap = unit(1., "mm"),
        cluster_row_slices = FALSE, show_row_dend = FALSE)
dev.off()


ptmat <- readRDS("backup/01_007_ptmat_fromC1toC5woC2C4.rds")
mat <- ptmat[, c("#merlot_pseudotime", markers)]
mat <- t(scale(mat[, -1]))
colnames(mat) <- ptmat[, 1]
row_order = c("ESR1", "NUPR1", "AREG", "GREB1", "NME1", "TBL1X", "HDAC2", "NRAS",
              "FOXA1", "RELA", "YY1", "MAPK1", "MYC", "CREB1", "AKT1",
              "NFKB1", "RAF1", "CD44", "GSTM3", "JAG1", "JUNB", "FOS", "FOSB",
              "JUN", "MTOR", "NOTCH2")
filename <- "figures/figure_01_0501.png"
png(file = filename, height = 1800, width = 1500, res = 300)
Heatmap(mat, name = "Expr", right_annotation = row_ha,
        row_names_gp = gpar(col = row_colors, fontsize = 12),
        column_names_gp = gpar(col = col_colors, fontsize = fontsize),
        column_names_rot = 90,
        row_split = factor(split$class, levels = unique(sort(split$class))),
        show_row_names = TRUE, show_column_dend = FALSE,
        column_order = colnames(mat), row_order = row_order,
        row_gap = unit(1., "mm"), cluster_row_slices = FALSE)
dev.off()



# Detail (fromC1toC5woC2C4)
# ptmat_dev
ptmat_dev <- readRDS("backup/01_007_ptmat_dev_fromC1toC5woC2C4.rds")
ptmat_dev_scaled <- ptmat_dev
for(j in 2:(ncol(ptmat_dev_scaled))){
  if((length(ptmat_dev_scaled[, j][ptmat_dev_scaled[, j] >= 0]) >= 1) &
     (length(ptmat_dev_scaled[, j][ptmat_dev_scaled[, j] < 0]) >= 1)){
    max_po <- max(ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j])
    min_po <- min(ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j])
    max_ne <- max(ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j])
    min_ne <- min(ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j])
    if(max_po != min_po){
      ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] <-
        (ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] - min_po) / (max_po - min_po)
    }else{
      ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] <- 0
    }
    if(max_ne != min_ne){
      ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j] <-
        -(max_ne - ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j]) / (max_ne - min_ne)
    }else{
      ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j] <- 0
    }
  }else if(length(ptmat_dev_scaled[, j][ptmat_dev_scaled[, j] >= 0]) >= 1){
    max_po <- max(ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j])
    min_po <- min(ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j])
    if(max_po != min_po){
      ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] <-
        (ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] - min_po) / (max_po - min_po)
    }else{
      ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] <- 0
    }
  }else if(length(ptmat_dev_scaled[, j][ptmat_dev_scaled[, j] < 0]) >= 1){
    max_ne <- max(ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j])
    min_ne <- min(ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j])
    if(max_ne != min_ne){
      ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j] <-
        -(max_ne - ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j]) / (max_ne - min_ne)
    }else{
      ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j] <- 0
    }
  }
#  ptmat_dev_scaled[, j][which(abs(ptmat_dev_scaled[, j]) != 1)] <- 0
}

markers <- c("MYC", "AKT1", "NFKB1", "JAG1", "NOTCH2")
ptmat_dev_scaled_red <- ptmat_dev_scaled[, c("#merlot_pseudotime", markers)]

fields <- factor(colnames(ptmat_dev_scaled_red)[-1], levels = colnames(ptmat_dev_scaled_red)[-1])
p <- ggplot(aes(x = ptmat_dev_scaled_red[, 1]), data = ptmat_dev_scaled_red)
p <- p + geom_hline(yintercept = 0, linetype = 'dashed', size = 1)
for(i in 1:length(fields)){
  loop_input = paste("geom_line(aes(y = ", fields[i],", color = '", fields[i],"'), size = 1.5)", sep = "")
  p <- p + eval(parse(text = loop_input))
}
p <- p + labs(title = "C1-C3-C5", x = "Pseudotime",
              y = "Scaled time derivative", color = "Gene") +
  theme_classic(base_size = 20) +
  theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  scale_color_manual(breaks = markers,
                     values = c("MYC" = "black",
                                "AKT1" = "grey30",
                                "NFKB1" = "grey45",
                                "JAG1" = "grey60",
                                "NOTCH2" = "grey75")) +
  guides(colour = guide_legend(override.aes = list(size = 4)))
filename <- "figures/figure_01_0505_ptmat_dif.png"
ggsave(file = filename, plot = p, dpi = 300, width = 8, height = 4)



markers <- c("TIMP1", "CD63", "ITGB1", "RHOA", "NFKB1", "RPN2", "ACTB")
ptmat_dev_scaled_red <- ptmat_dev_scaled[, c("#merlot_pseudotime", markers)]

fields <- factor(colnames(ptmat_dev_scaled_red)[-1], levels = colnames(ptmat_dev_scaled_red)[-1])
p <- ggplot(aes(x = ptmat_dev_scaled_red[, 1]), data = ptmat_dev_scaled_red)
p <- p + geom_hline(yintercept = 0, linetype = 'dashed', size = 1)
for(i in 1:length(fields)){
  loop_input = paste("geom_line(aes(y = ", fields[i],", color = '", fields[i],"'), size = 1.5)", sep = "")
  p <- p + eval(parse(text = loop_input))
}
p <- p + labs(title = "C1-C3-C5", x = "Pseudotime",
              y = "Scaled time derivative", color = "Gene") +
  theme_classic(base_size = 20) +
  theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  guides(colour = guide_legend(override.aes = list(size = 4)))
filename <- "figures/figure_01_0506_ptmat_dif.png"
ggsave(file = filename, plot = p, dpi = 300, width = 8, height = 4)



# ptmat
ptmat <- readRDS("backup/01_007_ptmat_fromC1toC5woC2C4.rds")
ptmat_scaled <- ptmat
ptmat_scaled <- scale(ptmat_scaled)
markers <- c("MYC", "AKT1", "NFKB1", "YY1", "JAG1", "NOTCH2")
ptmat_scaled_red <- as.data.frame(ptmat_scaled[, c("#merlot_pseudotime", markers)])
ptmat_scaled_red[, 1] <- ptmat[, 1]

fields <- factor(colnames(ptmat_scaled_red)[-1], levels = colnames(ptmat_scaled_red)[-1])
p <- ggplot(aes(x = ptmat_scaled_red[, 1]), data = ptmat_scaled_red)
p <- p + geom_hline(yintercept = 0, linetype = 'dashed', size = 1)
for(i in 1:length(fields)){
  loop_input = paste("geom_line(aes(y = ", fields[i],", color = '", fields[i],"'), size = 1.5)", sep = "")
  p <- p + eval(parse(text = loop_input))
}
p <- p + labs(title = "C1-C3-C5", x = "Pseudotime",
              y = "Scaled expression", color = "Gene") +
  theme_classic(base_size = 20) +
  theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  scale_color_manual(breaks = markers,
                     values = c("MYC" = "black",
                                "AKT1" = "grey25",
                                "NFKB1" = "grey40",
                                "YY1" = "grey55",
                                "JAG1" = "grey70",
                                "NOTCH2" = "grey85")) +
  guides(colour = guide_legend(override.aes = list(size = 4)))
filename <- "figures/figure_01_0506_ptmat.png"
ggsave(file = filename, plot = p, dpi = 300, width = 8, height = 4)



markers <- c("FOXA1", "BRD4", "ARID1A", "ATF4", "RET", "GDF15")
markers <- c("TIMP1", "CD63", "ITGB1", "RHOA", "NFKB1", "RPN2", "ACTB")
ptmat_dev_scaled_red <- ptmat_dev_scaled[, c("#merlot_pseudotime", markers)]

fields <- factor(colnames(ptmat_dev_scaled_red)[-1], levels = colnames(ptmat_dev_scaled_red)[-1])
p <- ggplot(aes(x = ptmat_dev_scaled_red[, 1]), data = ptmat_dev_scaled_red)
p <- p + geom_hline(yintercept = 0, linetype = 'dashed', size = 1)
for(i in 1:length(fields)){
  loop_input = paste("geom_line(aes(y = ", fields[i],", color = '", fields[i],"'), size = 1.5)", sep = "")
  p <- p + eval(parse(text = loop_input))
}
p <- p + labs(title = "C1-C3-C5", x = "Pseudotime",
              y = "Scaled time derivative", color = "Gene") +
  theme_classic(base_size = 20) +
  theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  guides(colour = guide_legend(override.aes = list(size = 4)))
filename <- "figures/figure_01_0507_ptmat_dif.png"
ggsave(file = filename, plot = p, dpi = 300, width = 8, height = 4)



# ptmat
ptmat <- readRDS("backup/01_007_ptmat_fromC1toC4woC2C5.rds")
ptmat_scaled <- ptmat
ptmat_scaled <- scale(ptmat_scaled)
ptmat_scaled_red <- as.data.frame(ptmat_scaled[, c("#merlot_pseudotime", markers)])
ptmat_scaled_red[, 1] <- ptmat[, 1]

fields <- factor(colnames(ptmat_scaled_red)[-1], levels = colnames(ptmat_scaled_red)[-1])
p <- ggplot(aes(x = ptmat_scaled_red[, 1]), data = ptmat_scaled_red)
p <- p + geom_hline(yintercept = 0, linetype = 'dashed', size = 1)
for(i in 1:length(fields)){
  loop_input = paste("geom_line(aes(y = ", fields[i],", color = '", fields[i],"'), size = 1.5)", sep = "")
  p <- p + eval(parse(text = loop_input))
}
p <- p + labs(title = "C1-C3-C5", x = "Pseudotime",
              y = "Scaled expression", color = "Gene") +
  theme_classic(base_size = 20) +
  theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  guides(colour = guide_legend(override.aes = list(size = 4)))
filename <- "figures/figure_01_0507_ptmat.png"
ggsave(file = filename, plot = p, dpi = 300, width = 8, height = 4)



<br>

#===============================================================================
# Schematic modeling (from C1 to C4)
#===============================================================================
mcf7 <- readRDS("backup/01_005_mcf7_seurat.rds")
#View(mcf7@misc[["FindAllMarkers"]])
pval_adj_th <- 1e-15
degs <- mcf7@misc[["FindAllMarkers"]][which(mcf7@misc[["FindAllMarkers"]]$p_val_adj <= pval_adj_th), ]

# Search as "<gene> estrogen receptor breast cancer ---------------------------"

selected_genes <- list()
selected_genes[[1]] <- c("GREB1", "AREG", "NUPR1", "NME1", "HDAC2", "FOS")
# We found that NME1 expression was negatively associated with EMT markers in many human cancers and was reduced in human breast tumor cell lines

selected_genes[[2]] <- c("GSTM3")
# Estrogen induces expression of BCAS3, a novel estrogen receptor-alpha coactivator...
# GSTM3: Induce AP.1 https://pubmed.ncbi.nlm.nih.gov/30054830/

selected_genes[[3]] <- c("NRAS")

#-------------------------------------------------------------------------------
# Manual selection
#-------------------------------------------------------------------------------
selected_genes[[4]] <- c("CREB1", "TBL1X", "YY1",
                         "AKT1", "MTOR",                    # Akt/MTOR
                         "RAF1", "MAPK1",                   # MAPK
                         "JUN", "JUNB", "FOSB",             # AP.1
                         "RELA", "NFKB1",                   # NFKB
                         "NOTCH2", "JAG1", "CD44",          # NOTCH and EMT
                         "ESR1", "FOXA1", "MYC")
markers <- unique(degs$gene)
markers <- markers[!grepl("^MT-", markers)]
markers <- intersect(markers, unlist(selected_genes))
markers <- c(markers, unlist(selected_genes))
markers <- unique(markers)
#View(t(t(markers)))

#-------------------------------------------------------------------------------
# From C1 to C4
#-------------------------------------------------------------------------------
ptmat_dev <- readRDS("backup/01_007_ptmat_dev_fromC1toC4woC2C5.rds")
ptmat_dev_scaled <- ptmat_dev
ptmat_dev_scaled <- ptmat_dev_scaled[, c("#merlot_pseudotime", markers)]
for(j in 2:(ncol(ptmat_dev_scaled))){
  if((length(ptmat_dev_scaled[, j][ptmat_dev_scaled[, j] >= 0]) >= 1) &
     (length(ptmat_dev_scaled[, j][ptmat_dev_scaled[, j] < 0]) >= 1)){
    max_po <- max(ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j])
    min_po <- min(ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j])
    max_ne <- max(ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j])
    min_ne <- min(ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j])
    if(max_po != min_po){
      ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] <-
        (ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] - min_po) / (max_po - min_po)
    }else{
      ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] <- 0
    }
    if(max_ne != min_ne){
      ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j] <-
        -(max_ne - ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j]) / (max_ne - min_ne)
    }else{
      ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j] <- 0
    }
  }else if(length(ptmat_dev_scaled[, j][ptmat_dev_scaled[, j] >= 0]) >= 1){
    max_po <- max(ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j])
    min_po <- min(ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j])
    if(max_po != min_po){
      ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] <-
        (ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] - min_po) / (max_po - min_po)
    }else{
      ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] <- 0
    }
  }else if(length(ptmat_dev_scaled[, j][ptmat_dev_scaled[, j] < 0]) >= 1){
    max_ne <- max(ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j])
    min_ne <- min(ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j])
    if(max_ne != min_ne){
      ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j] <-
        -(max_ne - ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j]) / (max_ne - min_ne)
    }else{
      ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j] <- 0
    }
  }
#  ptmat_dev_scaled[, j][which(abs(ptmat_dev_scaled[, j]) != 1)] <- 0
}

mat <- t(ptmat_dev_scaled[, -1])
colnames(mat) <- ptmat_dev_scaled[, 1]
hres <- hclust(amap::Dist(mat, method = "pearson"))
#hres <- hclust(amap::Dist(mat, method = "pearson"), method = "ward.D2")
split = data.frame(cutree(hres, k = 10))
colnames(split) <- "class"
split[which(split$class == 10), ] <- 11
split[which(split$class == 1), ] <- 22
split[which(split$class == 2), ] <- 33
split[which(split$class == 5), ] <- 44
split[which(split$class == 6), ] <- 55
split[which(split$class == 7), ] <- 66
split[which(split$class == 3), ] <- 77
split[which(split$class == 9), ] <- 88
split[which(split$class == 4), ] <- 99
split[which(split$class == 8), ] <- 100

split[which(split$class == 11), ] <- 1
split[which(split$class == 22), ] <- 2
split[which(split$class == 33), ] <- 3
split[which(split$class == 44), ] <- 4
split[which(split$class == 55), ] <- 5
split[which(split$class == 66), ] <- 6
split[which(split$class == 77), ] <- 7
split[which(split$class == 88), ] <- 8
split[which(split$class == 99), ] <- 9
split[which(split$class == 100), ] <- 10
row_colors <- c(rep("red", length(c(selected_genes[[1]], selected_genes[[2]],
                                      selected_genes[[3]]))),
                rep("black", length(c(selected_genes[[4]]))))
col_colors <- rep("black", 37)
fontsize <- c(rep(c(12, rep(0, 8)), 4), 12)
## Annotation ##################################################################
param_ptmat_dev <- readRDS("backup/01_012_param_bifurcation_reinterpret.rds")
param_others <- readRDS("individual_genes/backup/01_012_param_bifurcation_reinterpret.rds")
para <- list()
anno_row <- c()
for(i in seq_along(param_ptmat_dev)){
  para[[i]] <- rbind(param_ptmat_dev[[i]], param_others[[i]])
  para[[i]] <- para[[i]][which(para[[i]]$gene %in% markers), ]
  if(is.element(0, para[[i]]$NSP)){
    para[[i]][which(para[[i]]$NSP == 0), ]$NSP <- 1
  }
  anno_row <- cbind(anno_row, para[[i]]$NSP)
}
rownames(anno_row) <- para[[1]]$gene
colnames(anno_row) <- c("NSP_C1", "NSP_C2", "NSP_C4", "NSP_C5")
anno_row <- anno_row[rownames(mat), ]
anno_row <- as.data.frame(anno_row)
mypal <- paste(c("#000000", palette()[c(-1, -8)], "#9e9e9e"), "35", sep = "")
c1_1 <- mypal[3]
c1_2 <- scales::hue_pal()(5)[3]
c2_1 <- mypal[6]
c2_2 <- scales::hue_pal()(5)[5]
c4_1 <- mypal[2]
c4_2 <- scales::hue_pal()(5)[1]
c5_1 <- mypal[7]
c5_2 <- scales::hue_pal()(5)[2]
row_ha = rowAnnotation(NSP_C1 = anno_row$NSP_C1,
#                       NSP_C2 = anno_row$NSP_C2,
                       NSP_C4 = anno_row$NSP_C4,
#                       NSP_C5 = anno_row$NSP_C5,
                       col = list(NSP_C1 = c("1" = c1_1, "2" = c1_2),
#                                  NSP_C2 = c("1" = c2_1, "2" = c2_2),
                                  NSP_C4 = c("1" = c4_1, "2" = c4_2)),
#                                  NSP_C5 = c("1" = c5_1, "2" = c5_2)),
                       annotation_name_gp= gpar(fontsize = 12))
################################################################################
filename <- "figures/figure_01_0510.png"
png(file = filename, height = 1800, width = 1500, res = 300)
Heatmap(mat, name = "d[Expr]", right_annotation = row_ha,
        row_names_gp = gpar(col = row_colors, fontsize = 12),
        column_names_gp = gpar(col = col_colors, fontsize = fontsize),
        column_names_rot = 90,
        row_split = factor(split$class, levels = unique(sort(split$class))),
        show_row_names = TRUE, show_column_dend = FALSE,
        column_order = colnames(mat), row_gap = unit(1., "mm"),
        cluster_row_slices = FALSE, show_row_dend = FALSE)
dev.off()


ptmat <- readRDS("backup/01_007_ptmat_fromC1toC4woC2C5.rds")
mat <- ptmat[, c("#merlot_pseudotime", markers)]
mat <- t(scale(mat[, -1]))
colnames(mat) <- ptmat[, 1]
row_order = c("ESR1", "NUPR1", "AREG", "GREB1", "NME1", "TBL1X", "HDAC2",
              "NRAS", "AKT1", "CREB1", "MTOR", "RELA", "FOXA1", "MYC", "YY1",
              "MAPK1", "NOTCH2", "NFKB1", "RAF1", "CD44", "GSTM3",
              "JUNB", "FOSB", "JAG1", "FOS", "JUN")
filename <- "figures/figure_01_0511.png"
png(file = filename, height = 1800, width = 1500, res = 300)
Heatmap(mat, name = "Expr", right_annotation = row_ha,
        row_names_gp = gpar(col = row_colors, fontsize = 12),
        column_names_gp = gpar(col = col_colors, fontsize = fontsize),
        column_names_rot = 90,
        row_split = factor(split$class, levels = unique(sort(split$class))),
        show_row_names = TRUE, show_column_dend = FALSE,
        column_order = colnames(mat), row_order = row_order,
        row_gap = unit(1., "mm"), cluster_row_slices = FALSE)
dev.off()



# Detail (fromC1toC4woC2C5)
# ptmat_dev
ptmat_dev <- readRDS("backup/01_007_ptmat_dev_fromC1toC4woC2C5.rds")
ptmat_dev_scaled <- ptmat_dev
for(j in 2:(ncol(ptmat_dev_scaled))){
  if((length(ptmat_dev_scaled[, j][ptmat_dev_scaled[, j] >= 0]) >= 1) &
     (length(ptmat_dev_scaled[, j][ptmat_dev_scaled[, j] < 0]) >= 1)){
    max_po <- max(ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j])
    min_po <- min(ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j])
    max_ne <- max(ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j])
    min_ne <- min(ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j])
    if(max_po != min_po){
      ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] <-
        (ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] - min_po) / (max_po - min_po)
    }else{
      ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] <- 0
    }
    if(max_ne != min_ne){
      ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j] <-
        -(max_ne - ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j]) / (max_ne - min_ne)
    }else{
      ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j] <- 0
    }
  }else if(length(ptmat_dev_scaled[, j][ptmat_dev_scaled[, j] >= 0]) >= 1){
    max_po <- max(ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j])
    min_po <- min(ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j])
    if(max_po != min_po){
      ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] <-
        (ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] - min_po) / (max_po - min_po)
    }else{
      ptmat_dev_scaled[ptmat_dev_scaled[, j] >= 0, j] <- 0
    }
  }else if(length(ptmat_dev_scaled[, j][ptmat_dev_scaled[, j] < 0]) >= 1){
    max_ne <- max(ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j])
    min_ne <- min(ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j])
    if(max_ne != min_ne){
      ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j] <-
        -(max_ne - ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j]) / (max_ne - min_ne)
    }else{
      ptmat_dev_scaled[ptmat_dev_scaled[, j] < 0, j] <- 0
    }
  }
#  ptmat_dev_scaled[, j][which(abs(ptmat_dev_scaled[, j]) != 1)] <- 0
}

markers <- c("GREB1", "HDAC2", "CREB1", "NFKB1", "GSTM3", "FOS", "JUN")
ptmat_dev_scaled_red <- ptmat_dev_scaled[, c("#merlot_pseudotime", markers)]

fields <- factor(colnames(ptmat_dev_scaled_red)[-1], levels = colnames(ptmat_dev_scaled_red)[-1])
p <- ggplot(aes(x = ptmat_dev_scaled_red[, 1]), data = ptmat_dev_scaled_red)
p <- p + geom_hline(yintercept = 0, linetype = 'dashed', size = 1)
for(i in 1:length(fields)){
  loop_input = paste("geom_line(aes(y = ", fields[i],", color = '", fields[i],"'), size = 1.5)", sep = "")
  p <- p + eval(parse(text = loop_input))
}
p <- p + labs(title = "C1-C3-C4", x = "Pseudotime",
              y = "Scaled time derivative", color = "Gene") +
  theme_classic(base_size = 20) +
  theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  scale_color_manual(breaks = markers,
                     values = c("GREB1" = "black",
                                "HDAC2" = "grey15",
                                "CREB1" = "grey30",
                                "NFKB1" = "grey45",
                                "GSTM3" = "grey60",
                                "FOS" = "grey75",
                                "JUN" = "grey85")) +
  guides(colour = guide_legend(override.aes = list(size = 4)))
filename <- "figures/figure_01_0515_ptmat_dif.png"
ggsave(file = filename, plot = p, dpi = 300, width = 8, height = 4)



# ptmat
ptmat <- readRDS("backup/01_007_ptmat_fromC1toC4woC2C5.rds")
ptmat_scaled <- ptmat
ptmat_scaled <- scale(ptmat_scaled)
markers <- c("GREB1", "HDAC2", "CREB1", "NFKB1", "GSTM3", "FOS", "JUN")
ptmat_scaled_red <- as.data.frame(ptmat_scaled[, c("#merlot_pseudotime", markers)])
ptmat_scaled_red[, 1] <- ptmat[, 1]

fields <- factor(colnames(ptmat_scaled_red)[-1], levels = colnames(ptmat_scaled_red)[-1])
p <- ggplot(aes(x = ptmat_scaled_red[, 1]), data = ptmat_scaled_red)
p <- p + geom_hline(yintercept = 0, linetype = 'dashed', size = 1)
for(i in 1:length(fields)){
  loop_input = paste("geom_line(aes(y = ", fields[i],", color = '", fields[i],"'), size = 1.5)", sep = "")
  p <- p + eval(parse(text = loop_input))
}
p <- p + labs(title = "C1-C3-C4", x = "Pseudotime",
              y = "Scaled expression", color = "Gene") +
  theme_classic(base_size = 20) +
  theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  scale_color_manual(breaks = markers,
                     values = c("GREB1" = "black",
                                "HDAC2" = "grey15",
                                "CREB1" = "grey30",
                                "NFKB1" = "grey45",
                                "GSTM3" = "grey60",
                                "FOS" = "grey75",
                                "JUN" = "grey85")) +
  guides(colour = guide_legend(override.aes = list(size = 4)))
filename <- "figures/figure_01_0515_ptmat.png"
ggsave(file = filename, plot = p, dpi = 300, width = 8, height = 4)



markers <- c("FOXA1", "ARID1A", "BRD4", "RET", "ERBB3")
markers <- c("TIMP1", "CD63", "ITGB1", "RHOA", "NFKB1", "RPN2", "ACTB")
ptmat_dev_scaled_red <- ptmat_dev_scaled[, c("#merlot_pseudotime", markers)]

fields <- factor(colnames(ptmat_dev_scaled_red)[-1], levels = colnames(ptmat_dev_scaled_red)[-1])
p <- ggplot(aes(x = ptmat_dev_scaled_red[, 1]), data = ptmat_dev_scaled_red)
p <- p + geom_hline(yintercept = 0, linetype = 'dashed', size = 1)
for(i in 1:length(fields)){
  loop_input = paste("geom_line(aes(y = ", fields[i],", color = '", fields[i],"'), size = 1.5)", sep = "")
  p <- p + eval(parse(text = loop_input))
}
p <- p + labs(title = "C1-C3-C4", x = "Pseudotime",
              y = "Scaled time derivative", color = "Gene") +
  theme_classic(base_size = 20) +
  theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  guides(colour = guide_legend(override.aes = list(size = 4)))
filename <- "figures/figure_01_0515_ptmat_dif.png"
ggsave(file = filename, plot = p, dpi = 300, width = 8, height = 4)



# ptmat
ptmat <- readRDS("backup/01_007_ptmat_fromC1toC4woC2C5.rds")
ptmat_scaled <- ptmat
ptmat_scaled <- scale(ptmat_scaled)
ptmat_scaled_red <- as.data.frame(ptmat_scaled[, c("#merlot_pseudotime", markers)])
ptmat_scaled_red[, 1] <- ptmat[, 1]

fields <- factor(colnames(ptmat_scaled_red)[-1], levels = colnames(ptmat_scaled_red)[-1])
p <- ggplot(aes(x = ptmat_scaled_red[, 1]), data = ptmat_scaled_red)
p <- p + geom_hline(yintercept = 0, linetype = 'dashed', size = 1)
for(i in 1:length(fields)){
  loop_input = paste("geom_line(aes(y = ", fields[i],", color = '", fields[i],"'), size = 1.5)", sep = "")
  p <- p + eval(parse(text = loop_input))
}
p <- p + labs(title = "C1-C3-C4", x = "Pseudotime",
              y = "Scaled expression", color = "Gene") +
  theme_classic(base_size = 20) +
  theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  guides(colour = guide_legend(override.aes = list(size = 4)))
filename <- "figures/figure_01_0515_ptmat.png"
ggsave(file = filename, plot = p, dpi = 300, width = 8, height = 4)



<br>

################################################################################
# Potential computation for C1
# RPS6KB1 # MAP3K13 # YY1 # ARID1A, CDK7 # RBPJ # CDH1, CAV1, GSN, NME1 # CD9

gene <- "MAPK9"
type <- "fromC1toC1woC2C3C4C5"
f <- sprintf("gnuplot_output_%s/%s.txt", type, gene)
param <- read.table(file = f, sep = " ", header = F)
colnames(param) <- c("i", "t0", "x0", "a", "b", "c", "d", "m", "M", "n", "N", "FIT_STDFIT")
#
# For dxdt
#
f <- function(t, x, a, b, c, d, k, K, l, L, m, M, n, N){
  dxdt <- a * x**k / (x**k + K) * t**m / (t**m + M) - b * x**l / (x**l + L) * t**n / (t**n + N) + c - d * x
  return(dxdt)
}
param <- readRDS("backup/01_012_param_bifurcation_reinterpret.rds")[[type]]
param <- param[which(param$gene == gene), ]
t0 <- param$t0
x0 <- param$x0
a <- param$a
b <- param$b
c <- param$c
d <- param$d
k <- param$k
K <- param$K
l <- param$l
L <- param$L
m <- param$m
M <- param$M
n <- param$n
N <- param$N

tstep <- 400
dt <- 0.005
dx <- 0.01
potential <- data.frame(t = c(t0, t0 + dt * seq_len(tstep)))
poten <- c()
point <- c()
for(i in seq_len(nrow(potential))){
  t <- potential[i, 1]
  po <- c()
  if(i == 1){
    xx <- x0
  }else{
    xx <- xx + dt * f(t, xx, a, b, c, d, k, K, l, L, m, M, n, N)
  }
  dev <- 99
  for(j in 0:tstep){
    x <- dx * j
    fx <- x**(k + 1) * hypergeo(1, 1 + (1./k), 2 + (1./k), - x**k / K) / (k + 1) / K
    ft <- t**m / (t**m + M)
    gx <- x**(l + 1) * hypergeo(1, 1 + (1./l), 2 + (1./l), - x**l / L) / (l + 1) / L
    gt <- t**n / (t**n + N)
    tmp <- -to_real(a * fx * ft - b * gx * gt + c * x - 0.5 * d * x**2)[1]
    po <- c(po, tmp)
    dev_new <- abs(x - xx)
    if(dev_new < dev){
      yy <- tmp
    }
    dev <- dev_new
  }
  poten <- rbind(poten, po)
  point <- rbind(point, c(t, xx, yy))
}
potential <- cbind(potential, poten)
rownames(potential) <- as.character(potential[, 1])
potential <- potential[, -1]
colnames(potential) <- (as.integer(colnames(potential)) - 1) * dx
point <- as.data.frame(point)
colnames(point) <- c("t", "x", "y")
################################################################################
# Perturbation experiment
################################################################################
potential <- data.frame(t = c(t0, t0 + dt * seq_len(tstep)))
poten <- c()
point_perturbed <- c()
for(i in seq_len(nrow(potential))){
  t <- potential[i, 1]
  po <- c()
  if(i == 1){
    xx <- 0.8 * x0
  }else{
    xx <- xx + dt * f(t, xx, a, b, c, d, k, K, l, L, m, M, n, N)
  }
  dev <- 99
  for(j in 0:tstep){
    x <- dx * j
    fx <- x**(k + 1) * hypergeo(1, 1 + (1./k), 2 + (1./k), - x**k / K) / (k + 1) / K
    ft <- t**m / (t**m + M)
    gx <- x**(l + 1) * hypergeo(1, 1 + (1./l), 2 + (1./l), - x**l / L) / (l + 1) / L
    gt <- t**n / (t**n + N)
    tmp <- -to_real(a * fx * ft - b * gx * gt + c * x - 0.5 * d * x**2)[1]
    po <- c(po, tmp)
    dev_new <- abs(x - xx)
    if(dev_new < dev){
      yy <- tmp
    }
    dev <- dev_new
  }
  poten <- rbind(poten, po)
  point_perturbed <- rbind(point_perturbed, c(t, xx, yy))
}
potential <- cbind(potential, poten)
rownames(potential) <- as.character(potential[, 1])
potential <- potential[, -1]
colnames(potential) <- (as.integer(colnames(potential)) - 1) * dx
point_perturbed <- as.data.frame(point_perturbed)
colnames(point_perturbed) <- c("t", "x", "y")
################################################################################
# Perturbation experiment 2
################################################################################
potential <- data.frame(t = c(t0, t0 + dt * seq_len(tstep)))
poten <- c()
point_perturbed_2 <- c()
for(i in seq_len(nrow(potential))){
  t <- potential[i, 1]
  po <- c()
  if(i == 1){
    xx <- 0.6 * x0
  }else{
    xx <- xx + dt * f(t, xx, a, b, c, d, k, K, l, L, m, M, n, N)
  }
  dev <- 99
  for(j in 0:tstep){
    x <- dx * j
    fx <- x**(k + 1) * hypergeo(1, 1 + (1./k), 2 + (1./k), - x**k / K) / (k + 1) / K
    ft <- t**m / (t**m + M)
    gx <- x**(l + 1) * hypergeo(1, 1 + (1./l), 2 + (1./l), - x**l / L) / (l + 1) / L
    gt <- t**n / (t**n + N)
    tmp <- -to_real(a * fx * ft - b * gx * gt + c * x - 0.5 * d * x**2)[1]
    po <- c(po, tmp)
    dev_new <- abs(x - xx)
    if(dev_new < dev){
      yy <- tmp
    }
    dev <- dev_new
  }
  poten <- rbind(poten, po)
  point_perturbed_2 <- rbind(point_perturbed_2, c(t, xx, yy))
}
potential <- cbind(potential, poten)
rownames(potential) <- as.character(potential[, 1])
potential <- potential[, -1]
colnames(potential) <- (as.integer(colnames(potential)) - 1) * dx
point_perturbed_2 <- as.data.frame(point_perturbed_2)
colnames(point_perturbed_2) <- c("t", "x", "y")
################################################################################
################################################################################
# Perturbation experiment 3
################################################################################
potential <- data.frame(t = c(t0, t0 + dt * seq_len(tstep)))
poten <- c()
point_perturbed_3 <- c()
for(i in seq_len(nrow(potential))){
  t <- potential[i, 1]
  po <- c()
  if(i == 1){
    xx <- 1.2 * x0
  }else{
    xx <- xx + dt * f(t, xx, a, b, c, d, k, K, l, L, m, M, n, N)
  }
  dev <- 99
  for(j in 0:tstep){
    x <- dx * j
    fx <- x**(k + 1) * hypergeo(1, 1 + (1./k), 2 + (1./k), - x**k / K) / (k + 1) / K
    ft <- t**m / (t**m + M)
    gx <- x**(l + 1) * hypergeo(1, 1 + (1./l), 2 + (1./l), - x**l / L) / (l + 1) / L
    gt <- t**n / (t**n + N)
    tmp <- -to_real(a * fx * ft - b * gx * gt + c * x - 0.5 * d * x**2)[1]
    po <- c(po, tmp)
    dev_new <- abs(x - xx)
    if(dev_new < dev){
      yy <- tmp
    }
    dev <- dev_new
  }
  poten <- rbind(poten, po)
  point_perturbed_3 <- rbind(point_perturbed_3, c(t, xx, yy))
}
potential <- cbind(potential, poten)
rownames(potential) <- as.character(potential[, 1])
potential <- potential[, -1]
colnames(potential) <- (as.integer(colnames(potential)) - 1) * dx
point_perturbed_3 <- as.data.frame(point_perturbed_3)
colnames(point_perturbed_3) <- c("t", "x", "y")
################################################################################
# Perturbation experiment 4
################################################################################
potential <- data.frame(t = c(t0, t0 + dt * seq_len(tstep)))
poten <- c()
point_perturbed_4 <- c()
for(i in seq_len(nrow(potential))){
  t <- potential[i, 1]
  po <- c()
  if(i == 1){
    xx <- 1.4 * x0
  }else{
    xx <- xx + dt * f(t, xx, a, b, c, d, k, K, l, L, m, M, n, N)
  }
  dev <- 99
  for(j in 0:tstep){
    x <- dx * j
    fx <- x**(k + 1) * hypergeo(1, 1 + (1./k), 2 + (1./k), - x**k / K) / (k + 1) / K
    ft <- t**m / (t**m + M)
    gx <- x**(l + 1) * hypergeo(1, 1 + (1./l), 2 + (1./l), - x**l / L) / (l + 1) / L
    gt <- t**n / (t**n + N)
    tmp <- -to_real(a * fx * ft - b * gx * gt + c * x - 0.5 * d * x**2)[1]
    po <- c(po, tmp)
    dev_new <- abs(x - xx)
    if(dev_new < dev){
      yy <- tmp
    }
    dev <- dev_new
  }
  poten <- rbind(poten, po)
  point_perturbed_4 <- rbind(point_perturbed_4, c(t, xx, yy))
}
potential <- cbind(potential, poten)
rownames(potential) <- as.character(potential[, 1])
potential <- potential[, -1]
colnames(potential) <- (as.integer(colnames(potential)) - 1) * dx
point_perturbed_4 <- as.data.frame(point_perturbed_4)
colnames(point_perturbed_4) <- c("t", "x", "y")
################################################################################
df <- c()
for(j in seq_len(ncol(potential))){
  for(i in seq_len(nrow(potential))){
    df <- rbind(df, c(rownames(potential)[i], colnames(potential)[j], potential[i, j]))
  }
  df <- rbind(df, "\n")
}
df <- as.data.frame(df)
colnames(df)[1] <- "#V1"
dirname <- "data/potential"
dir.create(dirname)
filename <- sprintf("%s/poten_%s_%s.txt", dirname, gene, type)
write.table(df, file = filename,
            sep = "\t", row.names = FALSE, quote = FALSE)
dg <- point
colnames(dg)[1] <- "#t"
filename <- sprintf("%s/point_%s_%s.txt", dirname, gene, type)
write.table(dg, file = filename,
            sep = "\t", row.names = FALSE, quote = FALSE)
dg <- point_perturbed
colnames(dg)[1] <- "#t"
filename <- sprintf("%s/point_perturbed_%s_%s.txt", dirname, gene, type)
write.table(dg, file = filename,
            sep = "\t", row.names = FALSE, quote = FALSE)
dg <- point_perturbed_2
colnames(dg)[1] <- "#t"
filename <- sprintf("%s/point_perturbed_2_%s_%s.txt", dirname, gene, type)
write.table(dg, file = filename,
            sep = "\t", row.names = FALSE, quote = FALSE)
dg <- point_perturbed_3
colnames(dg)[1] <- "#t"
filename <- sprintf("%s/point_perturbed_3_%s_%s.txt", dirname, gene, type)
write.table(dg, file = filename,
            sep = "\t", row.names = FALSE, quote = FALSE)
dg <- point_perturbed_4
colnames(dg)[1] <- "#t"
filename <- sprintf("%s/point_perturbed_4_%s_%s.txt", dirname, gene, type)
write.table(dg, file = filename,
            sep = "\t", row.names = FALSE, quote = FALSE)

#Plot
dirname <- sprintf("fig_potential_%s_C5", gene)
dir.create(dirname)
for(i in seq_len(nrow(potential))){
  df <- data.frame(as.numeric(colnames(potential)), as.numeric(potential[i, ]))
  dg <- data.frame(point[i, 2], point[i, 3])
  dh <- data.frame(point_perturbed[i, 2], point_perturbed[i, 3])
  di <- data.frame(point_perturbed_2[i, 2], point_perturbed_2[i, 3])
  dj <- data.frame(point_perturbed_3[i, 2], point_perturbed_3[i, 3])
  dk <- data.frame(point_perturbed_4[i, 2], point_perturbed_4[i, 3])
  t1 = sprintf("t = %3.2f", t0 + dt * (i - 1))
  p <- ggplot() +
    geom_line(aes(x = df[, 1], y = df[, 2]), color = "black", size = 1, alpha = 1) +
    geom_point(aes(x = dk[, 1], y = dk[, 2]), shape = 21, fill = "green",
               color = "black", size = 5, alpha = 1, stroke = 1.5,) +
    geom_point(aes(x = di[, 1], y = di[, 2]), shape = 21, fill = "yellow",
               color = "black", size = 5, alpha = 1, stroke = 1.5) +
    geom_point(aes(x = dj[, 1], y = dj[, 2]), shape = 21, fill = "green3",
               color = "black", size = 5, alpha = 1, stroke = 1.5) +
    geom_point(aes(x = dh[, 1], y = dh[, 2]), shape = 21, fill = "orange",
               color = "black", size = 5, alpha = 1, stroke = 1.5) +
    geom_point(aes(x = dg[, 1], y = dg[, 2]), shape = 21, fill = "red",
               color = "black", size = 5, alpha = 1, stroke = 1.5) +
#    ylim(-40, 80) +
    labs(title = paste0(gene, " in ", type), x = "Expression", y = "Potential") +
    annotate("text", x = -Inf, y = Inf, label = t1, hjust = -0.35,
             vjust = 1.5, size = 6, colour = "black") +
    theme_classic(base_size = 20) +
    theme(plot.title = element_text(hjust = 0.5, size = 20)) +
    guides(colour = guide_legend(override.aes = list(size = 4)))
  filename <- sprintf("%s/%05d.png", dirname, i)
  ggsave(file = filename, plot = p, dpi = 100, width = 5, height = 4)
}



# Potential computation for C4
gene <- "MAPK9"
type <- "fromC4toC4woC1C2C3C5"
f <- sprintf("gnuplot_output_%s/%s.txt", type, gene)
param <- read.table(file = f, sep = " ", header = F)
colnames(param) <- c("i", "t0", "x0", "a", "b", "c", "d", "m", "M", "n", "N", "FIT_STDFIT")
#
# For dxdt
#
f <- function(t, x, a, b, c, d, k, K, l, L, m, M, n, N){
  dxdt <- a * x**k / (x**k + K) * t**m / (t**m + M) - b * x**l / (x**l + L) * t**n / (t**n + N) + c - d * x
  return(dxdt)
}
param <- readRDS("backup/01_012_param_bifurcation_reinterpret.rds")[[type]]
param <- param[which(param$gene == gene), ]
t0 <- param$t0
x0 <- param$x0
a <- param$a
b <- param$b
c <- param$c
d <- param$d
k <- param$k
K <- param$K
l <- param$l
L <- param$L
m <- param$m
M <- param$M
n <- param$n
N <- param$N

tstep <- 400
dt <- 0.005
dx <- 0.01
potential <- data.frame(t = c(t0, t0 + dt * seq_len(tstep)))
poten <- c()
point <- c()
for(i in seq_len(nrow(potential))){
  t <- potential[i, 1]
  po <- c()
  if(i == 1){
    xx <- x0
  }else{
    xx <- xx + dt * f(t, xx, a, b, c, d, k, K, l, L, m, M, n, N)
  }
  dev <- 99
  for(j in 0:tstep){
    x <- dx * j
    fx <- x**(k + 1) * hypergeo(1, 1 + (1./k), 2 + (1./k), - x**k / K) / (k + 1) / K
    ft <- t**m / (t**m + M)
    gx <- x**(l + 1) * hypergeo(1, 1 + (1./l), 2 + (1./l), - x**l / L) / (l + 1) / L
    gt <- t**n / (t**n + N)
    tmp <- -to_real(a * fx * ft - b * gx * gt + c * x - 0.5 * d * x**2)[1]
    po <- c(po, tmp)
    dev_new <- abs(x - xx)
    if(dev_new < dev){
      yy <- tmp
    }
    dev <- dev_new
  }
  poten <- rbind(poten, po)
  point <- rbind(point, c(t, xx, yy))
}
potential <- cbind(potential, poten)
rownames(potential) <- as.character(potential[, 1])
potential <- potential[, -1]
colnames(potential) <- (as.integer(colnames(potential)) - 1) * dx
point <- as.data.frame(point)
colnames(point) <- c("t", "x", "y")
################################################################################
# Perturbation experiment
################################################################################
potential <- data.frame(t = c(t0, t0 + dt * seq_len(tstep)))
poten <- c()
point_perturbed <- c()
for(i in seq_len(nrow(potential))){
  t <- potential[i, 1]
  po <- c()
  if(i == 1){
    xx <- 0.8 * x0
  }else{
    xx <- xx + dt * f(t, xx, a, b, c, d, k, K, l, L, m, M, n, N)
  }
  dev <- 99
  for(j in 0:tstep){
    x <- dx * j
    fx <- x**(k + 1) * hypergeo(1, 1 + (1./k), 2 + (1./k), - x**k / K) / (k + 1) / K
    ft <- t**m / (t**m + M)
    gx <- x**(l + 1) * hypergeo(1, 1 + (1./l), 2 + (1./l), - x**l / L) / (l + 1) / L
    gt <- t**n / (t**n + N)
    tmp <- -to_real(a * fx * ft - b * gx * gt + c * x - 0.5 * d * x**2)[1]
    po <- c(po, tmp)
    dev_new <- abs(x - xx)
    if(dev_new < dev){
      yy <- tmp
    }
    dev <- dev_new
  }
  poten <- rbind(poten, po)
  point_perturbed <- rbind(point_perturbed, c(t, xx, yy))
}
potential <- cbind(potential, poten)
rownames(potential) <- as.character(potential[, 1])
potential <- potential[, -1]
colnames(potential) <- (as.integer(colnames(potential)) - 1) * dx
point_perturbed <- as.data.frame(point_perturbed)
colnames(point_perturbed) <- c("t", "x", "y")
################################################################################
# Perturbation experiment 2
################################################################################
potential <- data.frame(t = c(t0, t0 + dt * seq_len(tstep)))
poten <- c()
point_perturbed_2 <- c()
for(i in seq_len(nrow(potential))){
  t <- potential[i, 1]
  po <- c()
  if(i == 1){
    xx <- 0.6 * x0
  }else{
    xx <- xx + dt * f(t, xx, a, b, c, d, k, K, l, L, m, M, n, N)
  }
  dev <- 99
  for(j in 0:tstep){
    x <- dx * j
    fx <- x**(k + 1) * hypergeo(1, 1 + (1./k), 2 + (1./k), - x**k / K) / (k + 1) / K
    ft <- t**m / (t**m + M)
    gx <- x**(l + 1) * hypergeo(1, 1 + (1./l), 2 + (1./l), - x**l / L) / (l + 1) / L
    gt <- t**n / (t**n + N)
    tmp <- -to_real(a * fx * ft - b * gx * gt + c * x - 0.5 * d * x**2)[1]
    po <- c(po, tmp)
    dev_new <- abs(x - xx)
    if(dev_new < dev){
      yy <- tmp
    }
    dev <- dev_new
  }
  poten <- rbind(poten, po)
  point_perturbed_2 <- rbind(point_perturbed_2, c(t, xx, yy))
}
potential <- cbind(potential, poten)
rownames(potential) <- as.character(potential[, 1])
potential <- potential[, -1]
colnames(potential) <- (as.integer(colnames(potential)) - 1) * dx
point_perturbed_2 <- as.data.frame(point_perturbed_2)
colnames(point_perturbed_2) <- c("t", "x", "y")
################################################################################
################################################################################
# Perturbation experiment 3
################################################################################
potential <- data.frame(t = c(t0, t0 + dt * seq_len(tstep)))
poten <- c()
point_perturbed_3 <- c()
for(i in seq_len(nrow(potential))){
  t <- potential[i, 1]
  po <- c()
  if(i == 1){
    xx <- 1.2 * x0
  }else{
    xx <- xx + dt * f(t, xx, a, b, c, d, k, K, l, L, m, M, n, N)
  }
  dev <- 99
  for(j in 0:tstep){
    x <- dx * j
    fx <- x**(k + 1) * hypergeo(1, 1 + (1./k), 2 + (1./k), - x**k / K) / (k + 1) / K
    ft <- t**m / (t**m + M)
    gx <- x**(l + 1) * hypergeo(1, 1 + (1./l), 2 + (1./l), - x**l / L) / (l + 1) / L
    gt <- t**n / (t**n + N)
    tmp <- -to_real(a * fx * ft - b * gx * gt + c * x - 0.5 * d * x**2)[1]
    po <- c(po, tmp)
    dev_new <- abs(x - xx)
    if(dev_new < dev){
      yy <- tmp
    }
    dev <- dev_new
  }
  poten <- rbind(poten, po)
  point_perturbed_3 <- rbind(point_perturbed_3, c(t, xx, yy))
}
potential <- cbind(potential, poten)
rownames(potential) <- as.character(potential[, 1])
potential <- potential[, -1]
colnames(potential) <- (as.integer(colnames(potential)) - 1) * dx
point_perturbed_3 <- as.data.frame(point_perturbed_3)
colnames(point_perturbed_3) <- c("t", "x", "y")
################################################################################
# Perturbation experiment 4
################################################################################
potential <- data.frame(t = c(t0, t0 + dt * seq_len(tstep)))
poten <- c()
point_perturbed_4 <- c()
for(i in seq_len(nrow(potential))){
  t <- potential[i, 1]
  po <- c()
  if(i == 1){
    xx <- 1.4 * x0
  }else{
    xx <- xx + dt * f(t, xx, a, b, c, d, k, K, l, L, m, M, n, N)
  }
  dev <- 99
  for(j in 0:tstep){
    x <- dx * j
    fx <- x**(k + 1) * hypergeo(1, 1 + (1./k), 2 + (1./k), - x**k / K) / (k + 1) / K
    ft <- t**m / (t**m + M)
    gx <- x**(l + 1) * hypergeo(1, 1 + (1./l), 2 + (1./l), - x**l / L) / (l + 1) / L
    gt <- t**n / (t**n + N)
    tmp <- -to_real(a * fx * ft - b * gx * gt + c * x - 0.5 * d * x**2)[1]
    po <- c(po, tmp)
    dev_new <- abs(x - xx)
    if(dev_new < dev){
      yy <- tmp
    }
    dev <- dev_new
  }
  poten <- rbind(poten, po)
  point_perturbed_4 <- rbind(point_perturbed_4, c(t, xx, yy))
}
potential <- cbind(potential, poten)
rownames(potential) <- as.character(potential[, 1])
potential <- potential[, -1]
colnames(potential) <- (as.integer(colnames(potential)) - 1) * dx
point_perturbed_4 <- as.data.frame(point_perturbed_4)
colnames(point_perturbed_4) <- c("t", "x", "y")
################################################################################
df <- c()
for(j in seq_len(ncol(potential))){
  for(i in seq_len(nrow(potential))){
    df <- rbind(df, c(rownames(potential)[i], colnames(potential)[j], potential[i, j]))
  }
  df <- rbind(df, "\n")
}
df <- as.data.frame(df)
colnames(df)[1] <- "#V1"
dirname <- "data/potential"
dir.create(dirname)
filename <- sprintf("%s/poten_%s_%s.txt", dirname, gene, type)
write.table(df, file = filename,
            sep = "\t", row.names = FALSE, quote = FALSE)
dg <- point
colnames(dg)[1] <- "#t"
filename <- sprintf("%s/point_%s_%s.txt", dirname, gene, type)
write.table(dg, file = filename,
            sep = "\t", row.names = FALSE, quote = FALSE)
dg <- point_perturbed
colnames(dg)[1] <- "#t"
filename <- sprintf("%s/point_perturbed_%s_%s.txt", dirname, gene, type)
write.table(dg, file = filename,
            sep = "\t", row.names = FALSE, quote = FALSE)
dg <- point_perturbed_2
colnames(dg)[1] <- "#t"
filename <- sprintf("%s/point_perturbed_2_%s_%s.txt", dirname, gene, type)
write.table(dg, file = filename,
            sep = "\t", row.names = FALSE, quote = FALSE)
dg <- point_perturbed_3
colnames(dg)[1] <- "#t"
filename <- sprintf("%s/point_perturbed_3_%s_%s.txt", dirname, gene, type)
write.table(dg, file = filename,
            sep = "\t", row.names = FALSE, quote = FALSE)
dg <- point_perturbed_4
colnames(dg)[1] <- "#t"
filename <- sprintf("%s/point_perturbed_4_%s_%s.txt", dirname, gene, type)
write.table(dg, file = filename,
            sep = "\t", row.names = FALSE, quote = FALSE)

#Plot
dirname <- sprintf("fig_potential_%s_C4", gene)
dir.create(dirname)
for(i in seq_len(nrow(potential))){
  df <- data.frame(as.numeric(colnames(potential)), as.numeric(potential[i, ]))
  dg <- data.frame(point[i, 2], point[i, 3])
  dh <- data.frame(point_perturbed[i, 2], point_perturbed[i, 3])
  di <- data.frame(point_perturbed_2[i, 2], point_perturbed_2[i, 3])
  dj <- data.frame(point_perturbed_3[i, 2], point_perturbed_3[i, 3])
  dk <- data.frame(point_perturbed_4[i, 2], point_perturbed_4[i, 3])
  t1 = sprintf("t = %3.2f", t0 + dt * (i - 1))
  p <- ggplot() +
    geom_line(aes(x = df[, 1], y = df[, 2]), color = "black", size = 1, alpha = 1) +
    geom_point(aes(x = dk[, 1], y = dk[, 2]), shape = 21, fill = "green",
               color = "black", size = 5, alpha = 1, stroke = 1.5,) +
    geom_point(aes(x = di[, 1], y = di[, 2]), shape = 21, fill = "yellow",
               color = "black", size = 5, alpha = 1, stroke = 1.5) +
    geom_point(aes(x = dj[, 1], y = dj[, 2]), shape = 21, fill = "green3",
               color = "black", size = 5, alpha = 1, stroke = 1.5) +
    geom_point(aes(x = dh[, 1], y = dh[, 2]), shape = 21, fill = "orange",
               color = "black", size = 5, alpha = 1, stroke = 1.5) +
    geom_point(aes(x = dg[, 1], y = dg[, 2]), shape = 21, fill = "red",
               color = "black", size = 5, alpha = 1, stroke = 1.5) +
#    ylim(-40, 80) +
    labs(title = paste0(gene, " in ", type), x = "Expression", y = "Potential") +
    annotate("text", x = -Inf, y = Inf, label = t1, hjust = -0.35,
             vjust = 1.5, size = 6, colour = "black") +
    theme_classic(base_size = 20) +
    theme(plot.title = element_text(hjust = 0.5, size = 20)) +
    guides(colour = guide_legend(override.aes = list(size = 4)))
  filename <- sprintf("%s/%05d.png", dirname, i)
  ggsave(file = filename, plot = p, dpi = 100, width = 5, height = 4)
}



# Potential computation for C5
gene <- "MAPK14"
type <- "fromC5toC5woC1C2C3C4"
f <- sprintf("gnuplot_output_%s/%s.txt", type, gene)
param <- read.table(file = f, sep = " ", header = F)
colnames(param) <- c("i", "t0", "x0", "a", "b", "c", "d", "m", "M", "n", "N", "FIT_STDFIT")
#
# For dxdt
#
f <- function(t, x, a, b, c, d, k, K, l, L, m, M, n, N){
  dxdt <- a * x**k / (x**k + K) * t**m / (t**m + M) - b * x**l / (x**l + L) * t**n / (t**n + N) + c - d * x
  return(dxdt)
}
param <- readRDS("backup/01_012_param_bifurcation_reinterpret.rds")[[type]]
param <- param[which(param$gene == gene), ]
t0 <- param$t0
x0 <- param$x0
a <- param$a
b <- param$b
c <- param$c
d <- param$d
k <- param$k
K <- param$K
l <- param$l
L <- param$L
m <- param$m
M <- param$M
n <- param$n
N <- param$N

tstep <- 400
dt <- 0.005
dx <- 0.01
potential <- data.frame(t = c(t0, t0 + dt * seq_len(tstep)))
poten <- c()
point <- c()
for(i in seq_len(nrow(potential))){
  t <- potential[i, 1]
  po <- c()
  if(i == 1){
    xx <- x0
  }else{
    xx <- xx + dt * f(t, xx, a, b, c, d, k, K, l, L, m, M, n, N)
  }
  dev <- 99
  for(j in 0:tstep){
    x <- dx * j
    fx <- x**(k + 1) * hypergeo(1, 1 + (1./k), 2 + (1./k), - x**k / K) / (k + 1) / K
    ft <- t**m / (t**m + M)
    gx <- x**(l + 1) * hypergeo(1, 1 + (1./l), 2 + (1./l), - x**l / L) / (l + 1) / L
    gt <- t**n / (t**n + N)
    tmp <- -to_real(a * fx * ft - b * gx * gt + c * x - 0.5 * d * x**2)[1]
    po <- c(po, tmp)
    dev_new <- abs(x - xx)
    if(dev_new < dev){
      yy <- tmp
    }
    dev <- dev_new
  }
  poten <- rbind(poten, po)
  point <- rbind(point, c(t, xx, yy))
}
potential <- cbind(potential, poten)
rownames(potential) <- as.character(potential[, 1])
potential <- potential[, -1]
colnames(potential) <- (as.integer(colnames(potential)) - 1) * dx
point <- as.data.frame(point)
colnames(point) <- c("t", "x", "y")
################################################################################
# Perturbation experiment
################################################################################
potential <- data.frame(t = c(t0, t0 + dt * seq_len(tstep)))
poten <- c()
point_perturbed <- c()
for(i in seq_len(nrow(potential))){
  t <- potential[i, 1]
  po <- c()
  if(i == 1){
    xx <- 0.8 * x0
  }else{
    xx <- xx + dt * f(t, xx, a, b, c, d, k, K, l, L, m, M, n, N)
  }
  dev <- 99
  for(j in 0:tstep){
    x <- dx * j
    fx <- x**(k + 1) * hypergeo(1, 1 + (1./k), 2 + (1./k), - x**k / K) / (k + 1) / K
    ft <- t**m / (t**m + M)
    gx <- x**(l + 1) * hypergeo(1, 1 + (1./l), 2 + (1./l), - x**l / L) / (l + 1) / L
    gt <- t**n / (t**n + N)
    tmp <- -to_real(a * fx * ft - b * gx * gt + c * x - 0.5 * d * x**2)[1]
    po <- c(po, tmp)
    dev_new <- abs(x - xx)
    if(dev_new < dev){
      yy <- tmp
    }
    dev <- dev_new
  }
  poten <- rbind(poten, po)
  point_perturbed <- rbind(point_perturbed, c(t, xx, yy))
}
potential <- cbind(potential, poten)
rownames(potential) <- as.character(potential[, 1])
potential <- potential[, -1]
colnames(potential) <- (as.integer(colnames(potential)) - 1) * dx
point_perturbed <- as.data.frame(point_perturbed)
colnames(point_perturbed) <- c("t", "x", "y")
################################################################################
# Perturbation experiment 2
################################################################################
potential <- data.frame(t = c(t0, t0 + dt * seq_len(tstep)))
poten <- c()
point_perturbed_2 <- c()
for(i in seq_len(nrow(potential))){
  t <- potential[i, 1]
  po <- c()
  if(i == 1){
    xx <- 0.6 * x0
  }else{
    xx <- xx + dt * f(t, xx, a, b, c, d, k, K, l, L, m, M, n, N)
  }
  dev <- 99
  for(j in 0:tstep){
    x <- dx * j
    fx <- x**(k + 1) * hypergeo(1, 1 + (1./k), 2 + (1./k), - x**k / K) / (k + 1) / K
    ft <- t**m / (t**m + M)
    gx <- x**(l + 1) * hypergeo(1, 1 + (1./l), 2 + (1./l), - x**l / L) / (l + 1) / L
    gt <- t**n / (t**n + N)
    tmp <- -to_real(a * fx * ft - b * gx * gt + c * x - 0.5 * d * x**2)[1]
    po <- c(po, tmp)
    dev_new <- abs(x - xx)
    if(dev_new < dev){
      yy <- tmp
    }
    dev <- dev_new
  }
  poten <- rbind(poten, po)
  point_perturbed_2 <- rbind(point_perturbed_2, c(t, xx, yy))
}
potential <- cbind(potential, poten)
rownames(potential) <- as.character(potential[, 1])
potential <- potential[, -1]
colnames(potential) <- (as.integer(colnames(potential)) - 1) * dx
point_perturbed_2 <- as.data.frame(point_perturbed_2)
colnames(point_perturbed_2) <- c("t", "x", "y")
################################################################################
################################################################################
# Perturbation experiment 3
################################################################################
potential <- data.frame(t = c(t0, t0 + dt * seq_len(tstep)))
poten <- c()
point_perturbed_3 <- c()
for(i in seq_len(nrow(potential))){
  t <- potential[i, 1]
  po <- c()
  if(i == 1){
    xx <- 1.2 * x0
  }else{
    xx <- xx + dt * f(t, xx, a, b, c, d, k, K, l, L, m, M, n, N)
  }
  dev <- 99
  for(j in 0:tstep){
    x <- dx * j
    fx <- x**(k + 1) * hypergeo(1, 1 + (1./k), 2 + (1./k), - x**k / K) / (k + 1) / K
    ft <- t**m / (t**m + M)
    gx <- x**(l + 1) * hypergeo(1, 1 + (1./l), 2 + (1./l), - x**l / L) / (l + 1) / L
    gt <- t**n / (t**n + N)
    tmp <- -to_real(a * fx * ft - b * gx * gt + c * x - 0.5 * d * x**2)[1]
    po <- c(po, tmp)
    dev_new <- abs(x - xx)
    if(dev_new < dev){
      yy <- tmp
    }
    dev <- dev_new
  }
  poten <- rbind(poten, po)
  point_perturbed_3 <- rbind(point_perturbed_3, c(t, xx, yy))
}
potential <- cbind(potential, poten)
rownames(potential) <- as.character(potential[, 1])
potential <- potential[, -1]
colnames(potential) <- (as.integer(colnames(potential)) - 1) * dx
point_perturbed_3 <- as.data.frame(point_perturbed_3)
colnames(point_perturbed_3) <- c("t", "x", "y")
################################################################################
# Perturbation experiment 4
################################################################################
potential <- data.frame(t = c(t0, t0 + dt * seq_len(tstep)))
poten <- c()
point_perturbed_4 <- c()
for(i in seq_len(nrow(potential))){
  t <- potential[i, 1]
  po <- c()
  if(i == 1){
    xx <- 1.4 * x0
  }else{
    xx <- xx + dt * f(t, xx, a, b, c, d, k, K, l, L, m, M, n, N)
  }
  dev <- 99
  for(j in 0:tstep){
    x <- dx * j
    fx <- x**(k + 1) * hypergeo(1, 1 + (1./k), 2 + (1./k), - x**k / K) / (k + 1) / K
    ft <- t**m / (t**m + M)
    gx <- x**(l + 1) * hypergeo(1, 1 + (1./l), 2 + (1./l), - x**l / L) / (l + 1) / L
    gt <- t**n / (t**n + N)
    tmp <- -to_real(a * fx * ft - b * gx * gt + c * x - 0.5 * d * x**2)[1]
    po <- c(po, tmp)
    dev_new <- abs(x - xx)
    if(dev_new < dev){
      yy <- tmp
    }
    dev <- dev_new
  }
  poten <- rbind(poten, po)
  point_perturbed_4 <- rbind(point_perturbed_4, c(t, xx, yy))
}
potential <- cbind(potential, poten)
rownames(potential) <- as.character(potential[, 1])
potential <- potential[, -1]
colnames(potential) <- (as.integer(colnames(potential)) - 1) * dx
point_perturbed_4 <- as.data.frame(point_perturbed_4)
colnames(point_perturbed_4) <- c("t", "x", "y")
################################################################################
df <- c()
for(j in seq_len(ncol(potential))){
  for(i in seq_len(nrow(potential))){
    df <- rbind(df, c(rownames(potential)[i], colnames(potential)[j], potential[i, j]))
  }
  df <- rbind(df, "\n")
}
df <- as.data.frame(df)
colnames(df)[1] <- "#V1"
dirname <- "data/potential"
dir.create(dirname)
filename <- sprintf("%s/poten_%s_%s.txt", dirname, gene, type)
write.table(df, file = filename,
            sep = "\t", row.names = FALSE, quote = FALSE)
dg <- point
colnames(dg)[1] <- "#t"
filename <- sprintf("%s/point_%s_%s.txt", dirname, gene, type)
write.table(dg, file = filename,
            sep = "\t", row.names = FALSE, quote = FALSE)
dg <- point_perturbed
colnames(dg)[1] <- "#t"
filename <- sprintf("%s/point_perturbed_%s_%s.txt", dirname, gene, type)
write.table(dg, file = filename,
            sep = "\t", row.names = FALSE, quote = FALSE)
dg <- point_perturbed_2
colnames(dg)[1] <- "#t"
filename <- sprintf("%s/point_perturbed_2_%s_%s.txt", dirname, gene, type)
write.table(dg, file = filename,
            sep = "\t", row.names = FALSE, quote = FALSE)
dg <- point_perturbed_3
colnames(dg)[1] <- "#t"
filename <- sprintf("%s/point_perturbed_3_%s_%s.txt", dirname, gene, type)
write.table(dg, file = filename,
            sep = "\t", row.names = FALSE, quote = FALSE)
dg <- point_perturbed_4
colnames(dg)[1] <- "#t"
filename <- sprintf("%s/point_perturbed_4_%s_%s.txt", dirname, gene, type)
write.table(dg, file = filename,
            sep = "\t", row.names = FALSE, quote = FALSE)

#Plot
dirname <- sprintf("fig_potential_%s_C5", gene)
dir.create(dirname)
for(i in seq_len(nrow(potential))){
  df <- data.frame(as.numeric(colnames(potential)), as.numeric(potential[i, ]))
  dg <- data.frame(point[i, 2], point[i, 3])
  dh <- data.frame(point_perturbed[i, 2], point_perturbed[i, 3])
  di <- data.frame(point_perturbed_2[i, 2], point_perturbed_2[i, 3])
  dj <- data.frame(point_perturbed_3[i, 2], point_perturbed_3[i, 3])
  dk <- data.frame(point_perturbed_4[i, 2], point_perturbed_4[i, 3])
  t1 = sprintf("t = %3.2f", t0 + dt * (i - 1))
  p <- ggplot() +
    geom_line(aes(x = df[, 1], y = df[, 2]), color = "black", size = 1, alpha = 1) +
    geom_point(aes(x = dk[, 1], y = dk[, 2]), shape = 21, fill = "green",
               color = "black", size = 5, alpha = 1, stroke = 1.5,) +
    geom_point(aes(x = di[, 1], y = di[, 2]), shape = 21, fill = "yellow",
               color = "black", size = 5, alpha = 1, stroke = 1.5) +
    geom_point(aes(x = dj[, 1], y = dj[, 2]), shape = 21, fill = "green3",
               color = "black", size = 5, alpha = 1, stroke = 1.5) +
    geom_point(aes(x = dh[, 1], y = dh[, 2]), shape = 21, fill = "orange",
               color = "black", size = 5, alpha = 1, stroke = 1.5) +
    geom_point(aes(x = dg[, 1], y = dg[, 2]), shape = 21, fill = "red",
               color = "black", size = 5, alpha = 1, stroke = 1.5) +
#    ylim(-40, 80) +
    labs(title = paste0(gene, " in ", type), x = "Expression", y = "Potential") +
    annotate("text", x = -Inf, y = Inf, label = t1, hjust = -0.35,
             vjust = 1.5, size = 6, colour = "black") +
    theme_classic(base_size = 20) +
    theme(plot.title = element_text(hjust = 0.5, size = 20)) +
    guides(colour = guide_legend(override.aes = list(size = 4)))
  filename <- sprintf("%s/%05d.png", dirname, i)
  ggsave(file = filename, plot = p, dpi = 100, width = 5, height = 4)
}



<br>

# Focus on control cells for defining treatment-naive and resistant cells
mcf7 <- readRDS("backup/01_002_mcf7_dataqc.rds")
ref <- readRDS("backup/01_005_mcf7_seurat.rds")
#-------------------------------------------------------------------------------
# RECODE
#-------------------------------------------------------------------------------
mat <- as.matrix(as.matrix(assay(mcf7, "counts")))
mat_recode <- RECODE::RECODE(mat)
mcf7 <- Seurat::CreateSeuratObject(counts = mat_recode,
                                   meta.data = ref[[]],
                                   project = "MCF7")
#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------
# Load the data.
#bayout <- readRDS("backup/01_003_mcf7_baynorm.rds")
#ref <- readRDS("backup/01_005_mcf7_seurat.rds")
# Create a Seurat object.
#mat <- do.call("cbind", bayout[["Bay_out_list"]])
#mcf7 <- Seurat::CreateSeuratObject(counts = mat,
#                                   meta.data = ref[[]],
#                                   project = "MCF7")

mcf7 <- mcf7[, which(mcf7$type == "Control")]
# Normalization
mcf7 <- Seurat::NormalizeData(mcf7, normalization.method = "LogNormalize")
# Variance stabilizing transform
# n <- 1500 # round(1.5 * ncol(mcf7)) <- good?
n <- 800
mcf7 <- Seurat::FindVariableFeatures(mcf7, selection.method = "vst",
                                     nfeatures = n)
# Scale data
mcf7 <- Seurat::ScaleData(mcf7)
# Principal component analysis
mcf7 <- Seurat::RunPCA(mcf7, features = Seurat::VariableFeatures(mcf7))

pc <- which(cumsum(mcf7@reductions[["pca"]]@stdev) /
        sum(mcf7@reductions[["pca"]]@stdev) > 0.5)[1]
#        sum(mcf7@reductions[["pca"]]@stdev) > 0.4)[1] <- good?
# Create k-nearest neighbor graph.
mcf7 <- Seurat::FindNeighbors(mcf7, reduction = "pca", dim = seq_len(pc))
# Cluster cells.
mcf7 <- Seurat::FindClusters(mcf7, resolution = 0.5)
# Run UMAP.
mcf7 <- Seurat::RunUMAP(mcf7, dims = seq_len(pc))
DimPlot(mcf7)

## Cell cycle inference (Control)
#Assign each cell a cell cycle score using `CellCycleScoring()`.
s.genes <- Seurat::cc.genes$s.genes
g2m.genes <- Seurat::cc.genes$g2m.genes
mcf7 <- Seurat::CellCycleScoring(mcf7, s.features = Seurat::cc.genes$s.genes,
                                 g2m.features = Seurat::cc.genes$g2m.genes)

mcf7 <- SetIdent(mcf7, value = "seurat_clusters")
mcf7@misc[["FindAllMarkers"]] <- FindAllMarkers(mcf7, only.pos = TRUE,
                                                min.pct = 0.25,
                                                logfc.threshold = 0.25)
View(mcf7@misc[["FindAllMarkers"]])

mcf7 <- SetIdent(mcf7, value = "seurat_clusters")
RidgePlot(mcf7, features = "RPS6KB1")
RidgePlot(mcf7, features = "MAP3K13")
RidgePlot(mcf7, features = "IGFBP5")
RidgePlot(mcf7, features = "EGR1")
RidgePlot(mcf7, features = "CCND1")

#-------------------------------------------------------------------------------
# GO enrichment
#-------------------------------------------------------------------------------
padj_cutoff = 0.001
n_groups <- length(unique(mcf7@meta.data[["seurat_clusters"]]))
tmp <- mcf7@misc[["FindAllMarkers"]]
cluster_names <- as.character(unique(tmp$cluster))
g <- list() ; glist <- list() ; label_names <- c()
for(i in seq_len(n_groups)){
  im1 <- i - 1
  df <- tmp[which(tmp$cluster == im1), ]
  g[[i]] <- df[which(df$p_val_adj <= padj_cutoff), ]$gene
  geneID <- clusterProfiler::bitr(g[[i]], fromType = "SYMBOL",
                                  toType = "ENTREZID",
                                  OrgDb = org.Hs.eg.db::org.Hs.eg.db)$ENTREZID
  glist[[i]] <- geneID
  label_names <- c(label_names, paste("Group_", im1, sep = ""))
}
names(glist) <- label_names

mcf7@misc[["compareCluster_GO"]] <- clusterProfiler::compareCluster(
  glist, fun = "enrichGO", OrgDb = org.Hs.eg.db::org.Hs.eg.db, ont = "BP",
  pAdjustMethod = "BH", pvalueCutoff = padj_cutoff)

p <- enrichplot::dotplot(mcf7@misc[["compareCluster_GO"]], showCategory = 5) +
  ggplot2::theme(panel.grid.major = ggplot2::element_line(size = 0.5,
                                                          color = "grey85"),
                 panel.border = ggplot2::element_rect(color = "black", fill = NA,
                                                      size = 1.5))
p
filename <- "figures/figure_02_0005b.png"
ggsave(file = filename, plot = p, dpi = 300, width = 6.8, height = 4)
#ggsave(file = filename, plot = p, dpi = 300, width = 6.8, height = 5)
--------------------------------------------------------------------------------

# Change the cluster name
mcf7$seurat_clusters <- as.integer(as.character(mcf7$seurat_clusters))
mcf7$seurat_clusters[which(mcf7$seurat_clusters == 0)] <- "C1-R"
mcf7$seurat_clusters[which(mcf7$seurat_clusters == 1)] <- "C1-S"
mcf7$seurat_clusters[which(mcf7$seurat_clusters == 2)] <- "C1-G"
mcf7$seurat_clusters[which(mcf7$seurat_clusters == 3)] <- "C1-U"
mcf7$seurat_clusters <- factor(mcf7$seurat_clusters, levels = c("C1-S", "C1-R", "C1-G", "C1-U"))

df <- data.frame(measure = mcf7$nReads, class = mcf7$seurat_clusters)
#p <- ggplot() + geom_boxplot(aes(x = df$class, y = df$measure)) +
#  labs(title = "", x = "", y = "Number of reads", color = "") +
#  theme_classic(base_size=24, base_family="Helvetica") +
#  theme(plot.title=element_text(hjust=0.5, size=24))
p <- ggplot() + geom_boxplot(aes(x = df$class, y = df$measure, fill = df$class), linewidth = 1.5) +
  ggplot2::scale_y_log10(
    breaks = 10^(3:6),
    labels = scales::trans_format("log10", scales::math_format(10^.x))
  ) +
  labs(title = "", x = "", y = "Number of reads", fill = "") +
  theme_classic(base_size=24, base_family="Helvetica") +
  theme(plot.title=element_text(hjust=0.5, size=24)) +
  ggplot2::scale_fill_manual(breaks = unique(sort(df$class)),
                             values = c("C1-R" = "darkgreen",
                                        "C1-U" = "violet",
                                        "C1-S" = "green",
                                        "C1-G" = "orange")) +
  guides(fill = guide_legend(override.aes = list(size = 12)))
p
filename <- "figures/figure_02_0030.png"
ggsave(file = filename, plot = p, dpi = 300, width = 6, height = 4)

df <- data.frame(measure = mcf7$percMT, class = mcf7$seurat_clusters)
#p <- ggplot() + geom_boxplot(aes(x = df$class, y = df$measure)) +
#  labs(title = "", x = "", y = "Number of reads", color = "") +
#  theme_classic(base_size=24, base_family="Helvetica") +
#  theme(plot.title=element_text(hjust=0.5, size=24))
p <- ggplot() + geom_boxplot(aes(x = df$class, y = df$measure, fill = df$class), linewidth = 1.5) +
  labs(title = "", x = "", y = "Perc of MT reads", fill = "") +
  theme_classic(base_size=24, base_family="Helvetica") +
  theme(plot.title=element_text(hjust=0.5, size=24)) +
  ggplot2::scale_fill_manual(breaks = unique(sort(df$class)),
                             values = c("C1-R" = "darkgreen",
                                        "C1-U" = "violet",
                                        "C1-S" = "green",
                                        "C1-G" = "orange")) +
  guides(fill = guide_legend(override.aes = list(size = 12)))
filename <- "figures/figure_02_0031.png"
ggsave(file = filename, plot = p, dpi = 300, width = 6, height = 4)

#Save the data.
saveRDS(mcf7, file = "backup/02_001_mcf7_seurat_control.rds")

#Load the data.
#mcf7 <- readRDS("backup/02_001_mcf7_seurat_control.rds")

#Visualize the clustering results.
title <- "MCF-7"
df <- as.data.frame(Seurat::Embeddings(object = mcf7, reduction = "umap"))

## Seurat clusters
df$label <- mcf7$seurat_clusters
p <- ggplot() +
  geom_point(aes(x = df[, 1], y = df[, 2], color = df$label),
             size = 2, alpha = 1) +
  labs(title = title, x = "UMAP_1", y = "UMAP_2", color = "Seurat class") +
  theme_classic(base_size = 20) +
  theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  guides(colour = guide_legend(override.aes = list(size = 4))) +
  ggplot2::scale_color_manual(breaks = unique(sort(df$label)),
                              values = c("C1-R" = "darkgreen",
                                         "C1-U" = "violet",
                                         "C1-S" = "green",
                                         "C1-G" = "orange"))
filename <- "figures/figure_02_0011.png"
ggsave(file = filename, plot = p, dpi = 300, width = 6.7, height = 5)

## QC metrics
df$ncounts <- mcf7$nCount_RNA
p <- ggplot() +
  geom_point(aes(x = df[, 1], y = df[, 2], color = df$ncounts),
             size = 2, alpha = 1) +
  labs(title = title, x = "UMAP_1", y = "UMAP_2", color = "nCount_RNA") +
  scale_color_continuous() +
  theme_classic(base_size = 20) +
  theme(plot.title = element_text(hjust = 0.5, size = 20))
filename <- "figures/figure_02_0012.png"
ggsave(file = filename, plot = p, dpi = 100, width = 7, height = 5)

df$ncounts <- mcf7$percMT
p <- ggplot() +
  geom_point(aes(x = df[, 1], y = df[, 2], color = df$ncounts),
             size = 2, alpha = 1) +
  labs(title = title, x = "UMAP_1", y = "UMAP_2", color = "percMT") +
  scale_color_continuous() +
  theme_classic(base_size = 20) +
  theme(plot.title = element_text(hjust = 0.5, size = 20))
filename <- "figures/figure_02_0013.png"
ggsave(file = filename, plot = p, dpi = 100, width = 6.2, height = 5)

## Cell cycle phase
df$label <- mcf7$Phase
p <- ggplot() +
  geom_point(aes(x = df[, 1], y = df[, 2],
                 color = factor(df$label, levels = c("G1", "S", "G2M"))),
             size = 2, alpha = 1) +
  labs(title = title, x = "UMAP_1", y = "UMAP_2", color = "Phase") +
  theme_classic(base_size = 20) +
  theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  guides(colour = guide_legend(override.aes = list(size = 4))) +
  scale_color_manual(values = c("G1" = rainbow(3)[3], "S" = "gold",
                                "G2M" = rainbow(3)[1]))
filename <- "figures/figure_02_0014.png"
ggsave(file = filename, plot = p, dpi = 300, width = 6, height = 5)

## Pseudotime
title <- "MCF-7"
merlot <- readRDS("robustness_abcdConstraint_C1/merlot_data/01_004_mcf7_merlot_C3_Nyk60.rds")
df <- data.frame(pt = mcf7$merlot_pseudotime / max(merlot[["Pseudotimes_highdim"]][["Times_cells"]]),
                 class = mcf7$seurat_clusters)
p <- ggplot() + geom_boxplot(aes(x = df$class, y = df$pt, fill = df$class), linewidth = 1.5) +
  labs(title = "", x = "", y = "Pseudotime", fill = "") +
  theme_classic(base_size=24, base_family="Helvetica") +
  theme(plot.title=element_text(hjust=0.5, size=24)) +
  ggplot2::scale_fill_manual(breaks = unique(sort(df$class)),
                             values = c("C1-R" = "darkgreen",
                                        "C1-U" = "violet",
                                        "C1-S" = "green",
                                        "C1-G" = "orange")) +
  guides(fill = guide_legend(override.aes = list(size = 12)))
p
filename <- "figures/figure_02_0015.png"
ggsave(file = filename, plot = p, dpi = 300, width = 6.2, height = 4)
## Mann-Whitney U test
wilcox.test(df[which(df$class == "C1-S"), ]$pt, df[which(df$class == "C1-R"), ]$pt)
# p-value < 2.2e-16

gene <- "IGFBP5"
df <- data.frame(x = mcf7$seurat_clusters, y = as.matrix(mcf7@assays[["RNA"]]@data)[gene, ])
#p <- FeaturePlot(mcf7, features = gene, reduction = "umap",
#                 pt.size = 1, min.cutoff = "q20", max.cutoff = "q90")
p <- ggplot() + geom_boxplot(aes(x = df$x, y = df$y, fill = df$x), linewidth = 1.5) +
  labs(title = "", x = "", y = gene, fill = "") +
  theme_classic(base_size=24, base_family="Helvetica") +
  theme(plot.title=element_text(hjust=0.5, size=24)) +
  ggplot2::scale_fill_manual(breaks = unique(sort(df$x)),
                             values = c("C1-R" = "darkgreen",
                                        "C1-U" = "violet",
                                        "C1-S" = "green",
                                        "C1-G" = "orange")) +
  guides(fill = guide_legend(override.aes = list(size = 12)))
p
filename <- sprintf("figures/figure_02_0020_%s.png", gene)
ggsave(file = filename, plot = p, dpi = 300, width = 7, height = 4)

gene <- "CCND1"
df <- data.frame(x = mcf7$seurat_clusters, y = as.matrix(mcf7@assays[["RNA"]]@data)[gene, ])
#p <- FeaturePlot(mcf7, features = gene, reduction = "umap",
#                 pt.size = 1, min.cutoff = "q20", max.cutoff = "q90")
p <- ggplot() + geom_boxplot(aes(x = df$x, y = df$y, fill = df$x), linewidth = 1.5) +
  labs(title = "", x = "", y = gene, fill = "") +
  theme_classic(base_size=24, base_family="Helvetica") +
  theme(plot.title=element_text(hjust=0.5, size=24)) +
  ggplot2::scale_fill_manual(breaks = unique(sort(df$x)),
                             values = c("C1-R" = "darkgreen",
                                        "C1-U" = "violet",
                                        "C1-S" = "green",
                                        "C1-G" = "orange")) +
  guides(fill = guide_legend(override.aes = list(size = 12)))
p
filename <- sprintf("figures/figure_02_0020_%s.png", gene)
ggsave(file = filename, plot = p, dpi = 300, width = 7, height = 4)

gene <- "FOS"
df <- data.frame(x = mcf7$seurat_clusters, y = as.matrix(mcf7@assays[["RNA"]]@data)[gene, ])
#p <- FeaturePlot(mcf7, features = gene, reduction = "umap",
#                 pt.size = 1, min.cutoff = "q20", max.cutoff = "q90")
p <- ggplot() + geom_boxplot(aes(x = df$x, y = df$y, fill = df$x), linewidth = 1.5) +
  labs(title = "", x = "", y = gene, fill = "") +
  theme_classic(base_size=24, base_family="Helvetica") +
  theme(plot.title=element_text(hjust=0.5, size=24)) +
  ggplot2::scale_fill_manual(breaks = unique(sort(df$x)),
                             values = c("C1-R" = "darkgreen",
                                        "C1-U" = "violet",
                                        "C1-S" = "green",
                                        "C1-G" = "orange")) +
  guides(fill = guide_legend(override.aes = list(size = 12)))
p
filename <- sprintf("figures/figure_02_0020_%s.png", gene)
ggsave(file = filename, plot = p, dpi = 300, width = 7, height = 4)

gene <- "JUN"
df <- data.frame(x = mcf7$seurat_clusters, y = as.matrix(mcf7@assays[["RNA"]]@data)[gene, ])
#p <- FeaturePlot(mcf7, features = gene, reduction = "umap",
#                 pt.size = 1, min.cutoff = "q20", max.cutoff = "q90")
p <- ggplot() + geom_boxplot(aes(x = df$x, y = df$y, fill = df$x), linewidth = 1.5) +
  labs(title = "", x = "", y = gene, fill = "") +
  theme_classic(base_size=24, base_family="Helvetica") +
  theme(plot.title=element_text(hjust=0.5, size=24)) +
  ggplot2::scale_fill_manual(breaks = unique(sort(df$x)),
                             values = c("C1-R" = "darkgreen",
                                        "C1-U" = "violet",
                                        "C1-S" = "green",
                                        "C1-G" = "orange")) +
  guides(fill = guide_legend(override.aes = list(size = 12)))
p
filename <- sprintf("figures/figure_02_0020_%s.png", gene)
ggsave(file = filename, plot = p, dpi = 300, width = 7, height = 4)

gene <- "CCNA2"
df <- data.frame(x = mcf7$seurat_clusters, y = as.matrix(mcf7@assays[["RNA"]]@data)[gene, ])
#p <- FeaturePlot(mcf7, features = gene, reduction = "umap",
#                 pt.size = 1, min.cutoff = "q20", max.cutoff = "q90")
p <- ggplot() + geom_boxplot(aes(x = df$x, y = df$y, fill = df$x), linewidth = 1.5) +
  labs(title = "", x = "", y = gene, fill = "") +
  theme_classic(base_size=24, base_family="Helvetica") +
  theme(plot.title=element_text(hjust=0.5, size=24)) +
  ggplot2::scale_fill_manual(breaks = unique(sort(df$x)),
                             values = c("C1-R" = "darkgreen",
                                        "C1-U" = "violet",
                                        "C1-S" = "green",
                                        "C1-G" = "orange")) +
  guides(fill = guide_legend(override.aes = list(size = 12)))
p
filename <- sprintf("figures/figure_02_0020_%s.png", gene)
ggsave(file = filename, plot = p, dpi = 300, width = 7, height = 4)

gene <- "PRDX1"
df <- data.frame(x = mcf7$seurat_clusters, y = as.matrix(mcf7@assays[["RNA"]]@data)[gene, ])
#p <- FeaturePlot(mcf7, features = gene, reduction = "umap",
#                 pt.size = 1, min.cutoff = "q20", max.cutoff = "q90")
p <- ggplot() + geom_boxplot(aes(x = df$x, y = df$y, fill = df$x), linewidth = 1.5) +
  labs(title = "", x = "", y = gene, fill = "") +
  theme_classic(base_size=24, base_family="Helvetica") +
  theme(plot.title=element_text(hjust=0.5, size=24)) +
  ggplot2::scale_fill_manual(breaks = unique(sort(df$x)),
                             values = c("C1-R" = "darkgreen",
                                        "C1-U" = "violet",
                                        "C1-S" = "green",
                                        "C1-G" = "orange")) +
  guides(fill = guide_legend(override.aes = list(size = 12)))
p
filename <- sprintf("figures/figure_02_0020_%s.png", gene)
ggsave(file = filename, plot = p, dpi = 300, width = 7, height = 4)

gene <- "MAPK14"
df <- data.frame(x = mcf7$seurat_clusters, y = as.matrix(mcf7@assays[["RNA"]]@data)[gene, ])
#p <- FeaturePlot(mcf7, features = gene, reduction = "umap",
#                 pt.size = 1, min.cutoff = "q20", max.cutoff = "q90")
p <- ggplot() + geom_boxplot(aes(x = df$x, y = df$y, fill = df$x), linewidth = 1.5) +
  labs(title = "", x = "", y = gene, fill = "") +
  theme_classic(base_size=24, base_family="Helvetica") +
  theme(plot.title=element_text(hjust=0.5, size=24)) +
  ggplot2::scale_fill_manual(breaks = unique(sort(df$x)),
                             values = c("C1-R" = "darkgreen",
                                        "C1-U" = "violet",
                                        "C1-S" = "green",
                                        "C1-G" = "orange")) +
  guides(fill = guide_legend(override.aes = list(size = 12)))
p
filename <- sprintf("figures/figure_02_0020_%s.png", gene)
ggsave(file = filename, plot = p, dpi = 300, width = 7, height = 4)

gene <- "MAP3K13"
df <- data.frame(x = mcf7$seurat_clusters, y = as.matrix(mcf7@assays[["RNA"]]@data)[gene, ])
#p <- FeaturePlot(mcf7, features = gene, reduction = "umap",
#                 pt.size = 1, min.cutoff = "q20", max.cutoff = "q90")
p <- ggplot() + geom_boxplot(aes(x = df$x, y = df$y, fill = df$x), linewidth = 1.5) +
  labs(title = "", x = "", y = gene, fill = "") +
  theme_classic(base_size=24, base_family="Helvetica") +
  theme(plot.title=element_text(hjust=0.5, size=24)) +
  ggplot2::scale_fill_manual(breaks = unique(sort(df$x)),
                             values = c("C1-R" = "darkgreen",
                                        "C1-U" = "violet",
                                        "C1-S" = "green",
                                        "C1-G" = "orange")) +
  guides(fill = guide_legend(override.aes = list(size = 12)))
p
filename <- sprintf("figures/figure_02_0020_%s.png", gene)
ggsave(file = filename, plot = p, dpi = 300, width = 7, height = 4)

gene <- "RPS6KB1"
df <- data.frame(x = mcf7$seurat_clusters, y = as.matrix(mcf7@assays[["RNA"]]@data)[gene, ])
#p <- FeaturePlot(mcf7, features = gene, reduction = "umap",
#                 pt.size = 1, min.cutoff = "q20", max.cutoff = "q90")
p <- ggplot() + geom_boxplot(aes(x = df$x, y = df$y, fill = df$x), linewidth = 1.5) +
  labs(title = "", x = "", y = gene, fill = "") +
  theme_classic(base_size=24, base_family="Helvetica") +
  theme(plot.title=element_text(hjust=0.5, size=24)) +
  ggplot2::scale_fill_manual(breaks = unique(sort(df$x)),
                             values = c("C1-R" = "darkgreen",
                                        "C1-U" = "violet",
                                        "C1-S" = "green",
                                        "C1-G" = "orange")) +
  guides(fill = guide_legend(override.aes = list(size = 12)))
p
filename <- sprintf("figures/figure_02_0020_%s.png", gene)
ggsave(file = filename, plot = p, dpi = 300, width = 7, height = 4)

gene <- "XBP1"
df <- data.frame(x = mcf7$seurat_clusters, y = as.matrix(mcf7@assays[["RNA"]]@data)[gene, ])
#p <- FeaturePlot(mcf7, features = gene, reduction = "umap",
#                 pt.size = 1, min.cutoff = "q20", max.cutoff = "q90")
p <- ggplot() + geom_boxplot(aes(x = df$x, y = df$y, fill = df$x), linewidth = 1.5) +
  labs(title = "", x = "", y = gene, fill = "") +
  theme_classic(base_size=24, base_family="Helvetica") +
  theme(plot.title=element_text(hjust=0.5, size=24)) +
  ggplot2::scale_fill_manual(breaks = unique(sort(df$x)),
                             values = c("C1-R" = "darkgreen",
                                        "C1-U" = "violet",
                                        "C1-S" = "green",
                                        "C1-G" = "orange")) +
  guides(fill = guide_legend(override.aes = list(size = 12)))
p
filename <- sprintf("figures/figure_02_0020_%s.png", gene)
ggsave(file = filename, plot = p, dpi = 300, width = 6, height = 4)

gene <- "CCNB2"
df <- data.frame(x = mcf7$seurat_clusters, y = as.matrix(mcf7@assays[["RNA"]]@data)[gene, ])
#p <- FeaturePlot(mcf7, features = gene, reduction = "umap",
#                 pt.size = 1, min.cutoff = "q20", max.cutoff = "q90")
p <- ggplot() + geom_boxplot(aes(x = df$x, y = df$y, fill = df$x), linewidth = 1.5) +
  labs(title = "", x = "", y = gene, fill = "") +
  theme_classic(base_size=24, base_family="Helvetica") +
  theme(plot.title=element_text(hjust=0.5, size=24)) +
  ggplot2::scale_fill_manual(breaks = unique(sort(df$x)),
                             values = c("C1-R" = "darkgreen",
                                        "C1-U" = "violet",
                                        "C1-S" = "green",
                                        "C1-G" = "orange")) +
  guides(fill = guide_legend(override.aes = list(size = 12)))
p
filename <- sprintf("figures/figure_02_0020_%s.png", gene)
ggsave(file = filename, plot = p, dpi = 300, width = 6.3, height = 4)

gene <- "NME1"
df <- data.frame(x = mcf7$seurat_clusters, y = as.matrix(mcf7@assays[["RNA"]]@data)[gene, ])
#p <- FeaturePlot(mcf7, features = gene, reduction = "umap",
#                 pt.size = 1, min.cutoff = "q20", max.cutoff = "q90")
p <- ggplot() + geom_boxplot(aes(x = df$x, y = df$y, fill = df$x), linewidth = 1.5) +
  labs(title = "", x = "", y = gene, fill = "") +
  theme_classic(base_size=24, base_family="Helvetica") +
  theme(plot.title=element_text(hjust=0.5, size=24)) +
  ggplot2::scale_fill_manual(breaks = unique(sort(df$x)),
                             values = c("C1-R" = "darkgreen",
                                        "C1-U" = "violet",
                                        "C1-S" = "green",
                                        "C1-G" = "orange")) +
  guides(fill = guide_legend(override.aes = list(size = 12)))
p
filename <- sprintf("figures/figure_02_0020_%s.png", gene)
ggsave(file = filename, plot = p, dpi = 300, width = 7, height = 4)

gene <- "CD63"
df <- data.frame(x = mcf7$seurat_clusters, y = as.matrix(mcf7@assays[["RNA"]]@data)[gene, ])
#p <- FeaturePlot(mcf7, features = gene, reduction = "umap",
#                 pt.size = 1, min.cutoff = "q20", max.cutoff = "q90")
p <- ggplot() + geom_boxplot(aes(x = df$x, y = df$y, fill = df$x), linewidth = 1.5) +
  labs(title = "", x = "", y = gene, fill = "") +
  theme_classic(base_size=24, base_family="Helvetica") +
  theme(plot.title=element_text(hjust=0.5, size=24)) +
  ggplot2::scale_fill_manual(breaks = unique(sort(df$x)),
                             values = c("C1-R" = "darkgreen",
                                        "C1-U" = "violet",
                                        "C1-S" = "green",
                                        "C1-G" = "orange")) +
  guides(fill = guide_legend(override.aes = list(size = 12)))
p
filename <- sprintf("figures/figure_02_0020_%s.png", gene)
ggsave(file = filename, plot = p, dpi = 300, width = 7, height = 4)

mcf7 <- SetIdent(mcf7, value = "seurat_clusters")
RidgePlot(mcf7, features = "TBL1XR1")
FeaturePlot(mcf7, features = "KRT8P33", reduction = "umap",
                 pt.size = 1, min.cutoff = "q20", max.cutoff = "q90")
VlnPlot(mcf7, features = "MAPK14")
#cormat <- apply(mat, 1, function(x){ return(cor(mat["KRT19", ], x))})



<br>

## From cluster 1 to 5 without cluster 2
mcf7 <- readRDS("backup/01_006_mcf7_reduced.rds")
ctrl <- readRDS("backup/02_001_mcf7_seurat_control.rds")
merlot <- readRDS("robustness_abcdConstraint_C1/merlot_data/01_004_mcf7_merlot_C3_Nyk60.rds")

### Chagne the cluster name
ctrl$seurat_clusters <- as.character(ctrl$seurat_clusters)
mcf7$detail_clusters <- as.character(mcf7$merlot_clusters)
for(i in seq_len(ncol(mcf7))){
  if(colnames(mcf7)[i] %in% colnames(ctrl)){
    mcf7$detail_clusters[i] <- ctrl[, colnames(mcf7)[i]]$seurat_clusters
  }
}
mcf7$detail_clusters[mcf7$detail_clusters == "C1-S"] <- 11
mcf7$detail_clusters[mcf7$detail_clusters == "C1-R"] <- 22
mcf7$detail_clusters[mcf7$detail_clusters == "C1-U"] <- 88
mcf7$detail_clusters[mcf7$detail_clusters == "C1-G"] <- 66
mcf7$detail_clusters[mcf7$detail_clusters == "2"] <- 77
mcf7$detail_clusters[mcf7$detail_clusters == "3"] <- 33
mcf7$detail_clusters[mcf7$detail_clusters == "4"] <- 44
mcf7$detail_clusters[mcf7$detail_clusters == "5"] <- 55
mcf7$detail_clusters[mcf7$detail_clusters == "11"] <- 1
mcf7$detail_clusters[mcf7$detail_clusters == "22"] <- 2
mcf7$detail_clusters[mcf7$detail_clusters == "33"] <- 3
mcf7$detail_clusters[mcf7$detail_clusters == "44"] <- 4
mcf7$detail_clusters[mcf7$detail_clusters == "55"] <- 5
mcf7$detail_clusters[mcf7$detail_clusters == "66"] <- 6
mcf7$detail_clusters[mcf7$detail_clusters == "77"] <- 7
mcf7$detail_clusters[mcf7$detail_clusters == "88"] <- 8
mcf7$detail_clusters <- as.integer(mcf7$detail_clusters)

#Save the data.
saveRDS(mcf7, file = "backup/01_020_mcf7_seurat_detail.rds")

#Load the data.
#mcf7 <- readRDS("backup/01_020_mcf7_seurat_detail.rds")

mcf7_sub <- mcf7[, which(mcf7$detail_clusters %in% c(1, 2, 3, 4, 5))]
class <- as.integer(mcf7$detail_clusters)
#barplot(rep(1, 5), col = scales::hue_pal()(5))
#barplot(rep(1, 2), col = RColorBrewer::brewer.pal(3, "Accent")[c(1, 2)])
merlot[["Pseudotimes_highdim"]][["Cells2Branches"]] <- class
merlot_sub <- merlot
merlot_sub[["Pseudotimes_highdim"]][["Cells2Branches"]] <- mcf7$merlot_clusters
mat <- merlot_sub[["GenesSpaceEmbedding"]][["CellCoords"]]
inds_cell <- which(rownames(mat) %in% colnames(mcf7_sub))
inds_gene <- which(colnames(mat) %in% rownames(mcf7_sub))

### Modify merlot object.
### GenesSpaceEmbedding
merlot_sub[["GenesSpaceEmbedding"]][["Nodes"]] <-
  merlot[["GenesSpaceEmbedding"]][["Nodes"]][, inds_gene]
merlot_sub[["GenesSpaceEmbedding"]][["Topology"]][["Endpoints"]] <- c(1, 2, 4)
merlot_sub[["GenesSpaceEmbedding"]][["CellCoords"]] <-
  merlot[["GenesSpaceEmbedding"]][["CellCoords"]][inds_cell, inds_gene]
merlot_sub[["GenesSpaceEmbedding"]][["Cells2TreeNodes"]] <-
  merlot[["GenesSpaceEmbedding"]][["Cells2TreeNodes"]][inds_cell, ]
merlot_sub[["GenesSpaceEmbedding"]][["Cells2Branches"]] <-
  merlot[["GenesSpaceEmbedding"]][["Cells2Branches"]][inds_cell]
merlot_sub[["GenesSpaceEmbedding"]][["Branches"]][[5]] <- NULL

### Pseudotimes_highdim
merlot_sub[["Pseudotimes_highdim"]][["Times_cells"]] <-
  merlot[["Pseudotimes_highdim"]][["Times_cells"]][inds_cell]
merlot_sub[["Pseudotimes_highdim"]][["Projected_Times_Cells"]] <-
  merlot[["Pseudotimes_highdim"]][["Projected_Times_Cells"]][inds_cell]
merlot_sub[["Pseudotimes_highdim"]][["Branches"]][[5]] <- NULL
merlot_sub[["Pseudotimes_highdim"]][["Cells2TreeNodes"]] <-
  merlot[["Pseudotimes_highdim"]][["Cells2TreeNodes"]][inds_cell, ]
merlot_sub[["Pseudotimes_highdim"]][["Cells2Branches"]] <-
  merlot[["Pseudotimes_highdim"]][["Cells2Branches"]][inds_cell]
merlot_sub[["Pseudotimes_highdim"]][["PseudoExpressionMatrix"]] <-
  merlot[["Pseudotimes_highdim"]][["PseudoExpressionMatrix"]][inds_cell, inds_gene]
# ----------------------------------------
# Show gene expressions as a function of pseudotime
# ----------------------------------------
source("R/myfuncs.R")
genes <- rownames(mcf7_sub)
#View(t(t(genes)))
dirname <- "fig_pt_detail"
dir.create(dirname)
# ESR1 (769), GREB1 (5284), XBP1 (865), CREB1 (1890), NFKB1 (1412), FOS (4470),
# RHOA (403), NOTCH2 (2634), MAPK9 (259), RPS6KB1 (1369),
# MAPK14 (1534), MAP3K13 (490), JUN (4819), JUND (2435), IGFBP5 (1728), EGR1 (2016)
goi <- c(769, 5284, 865, 1890, 1412, 4470, 403, 2634, 259, 1369, 1534, 490, 4819, 2435, 1728, 2016)
for(i in seq_along(goi)){
  filename <- sprintf("%s/%s.png", dirname, genes[goi[i]])
#  png(file = filename, width = 600, height = 450, res = 100)
  png(file = filename, width = 1250, height = 1200, res = 300)
  plot_pseudotime_expression_detail(
    GeneName = genes[goi[i]], merlot = merlot,
    EmbeddedTree = merlot_sub[["GenesSpaceEmbedding"]],
    Pseudotimes = merlot_sub[["Pseudotimes_highdim"]],
    branch_tags = c(), addlegend = F, range_y = "tree", switch = 1)
  dev.off()
}

# Elastic tree
source("R/plot_additional.R")
labels <- mcf7$detail_clusters
colors <- labels
merlot[["GenesSpaceEmbedding"]]$Cells2Branches <- colors
colors <- c("green", "green4", scales::hue_pal()(5)[c(4, 1, 2)], "orange", scales::hue_pal()(5)[c(5)], "violet")[colors]

ElasticTree <- merlot$ElasticTree
GeneSpaceEmbedding <- merlot[["GenesSpaceEmbedding"]]
n_groups <- length(unique(merlot[["GenesSpaceEmbedding"]]$Cells2Branches))
df <- as.data.frame(ElasticTree$CellCoords)
df$branch <- GeneSpaceEmbedding$Cells2Branches
df$label <- labels
df$color <- colors
df <- df[order(df$label), ]
dg <- as.data.frame(ElasticTree$Nodes)
dh <- as.data.frame(ElasticTree$EndpointsCoords)
di <- ElasticTree$BranchpointsCoords
if(is.null(dim(di))){
  di <- as.data.frame(t(ElasticTree$BranchpointsCoords))
}else{
  di <- as.data.frame(ElasticTree$BranchpointsCoords)
}
dj <- list()
for(j in seq_len(n_groups)){
  dj[[j]] <- as.data.frame(dg[df$branch[[j]], ])
}
mycolors <- unique(df[order(df$branch), ]$color)
filename <- "figures/figure_01_0050.png"
png(file = filename, height = 1500, width = 1500, res = 300)
plot3D::scatter3D(df[, 1], df[, 2], df[, 3], main = "MCF-7",
                  xlab = "", ylab = "", zlab = "",
                  box = FALSE, bty = "b2", axes = TRUE, nticks = 5,
                  theta = 230, phi = 30, pch = 16, cex = 1, alpha = 0.1,
                  col = mycolors[df$branch], colvar = NA, colkey = FALSE)
for(j in seq_len(n_groups)){
  plot3D::scatter3D(dj[[j]][, 1], dj[[j]][, 2], dj[[j]][, 3], 
                    lwd = 2, alpha = 1, col = "black", add = TRUE)
}
plot3D::scatter3D(dh[, 1], dh[, 2], dh[, 3], pch = 16, cex = 3, alpha = 0.25,
                  col = "black", add = TRUE)
plot3D::scatter3D(di[, 1], di[, 2], di[, 3], pch = 16, cex = 3, alpha = 0.40,
                  col = "black", add = TRUE)
plot3D::scatter3D(dg[, 1], dg[, 2], dg[, 3], pch = 16, cex = 1, alpha = 1,
                  col = "black", add = TRUE)
dev.off()



# Bar chart
library(dplyr)
Plot_bar_cellcycle <- function(obj, ymax, title, xlabel, ylabel){
  df <- obj[[]][, c("merlot_clusters", "type")]
  df$merlot_clusters <- as.integer(df$merlot_clusters)
  df$merlot_clusters[df$merlot_clusters == 2] <- 55
  df$merlot_clusters[df$merlot_clusters == 3] <- 22
  df$merlot_clusters[df$merlot_clusters == 4] <- 33
  df$merlot_clusters[df$merlot_clusters == 5] <- 44
  df$merlot_clusters[df$merlot_clusters == 22] <- 2
  df$merlot_clusters[df$merlot_clusters == 33] <- 3
  df$merlot_clusters[df$merlot_clusters == 44] <- 4
  df$merlot_clusters[df$merlot_clusters == 55] <- 5
  df <- group_by(df, type, merlot_clusters) %>% summarise (n=n())
  df <- rbind(df, data.frame(type = "Control", merlot_clusters = 2, n = 0))
  df <- rbind(df, data.frame(type = "Control", merlot_clusters = 3, n = 0))
  df <- rbind(df, data.frame(type = "TAM_3weeks", merlot_clusters = 1, n = 0))
  df <- rbind(df, data.frame(type = "TAM_6weeks", merlot_clusters = 1, n = 0))
  df <- rbind(df, data.frame(type = "TAM_9weeks", merlot_clusters = 1, n = 0))
  df$type <- factor(df$type, levels = c("Control", "TAM_3weeks", "TAM_6weeks", "TAM_9weeks"))
  df$merlot_clusters <- as.integer(as.character(df$merlot_clusters))
  pop <- c()
  for(j in 1:length(unique(df$merlot_clusters))){
    c <- df[which(df$merlot_clusters == j),] ; c$n <- c$n / sum(c$n)
    pop <- rbind(pop, c)
  }
  n_groups <- length(unique(df$type))
#  my_colors <- c("green4", "magenta", "cyan", "gold")
  my_colors <- c("grey20", "grey40", "grey60", "grey80")
  p <- ggplot(pop) +
    geom_bar(aes(x = as.factor(merlot_clusters), y = n, fill = type),
           stat="identity", position=position_dodge2(width=0.8), width=0.8) +
    theme_classic(base_size=20, base_family="Helvetica") +
    theme(plot.title=element_text(hjust=0.5)) +
    ylim(0, ymax) + labs(title = title, x = xlabel, y = ylabel) +
    scale_fill_manual(values=my_colors)
  return(p)
}
p <- Plot_bar_cellcycle(obj = mcf7, ymax = 1, title = "MCF-7",
                        xlabel = "", ylabel = "Population ratio")
filename <- "figures/figure_01_0051.png"
ggsave(file = filename, plot = p, dpi = 300, width = 7.5, height = 4)

Plot_bar_cellcycle <- function(obj, ymax, title, xlabel, ylabel){
  df <- obj[[]][, c("merlot_clusters", "Phase")]
  df$merlot_clusters <- as.integer(df$merlot_clusters)
  df$merlot_clusters <- as.integer(df$merlot_clusters)
  df$merlot_clusters[df$merlot_clusters == 2] <- 55
  df$merlot_clusters[df$merlot_clusters == 3] <- 22
  df$merlot_clusters[df$merlot_clusters == 4] <- 33
  df$merlot_clusters[df$merlot_clusters == 5] <- 44
  df$merlot_clusters[df$merlot_clusters == 22] <- 2
  df$merlot_clusters[df$merlot_clusters == 33] <- 3
  df$merlot_clusters[df$merlot_clusters == 44] <- 4
  df$merlot_clusters[df$merlot_clusters == 55] <- 5
  df <- group_by(df, Phase, merlot_clusters) %>% summarise (n=n())
  df$Phase <- factor(df$Phase, levels = c("G1", "S", "G2M"))
  df$merlot_clusters <- as.integer(as.character(df$merlot_clusters))
  pop <- c()
  for(j in 1:length(unique(df$merlot_clusters))){
    c <- df[which(df$merlot_clusters == j),] ; c$n <- c$n / sum(c$n)
    pop <- rbind(pop, c)
  }
  n_groups <- length(unique(df$Phase))
  my_colors <- c(rainbow(n_groups)[3], "gold", rainbow(n_groups)[1])
  p <- ggplot(pop) +
    geom_bar(aes(x=as.factor(merlot_clusters), y=n, fill=Phase),
           stat="identity", position=position_dodge2(width=0.7), width=0.7) +
    theme_classic(base_size=20, base_family="Helvetica") +
    theme(plot.title=element_text(hjust=0.5)) +
    ylim(0, ymax) + labs(title = title, x = xlabel, y = ylabel) +
    scale_fill_manual(values = my_colors)
  return(p)
}
p <- Plot_bar_cellcycle(obj = mcf7, ymax = 1, title = "MCF-7",
                        xlabel = "MERLoT Cluster", ylabel = "Population ratio")
filename <- "figures/figure_01_0052.png"
ggsave(file = filename, plot = p, dpi = 300, width = 6.5, height = 4)



<br>

# Validaton by analyzing scRNA-seq data
#param_ptmat_dev <- readRDS("backup/01_012_param_bifurcation_reinterpret.rds")
mcf7 <- readRDS("backup/01_002_mcf7_dataqc.rds")
#-------------------------------------------------------------------------------
# RECODE
#-------------------------------------------------------------------------------
mat <- as.matrix(as.matrix(assay(mcf7, "counts")))
mat_recode <- RECODE::RECODE(mat)
mcf7 <- Seurat::CreateSeuratObject(counts = mat_recode,
                                   meta.data = as.data.frame(colData(mcf7)),
                                   project = "MCF7")
#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------
#mcf7 <- Seurat::CreateSeuratObject(counts = as.matrix(assay(mcf7, "counts")),
#                                   meta.data = as.data.frame(colData(mcf7)),
#                                   project = "MCF7")

#Save the data.
saveRDS(mcf7, file = "backup/02_002_mcf7_prep_validaton.rds")

#Load the data.
#mcf7 <- readRDS("backup/02_002_mcf7_prep_validaton.rds")

ref <- readRDS("backup/01_020_mcf7_seurat_detail.rds")
ref$detail_clusters[ref$detail_clusters == 1] <- "C1-S"
ref$detail_clusters[ref$detail_clusters == 2] <- "C1-R"
ref$detail_clusters[ref$detail_clusters == 3] <- "C2"
ref$detail_clusters[ref$detail_clusters == 4] <- "C3"
ref$detail_clusters[ref$detail_clusters == 5] <- "C4"
ref$detail_clusters[ref$detail_clusters == 6] <- "C1-G"
ref$detail_clusters[ref$detail_clusters == 7] <- "C5"
ref$detail_clusters[ref$detail_clusters == 8] <- "C1-U"

#ref$detail_clusters <- factor(ref$detail_clusters,
#                              levels = c("C1-S", "C1-R", "C2", "C3", "C4", "C1-G", "C5", "C1-U"))
ref$detail_clusters <- factor(ref$detail_clusters,
                              levels = c("C1-U", "C5", "C1-G", "C4", "C3", "C2", "C1-R", "C1-S"))
mcf7$detail_clusters <- ref$detail_clusters
mcf7 <- Seurat::NormalizeData(mcf7, normalization.method = "LogNormalize")
mcf7_sub <- mcf7[, which(mcf7$detail_clusters %in% c("C1-S", "C1-R", "C2", "C3", "C4"))]
mcf7_sub <- mcf7[, which(mcf7$detail_clusters %in% c("C4", "C3", "C2", "C1-R", "C1-S"))]

Idents(object = mcf7_sub) <- "detail_clusters"
RidgePlot(object = mcf7_sub, features = "EGR1")
RidgePlot(object = mcf7_sub, features = "RPS6KB1")
RidgePlot(object = mcf7_sub, features = "MAP3K13")
#-------------------------------------------------------------------------------
library(ggridges)
library(ggplot2)
#genes <- c("MAP3K13", "MAPK14", "IGFBP5", "FOS", "RPS6KB1", "XBP1", "NME1", "MAPK9")
genes <- c("RPS6KB1")
genes <- c("JUND")
for(i in seq_along(genes)){
  df <- data.frame(x = as.matrix(mcf7_sub@assays[["RNA"]]@data)[genes[i], ],
                   y = mcf7_sub$detail_clusters)
  p <- ggplot(df, aes(x = x, y = y, fill = y)) +
    geom_density_ridges() +
    theme_ridges() +
#    xlim(c(NA, 1.2)) +
    xlim(c(NA, 3)) +
    theme(legend.position = "none", plot.title = element_text(hjust = 0.5, size = 20)) +
    labs(title = "", x = genes[i], y = "", fill = "Class") +
    scale_fill_manual(values = c("C1-S" = "green",
                                 "C1-R" = "green4",
                                 "C2" = scales::hue_pal()(5)[4],
                                 "C3" = scales::hue_pal()(5)[1],
                                 "C4" = scales::hue_pal()(5)[2])) +
    theme_classic(base_size = 20) +
    ggplot2::theme(plot.title = element_text(hjust = 0.5, size = 20)) +
    ggplot2::guides(colour = guide_legend(override.aes = list(size = 4)))
  filename <- sprintf("figures/figure_01_1000_%s.png", genes[i])
  ggsave(file = filename, plot = p, dpi = 300, width = 5, height = 4.3)
}
#-------------------------------------------------------------------------------

#Save the data.
saveRDS(mcf7_sub, file = "backup/02_003_mcf7_sub_validaton.rds")

#Load the data.
#mcf7_sub <- readRDS("backup/02_003_mcf7_sub_validaton.rds")



<br>

################################################################################
################################################################################
################################################################################
#
# BRCA + TNBC data
#
################################################################################
################################################################################
################################################################################
brca <- readRDS("backup/10_002_brca_xcell.rds")
brca <- Seurat::CreateSeuratObject(counts = as.matrix(assay(brca, "counts")),
                                   meta.data = as.data.frame(colData(brca)),
                                   project = "BRCA")
brca <- brca[, which(brca$subtype != "NA")]
brca <- Seurat::NormalizeData(brca, normalization.method = "LogNormalize")
brca <- SetIdent(brca, value = "subtype")
gene <- "JUN"
p <- RidgePlot(brca, features = gene)
filename <- sprintf("figures/figure_50_0010_%s.png", gene)
ggsave(file = filename, plot = p, dpi = 300, width = 6, height = 5)

res <- Seurat::FindAllMarkers(brca, only.pos = TRUE,
                              min.pct = 0.25, logfc.threshold = 0.25)
View(res)


## Perform Seurat preprocessing
# Variance stabilizing transform
n <- round(4 * ncol(brca))
brca <- Seurat::FindVariableFeatures(brca, selection.method = "vst",
                                     nfeatures = n)
# Scale data
brca <- Seurat::ScaleData(brca)
# Principal component analysis
brca <- Seurat::RunPCA(brca, features = Seurat::VariableFeatures(brca))

## Cluster cells
pc <- which(cumsum(brca@reductions[["pca"]]@stdev) /
              sum(brca@reductions[["pca"]]@stdev) > 0.9)[1]
# Create k-nearest neighbor graph.
brca <- Seurat::FindNeighbors(brca, reduction = "pca", dim = seq_len(pc))
# Cluster cells.
brca <- Seurat::FindClusters(brca, resolution = 0.4)
# Run UMAP.
brca <- Seurat::RunUMAP(brca, dims = seq_len(pc))

#Visualize the clustering results.
DimPlot(brca, group.by = "subtype")
title <- "TCGA (BRCA)"
df <- as.data.frame(Seurat::Embeddings(object = brca, reduction = "umap"))

# Conditions
df$label <- factor(brca$subtype, levels = unique(sort(brca$subtype)))
p <- ggplot2::ggplot() +
  ggplot2::geom_point(ggplot2::aes(x = df[, 1], y = df[, 2], color = df$label),
                      size = 1, alpha = 1) +
  ggplot2::labs(title = title, x = "UMAP_1", y = "UMAP_2", color = "") +
  ggplot2::theme_classic(base_size = 20) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  ggplot2::guides(colour = guide_legend(override.aes = list(size = 4)))
filename <- "figures/figure_50_0020.png"
ggplot2::ggsave(file = filename, plot = p, dpi = 200, width = 6.5, height = 5)

# Seurat clusters
df$label <- brca$seurat_clusters
p <- ggplot2::ggplot() +
  ggplot2::geom_point(ggplot2::aes(x = df[, 1], y = df[, 2], color = df$label),
                      size = 1, alpha = 1) +
  ggplot2::labs(title = title, x = "UMAP_1", y = "UMAP_2", color = "") +
  ggplot2::theme_classic(base_size = 20) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  ggplot2::guides(colour = guide_legend(override.aes = list(size = 4)))
filename <- "figures/figure_50_0021.png"
ggplot2::ggsave(file = filename, plot = p, dpi = 200, width = 5.8, height = 5)

# QC metrics
df$ncounts <- brca$nCount_RNA
p <- ggplot2::ggplot() +
  ggplot2::geom_point(ggplot2::aes(x = df[, 1], y = df[, 2], color=df$ncounts),
                      size = 1, alpha = 1) +
  ggplot2::labs(title = title, x = "UMAP_1", y = "UMAP_2", color="nCount_RNA") +
  ggplot2::scale_color_continuous() + ggplot2::theme_classic(base_size = 20) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, size = 20))
filename <- "figures/figure_50_0022.png"
ggplot2::ggsave(file = filename, plot = p, dpi = 80, width = 7, height = 5)

df$ncounts <- brca$percMT
p <- ggplot2::ggplot() +
  ggplot2::geom_point(ggplot2::aes(x = df[, 1], y = df[, 2], color=df$ncounts),
                      size = 1, alpha = 1) +
  ggplot2::labs(title = title, x = "UMAP_1", y = "UMAP_2", color = "percMT") +
  ggplot2::scale_color_continuous() + ggplot2::theme_classic(base_size = 20) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, size = 20))
filename <- "figures/figure_50_0023.png"
ggplot2::ggsave(file = filename, plot = p, dpi = 80, width = 6.2, height = 5)

## Save the data.
saveRDS(brca, file = "backup/10_003_brca_seurat.rds")

#Load the data.
#brca <- readRDS("backup/10_003_brca_seurat.rds")


# Pseudotime analysis using MERLoT
## Attach R libraries.
library(plot3D)

## Diffusion map
## Perform a dimensionality reduction using a diffusion map.
mat <- Seurat::Embeddings(object = brca, reduction = "pca")
set.seed(1)
res <- destiny::DiffusionMap(mat, sigma = "local", distance = "euclidean")

## Check the results.
df <- as.data.frame(res@eigenvectors)
df$label <- factor(brca$subtype, levels = unique(sort(brca$subtype)))
plotly::plot_ly(df, x = ~df[, 1], y = ~df[, 2], z = ~df[, 3],
                marker = list(size = 5)) %>% add_markers(color = ~df$label)

df <- as.data.frame(res@eigenvectors)[, seq_len(3)]
df$label <- factor(mcf7$type, levels = unique(sort(mcf7$type)))
colors <- as.character(df$label)
colors[which(colors == "Normal")] <- "red"
colors[which(colors == "LumA")]   <- "orange"
colors[which(colors == "LumB")]   <- "yellow"
colors[which(colors == "Her2")]   <- "green"
colors[which(colors == "Basal")]  <- "blue"
filename <- "figures/figure_50_0030.png"
png(file = filename, height = 1500, width = 1500, res = 300)
scatter3D(df[, 1], df[, 2], df[, 3],
          main = "TCGA (BRCA)", xlab = "DC_1", ylab = "DC_2", zlab = "DC_3",
          box = TRUE, bty = "b2", axes = TRUE, nticks = 5,
          theta = 230, phi = 30, pch = 16, cex = 1.0, alpha = 1.0,
          col = colors, colvar = NA, colkey = FALSE)
graphics::legend("bottomright", legend = unique(sort(df$label)),
                 pch = 16, col = unique(sort(colors)), cex = 1,
                 inset = c(-0.0))
dev.off()



<br>

################################################################################
################################################################################
################################################################################
#
# ASURAT
#
################################################################################
################################################################################
################################################################################
## Load the data.
ref <- readRDS("backup/02_002_mcf7_prep_validaton.rds")
ref <- FindVariableFeatures(ref, selection.method = "vst", nfeatures = 2000)
features = Seurat::VariableFeatures(ref)
View(t(t(features)))

mcf7 <- readRDS("backup/01_002_mcf7_dataqc.rds")
mcf7 <- mcf7[features, ]
bayout <- readRDS("backup/01_003_mcf7_baynorm.rds")
mat <- do.call("cbind", bayout[["Bay_out_list"]])
mat <- mat[features, colnames(mcf7)]
assay(mcf7, "normalized") <- mat

assay(mcf7, "logcounts") <- log(assay(mcf7, "normalized") + 1)
mat <- assay(mcf7, "logcounts")
assay(mcf7, "centered") <- sweep(mat, 1, apply(mat, 1, mean), FUN = "-")

sname <- "logcounts"
altExp(mcf7, sname) <- SummarizedExperiment(list(counts = assay(mcf7, sname)))

dictionary <- AnnotationDbi::select(org.Hs.eg.db::org.Hs.eg.db,
                                    key = rownames(mcf7),
                                    columns = "ENTREZID", keytype = "SYMBOL")
dictionary <- dictionary[!duplicated(dictionary$SYMBOL), ]
rowData(mcf7)$geneID <- dictionary$ENTREZID

saveRDS(mcf7, file = "backup/20_003_mcf7_normalized.rds")
mcf7 <- readRDS("backup/20_003_mcf7_normalized.rds")

mat <- t(as.matrix(assay(mcf7, "centered")))
cormat <- cor(mat, method = "spearman")

saveRDS(cormat, file = "backup/20_003_mcf7_cormat.rds")
cormat <- readRDS("backup/20_003_mcf7_cormat.rds")

urlpath <- "https://github.com/keita-iida/ASURATDB/blob/main/genes2bioterm/"
#load(url(paste0(urlpath, "20201213_human_GO_red.rda?raw=TRUE")))     # GO
#load(url(paste0(urlpath, "20201213_human_KEGG.rda?raw=TRUE")))       # KEGG
load("/Users/kiida/ASURATDB/genes2bioterm/20201213_human_GO_red.rda") # GO
load("/Users/kiida/ASURATDB/genes2bioterm/20201213_human_KEGG.rda") # KEGG

mcf7s <- list(GO = mcf7, KG = mcf7)
metadata(mcf7s$GO) <- list(sign = human_GO[["BP"]])
metadata(mcf7s$KG) <- list(sign = human_KEGG[["pathway"]])

mcf7s$GO <- remove_signs(sce = mcf7s$GO, min_ngenes = 2, max_ngenes = 1000)
mcf7s$KG <- remove_signs(sce = mcf7s$KG, min_ngenes = 2, max_ngenes = 1000)

set.seed(1)
mcf7s$GO <- cluster_genesets(sce = mcf7s$GO, cormat = cormat,
                             th_posi = 0.25, th_nega = -0.25)
set.seed(1)
mcf7s$KG <- cluster_genesets(sce = mcf7s$KG, cormat = cormat,
                             th_posi = 0.25, th_nega = -0.25)

mcf7s$GO <- create_signs(sce = mcf7s$GO, min_cnt_strg = 3, min_cnt_vari = 3)
mcf7s$KG <- create_signs(sce = mcf7s$KG, min_cnt_strg = 3, min_cnt_vari = 3)

simmat <- human_GO$similarity_matrix$BP
mcf7s$GO <- remove_signs_redundant(sce = mcf7s$GO, similarity_matrix = simmat,
                                   threshold = 0.80, keep_rareID = TRUE)

mcf7s$GO <- makeSignMatrix(sce = mcf7s$GO, weight_strg = 0.5, weight_vari = 0.5)
mcf7s$KG <- makeSignMatrix(sce = mcf7s$KG, weight_strg = 0.5, weight_vari = 0.5)

dbs <- c("GO", "KG")
for(i in seq_along(dbs)){
  set.seed(1)
  mat <- t(as.matrix(assay(mcf7s[[dbs[i]]], "counts")))
  res <- Rtsne::Rtsne(mat, dim = 2, pca = TRUE, initial_dims = 100)
  reducedDim(mcf7s[[dbs[i]]], "TSNE") <- res[["Y"]]
}

saveRDS(mcf7s, file = "backup/20_004_mcf7s_ssm.rds")
mcf7s <- readRDS("backup/20_004_mcf7s_ssm.rds")

resolutions <- c(0.10, 0.20)
dims <- list(seq_len(40), seq_len(30))
dbs <- c("GO", "KG")
for(i in seq_along(dbs)){
  surt <- Seurat::as.Seurat(mcf7s[[dbs[i]]], counts = "counts", data = "counts")
  mat <- as.matrix(assay(mcf7s[[dbs[i]]], "counts"))
  surt[["SSM"]] <- Seurat::CreateAssayObject(counts = mat)
  Seurat::DefaultAssay(surt) <- "SSM"
  surt <- Seurat::ScaleData(surt, features = rownames(surt))
  surt <- Seurat::RunPCA(surt, features = rownames(surt))
  surt <- Seurat::FindNeighbors(surt, reduction = "pca", dims = dims[[i]])
  surt <- Seurat::FindClusters(surt, resolution = resolutions[i])
  temp <- Seurat::as.SingleCellExperiment(surt)
  colData(mcf7s[[dbs[i]]])$seurat_clusters <- colData(temp)$seurat_clusters
}

titles <- c("MCF7 (function)", "MCF7 (pathway)")
dbs <- c("GO", "KG")
for(i in seq_along(dbs)){
  labels <- colData(mcf7s[[dbs[i]]])$seurat_clusters
  df <- as.data.frame(reducedDim(mcf7s[[dbs[i]]], "TSNE"))
  p <- ggplot2::ggplot() +
    ggplot2::geom_point(ggplot2::aes(x = df[, 1], y = df[, 2], color = labels),
                        size = 1, alpha = 1) +
    ggplot2::labs(title = titles[i], x = "tSNE_1", y = "tSNE_2", color = "") +
    ggplot2::theme_classic(base_size = 20) +
    ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5, size = 18)) +
    ggplot2::guides(colour = ggplot2::guide_legend(override.aes = list(size = 4)))
  if(i == 1){
    p <- p + ggplot2::scale_colour_brewer(palette = "Set1")
  }else if(i == 2){
    p <- p + ggplot2::scale_colour_brewer(palette = "Set2")
  }
  filename <- sprintf("figures/figure_20_%04d.png", 9 + i)
  ggplot2::ggsave(file = filename, plot = p, dpi = 200, width = 5.1, height = 4.3)
}



set.seed(1)
surt <- Seurat::as.Seurat(mcf7s$GO, counts = "counts", data = "counts")
mat <- as.matrix(assay(altExp(mcf7s$GO), "counts"))
surt[["GEM"]] <- Seurat::CreateAssayObject(counts = mat)
Seurat::DefaultAssay(surt) <- "GEM"
surt <- Seurat::SetIdent(surt, value = "seurat_clusters")
res <- Seurat::FindAllMarkers(surt, only.pos = TRUE,
                              min.pct = 0.25, logfc.threshold = 0.25)
metadata(mcf7s$GO)$marker_genes$all <- res



set.seed(1)
surt <- Seurat::as.Seurat(mcf7s$KG, counts = "counts", data = "counts")
mat <- as.matrix(assay(altExp(mcf7s$KG), "counts"))
surt[["GEM"]] <- Seurat::CreateAssayObject(counts = mat)
Seurat::DefaultAssay(surt) <- "GEM"
surt <- Seurat::SetIdent(surt, value = "seurat_clusters")
res <- Seurat::FindAllMarkers(surt, only.pos = TRUE,
                              min.pct = 0.25, logfc.threshold = 0.25)
metadata(mcf7s$KG)$marker_genes$all <- res


saveRDS(mcf7s, file = "backup/20_005_mcf7s_asurat.rds")
mcf7s <- readRDS("backup/20_005_mcf7s_asurat.rds")



#-------------------------------------------------------------------------------
# Bistable gene analysis
#-------------------------------------------------------------------------------
#Load the data.
mcf7 <- readRDS("backup/01_004_mcf7_seurat.rds")
param_ptmat_dev <- readRDS("backup/01_012_param_bifurcation_reinterpret.rds")
gene_C1 <- param_ptmat_dev[["fromC1toC1woC2C3C4C5"]]
gene_C1 <- gene_C1[which(gene_C1$NSP >= 2), ]$gene
gene_C4 <- param_ptmat_dev[["fromC4toC4woC1C2C3C5"]]
gene_C4 <- gene_C4[which(gene_C4$NSP >= 2), ]$gene
gene_C5 <- param_ptmat_dev[["fromC5toC5woC1C2C3C4"]]
gene_C5 <- gene_C5[which(gene_C5$NSP >= 2), ]$gene
genes <- unique(c(gene_C1, gene_C4, gene_C5))
mcf7 <- mcf7[genes, ]
mcf7 <- mcf7[, which(mcf7$type %in% c("TAM_3weeks", "TAM_6weeks", "TAM_9weeks"))]

# Scale data
mcf7 <- Seurat::ScaleData(mcf7)
# Principal component analysis
mcf7 <- Seurat::RunPCA(mcf7)
pc <- which(cumsum(mcf7@reductions[["pca"]]@stdev) /
              sum(mcf7@reductions[["pca"]]@stdev) > 0.5)[1]
# Create k-nearest neighbor graph.
mcf7 <- Seurat::FindNeighbors(mcf7, reduction = "pca", dim = seq_len(pc))
# Cluster cells.
mcf7 <- Seurat::FindClusters(mcf7, resolution = 0.2)
# Run UMAP.
mcf7 <- Seurat::RunUMAP(mcf7, dims = seq_len(pc))
DimPlot(mcf7)

# DEG analysis
mcf7 <- SetIdent(mcf7, value = "seurat_clusters")
mcf7@misc[["FindAllMarkers"]] <- FindAllMarkers(mcf7, only.pos = TRUE,
                                                min.pct = 0.25,
                                                logfc.threshold = 0.25)
View(mcf7@misc[["FindAllMarkers"]])

title <- "MCF-7"
df <- as.data.frame(Seurat::Embeddings(object = mcf7, reduction = "umap"))
# Seurat clusters
df$label <- mcf7$seurat_clusters
p <- ggplot2::ggplot() +
  ggplot2::geom_point(ggplot2::aes(x = df[, 1], y = df[, 2], color = df$label),
                      size = 1, alpha = 1) +
  ggplot2::labs(title = title, x = "UMAP_1", y = "UMAP_2", color = "") +
  ggplot2::theme_classic(base_size = 20) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, size = 20)) +
  ggplot2::guides(colour = guide_legend(override.aes = list(size = 4)))
#filename <- "figures/figure_01_0021.png"
#ggplot2::ggsave(file = filename, plot = p, dpi = 80, width = 5.8, height = 5)



#-------------------------------------------------------------------------------
# RPS6KB1 (C1)
#-------------------------------------------------------------------------------
gene <- "RPS6KB1"
type <- "fromC1toC1woC2C3C4C5"
f <- sprintf("gnuplot_output_%s/%s.txt", type, gene)
param <- read.table(file = f, sep = " ", header = F)
colnames(param) <- c("i", "t0", "x0", "a", "b", "c", "d", "m", "M", "n", "N", "FIT_STDFIT")
#
# For dxdt
#
f <- function(t, x, a, b, c, d, k, K, l, L, m, M, n, N){
  dxdt <- a * x**k / (x**k + K) * t**m / (t**m + M) - b * x**l / (x**l + L) * t**n / (t**n + N) + c - d * x
  return(dxdt)
}
param <- readRDS("backup/01_012_param_bifurcation_reinterpret.rds")[[type]]
param <- param[which(param$gene == gene), ]
# parameter for Error < 0.1 but almost-same sets were removed.
a <- c(5.60715891440591, 6.95543888381215, 5.61506753018533)
b <- c(2.15020365091602, 5.39487067459284, 2.15601964396678)
c <- c(4.84885574543932, 7.46757029814005, 4.84695790303724)
d <- c(4.23084311910766, 6.51883288634458, 4.22918403551746)
k <- c(1.78894158244902, 3.76840094787012, 1.00350094487201)
K <- c(-1.22798419760446e-05, 7.97131164843362, 0.0834288708205815)
l <- c(0.702149159896281, 0.511961688403384, 0.954224487375553)
L <- c(0.00946306873781471, 2.55742777986908e-05, 0.0924617650894001)
m <- c(1.23230444274633, 1.45311566146895, 1.23220267305097)
M <- c(0.94475969546108, -0.226462152616938, 0.944063531188496)
n <- c(2.08887507376226, 1.68963158329994, 2.08861618333145)
N <- c(0.094953344063101, 0.162842567041471, 0.0949653581438872)
df <- data.frame(a, b, c, d, k, K, l, L, m, M, n, N)
View(df)
t0 <- param$t0
x0 <- param$x0
a <- a**2
b <- b**2
c <- c**2
d <- d**2
k <- k**2
K <- K**2
l <- l**2
L <- L**2
m <- m**2
M <- M**2
n <- n**2
N <- N**2

tstep <- 4000
dt <- 0.005
potential <- data.frame(t = c(t0, t0 + dt * seq_len(tstep)))
for(kk in seq_along(a)){
  poten <- c()
  point <- c()
  for(i in seq_len(nrow(potential))){
    t <- potential[i, 1]
    if(i == 1){
      xx <- x0
    }else{
      xx <- xx + dt * f(t, xx, a[kk], b[kk], c[kk], d[kk], k[kk], K[kk], l[kk], L[kk], m[kk], M[kk], n[kk], N[kk])
    }
    point <- rbind(point, c(t, xx))
  }
  point <- as.data.frame(point)
  colnames(point) <- c("t", "x")
  
  dg <- point
  colnames(dg)[1] <- "#t"
  filename <- sprintf("RPS6KB1_paper/point_%s_%s_param_%d.txt", gene, type, kk)
  write.table(dg, file = filename,
              sep = "\t", row.names = FALSE, quote = FALSE)
}

#-------------------------------------------------------------------------------
# RPS6KB1 (C4)
#-------------------------------------------------------------------------------
gene <- "RPS6KB1"
type <- "fromC4toC4woC1C2C3C5"
f <- sprintf("gnuplot_output_%s/%s.txt", type, gene)
param <- read.table(file = f, sep = " ", header = F)
colnames(param) <- c("i", "t0", "x0", "a", "b", "c", "d", "m", "M", "n", "N", "FIT_STDFIT")
#
# For dxdt
#
f <- function(t, x, a, b, c, d, k, K, l, L, m, M, n, N){
  dxdt <- a * x**k / (x**k + K) * t**m / (t**m + M) - b * x**l / (x**l + L) * t**n / (t**n + N) + c - d * x
  return(dxdt)
}
param <- readRDS("backup/01_012_param_bifurcation_reinterpret.rds")[[type]]
param <- param[which(param$gene == gene), ]
# parameter for Error < 0.02
a <- c(1.55102726910428, 1.42450342750481, 1.56980810163961,
       1.44978650251387, 6.02261299880295)
b <- c(1.86000905851167, 1.49436980483143, 1.64014513800044,
       1.53022023583461, 6.05781592557455)
c <- c(1.26887676791826, 1.1643141472431, 1.45240227365472,
       0.934386068297315, 0.741319264401254)
d <- c(1.00488710241844, 0.922234474826163, 1.14953157731569,
       0.74046047381832, 0.58943038881012)
k <- c(0.966748776399256, 0.997039232592534, 0.0998695817723216,
       0.0999805689622979, 0.0956449191452329)
K <- c(0.499652316494465, 0.0990664833871207, 0.65762384562846,
       0.0995849709856449, 0.0777820902598886)
l <- c(1.70304155275077, 1.00840412299607, 0.100962344277414,
       0.100031998579688, 0.106177092712617)
L <- c(1.37117609253806, 0.10079076895992, 0.627611544969449,
       0.0993929234503134, 0.0909127649141223)
m <- c(4.40956940961314, 4.43837542580305, 4.50636217336347,
       4.44268483615653, 4.00427541761679)
M <- c(0.358664086000866, 0.351459981665446, 0.340153332628892,
       0.356681317639938, 0.453978338445044)
n <- c(3.64819539640447, 3.63345552907139, 3.59456849446045,
       3.61244498284406, 3.94709443926924)
N <- c(0.454740904985079, 0.455899300783156, 0.448001830073055,
       0.48104671693544, 0.464835755531639)
df <- data.frame(a, b, c, d, k, K, l, L, m, M, n, N)
View(df)
t0 <- param$t0
x0 <- param$x0
a <- a**2
b <- b**2
c <- c**2
d <- d**2
k <- k**2
K <- K**2
l <- l**2
L <- L**2
m <- m**2
M <- M**2
n <- n**2
N <- N**2

tstep <- 4000
dt <- 0.005
potential <- data.frame(t = c(t0, t0 + dt * seq_len(tstep)))
for(kk in seq_along(a)){
  poten <- c()
  point <- c()
  for(i in seq_len(nrow(potential))){
    t <- potential[i, 1]
    if(i == 1){
      xx <- x0
    }else{
      xx <- xx + dt * f(t, xx, a[kk], b[kk], c[kk], d[kk], k[kk], K[kk], l[kk], L[kk], m[kk], M[kk], n[kk], N[kk])
    }
    point <- rbind(point, c(t, xx))
  }
  point <- as.data.frame(point)
  colnames(point) <- c("t", "x")
  
  dg <- point
  colnames(dg)[1] <- "#t"
  filename <- sprintf("RPS6KB1_paper/point_%s_%s_param_%d.txt", gene, type, kk)
  write.table(dg, file = filename,
              sep = "\t", row.names = FALSE, quote = FALSE)
}

#-------------------------------------------------------------------------------
# RPS6KB1 (C5)
#-------------------------------------------------------------------------------
gene <- "RPS6KB1"
type <- "fromC5toC5woC1C2C3C4"
f <- sprintf("gnuplot_output_%s/%s.txt", type, gene)
param <- read.table(file = f, sep = " ", header = F)
colnames(param) <- c("i", "t0", "x0", "a", "b", "c", "d", "m", "M", "n", "N", "FIT_STDFIT")
#
# For dxdt
#
f <- function(t, x, a, b, c, d, k, K, l, L, m, M, n, N){
  dxdt <- a * x**k / (x**k + K) * t**m / (t**m + M) - b * x**l / (x**l + L) * t**n / (t**n + N) + c - d * x
  return(dxdt)
}
param <- readRDS("backup/01_012_param_bifurcation_reinterpret.rds")[[type]]
param <- param[which(param$gene == gene), ]
# parameter for Error < 0.02
a <- c(9.2945855784555, 9.85673528178194, 6.93173890263515)
b <- c(5.96511014137327, 6.6777893133142, 6.42596508163344)
c <- c(0.0185066320311762, 0.000750263815286677, 1.22118745244898)
d <- c(4.15606498000519, 4.11406595645723, 2.4952615977341)
k <- c(1.95273251062397, 1.93642122479491, 0.698560412833255)
K <- c(1.65657802565775, 1.66805518105509, 0.0494138023301595)
l <- c(2.81809777289628, 2.68538495609896, 2.48449767755541)
L <- c(2.61787383416711, 2.35379197229098, 0.808539866499963)
m <- c(2.13932050985669, 2.18783087898127, 1.98753053855417)
M <- c(0.198661613553677, 0.19539659791799, 0.471231807757278)
n <- c(2.67122795713133, 2.63672290330624, 2.45701605705581)
N <- c(0.169897265080544, 0.170596637296396, 0.380769354862149)
df <- data.frame(a, b, c, d, k, K, l, L, m, M, n, N)
View(df)
t0 <- param$t0
x0 <- param$x0
a <- a**2
b <- b**2
c <- c**2
d <- d**2
k <- k**2
K <- K**2
l <- l**2
L <- L**2
m <- m**2
M <- M**2
n <- n**2
N <- N**2

tstep <- 4000
dt <- 0.005
potential <- data.frame(t = c(t0, t0 + dt * seq_len(tstep)))
for(kk in seq_along(a)){
  poten <- c()
  point <- c()
  for(i in seq_len(nrow(potential))){
    t <- potential[i, 1]
    if(i == 1){
      xx <- x0
    }else{
      xx <- xx + dt * f(t, xx, a[kk], b[kk], c[kk], d[kk], k[kk], K[kk], l[kk], L[kk], m[kk], M[kk], n[kk], N[kk])
    }
    point <- rbind(point, c(t, xx))
  }
  point <- as.data.frame(point)
  colnames(point) <- c("t", "x")
  
  dg <- point
  colnames(dg)[1] <- "#t"
  filename <- sprintf("RPS6KB1_paper/point_%s_%s_param_%d.txt", gene, type, kk)
  write.table(dg, file = filename,
              sep = "\t", row.names = FALSE, quote = FALSE)
}